<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>geneflow.geneflow &mdash; GeneFlow 0.8 documentation</title><link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  <script id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/language_data.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html"><img src="../../_static/geneflow_logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption"><span class="caption-text">GeneFlow:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../geneflow.html">GeneFlow</a></li>
</ul>
<p class="caption"><span class="caption-text">Modules:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../src.html">DataObject</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../src.html#module-src.etl">ETL</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../src.html#module-src.model">Model</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../src.html#module-src.objects">Objects</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../src.html#module-src.processing">Processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../src.html#module-src.utils">Utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../src.html#module-src.visualize">Visualize</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">GeneFlow</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../index.html">Module code</a> &raquo;</li>
      <li>geneflow.geneflow</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for geneflow.geneflow</h1><div class="highlight"><pre>
<span></span><span class="c1"># -*- coding: utf-8 -*-</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">@author: Alba Casillas Rodr√≠guez (albacaro@correo.ugr.es)</span>

<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">from</span> <span class="nn">src</span> <span class="kn">import</span> <span class="n">utils</span> <span class="k">as</span> <span class="n">ut</span>
<span class="kn">from</span> <span class="nn">src</span> <span class="kn">import</span> <span class="n">processing</span> <span class="k">as</span> <span class="n">pro</span>
<span class="kn">from</span> <span class="nn">src</span> <span class="kn">import</span> <span class="n">dataobject</span> <span class="k">as</span> <span class="n">dobj</span>
<span class="kn">from</span> <span class="nn">src</span> <span class="kn">import</span> <span class="n">visualize</span> <span class="k">as</span> <span class="n">vis</span>
<span class="kn">from</span> <span class="nn">src.processing</span> <span class="kn">import</span> <span class="n">Task</span>
<span class="kn">from</span> <span class="nn">src.dataobject</span> <span class="kn">import</span> <span class="n">DataObject</span>
<span class="kn">from</span> <span class="nn">src.etl</span> <span class="kn">import</span> <span class="n">ProcessGDC</span>
<span class="kn">from</span> <span class="nn">src.objects</span> <span class="kn">import</span> <span class="n">GDCQuery</span>
<span class="kn">from</span> <span class="nn">src.objects</span> <span class="kn">import</span> <span class="n">DataProject</span>
<span class="kn">from</span> <span class="nn">src.objects</span> <span class="kn">import</span> <span class="n">FileProject</span>
<span class="kn">from</span> <span class="nn">src.objects</span> <span class="kn">import</span> <span class="n">Workflow</span>
<span class="kn">from</span> <span class="nn">src.model</span> <span class="kn">import</span> <span class="n">ModelSelection</span>
<span class="kn">from</span> <span class="nn">src.model</span> <span class="kn">import</span> <span class="n">LogisticRegression</span>
<span class="kn">from</span> <span class="nn">src.model</span> <span class="kn">import</span> <span class="n">RandomForest</span>
<span class="kn">from</span> <span class="nn">src.model</span> <span class="kn">import</span> <span class="n">ScikitLearnModel</span>
<span class="kn">from</span> <span class="nn">src.model</span> <span class="kn">import</span> <span class="n">SupportVectorClassif</span>
<span class="kn">from</span> <span class="nn">src.model</span> <span class="kn">import</span> <span class="n">NeuralNetwork</span>



<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">######################################################</span>

<span class="sd">                    TASK</span>

<span class="sd">######################################################</span>
<span class="sd">&quot;&quot;&quot;</span>


<div class="viewcode-block" id="from_dict"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.from_dict">[docs]</a><span class="k">def</span> <span class="nf">from_dict</span><span class="p">(</span><span class="n">json_task</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns a object Task replicate from json or dictionary</span>

<span class="sd">    :return: a object Task replicated</span>
<span class="sd">    :rtype: Task</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">Task</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">json_task</span><span class="p">)</span></div>


<div class="viewcode-block" id="workflow_to_json"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.workflow_to_json">[docs]</a><span class="k">def</span> <span class="nf">workflow_to_json</span><span class="p">(</span><span class="n">workfw</span> <span class="p">:</span> <span class="n">Workflow</span><span class="p">,</span> <span class="n">path</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Generate an external file to save every Log object of Workflow</span>

<span class="sd">    :param path: file where dictionary generates will be saved</span>
<span class="sd">    :type: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">workfw</span><span class="o">.</span><span class="n">generate_json</span><span class="p">(</span><span class="n">path</span><span class="p">)</span></div>


<div class="viewcode-block" id="workflow_from_json"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.workflow_from_json">[docs]</a><span class="k">def</span> <span class="nf">workflow_from_json</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Open a json file with some Log saved as dictionary</span>
<span class="sd">        and creates a new Workflow throught them</span>

<span class="sd">    :param path: The file that contains object Task as dictionary for lines</span>
<span class="sd">    :type: str</span>

<span class="sd">    :return: a replicate Workflow from json dictionary</span>
<span class="sd">    :rtype: Workflow    </span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="n">flow</span> <span class="o">=</span> <span class="n">Workflow</span><span class="p">()</span>
    <span class="n">opened</span> <span class="o">=</span> <span class="s2">&quot;workflow.json&quot;</span>

    <span class="k">if</span> <span class="n">path</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
        <span class="n">opened</span> <span class="o">=</span> <span class="n">path</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">opened</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
        <span class="n">task_json</span> <span class="o">=</span> <span class="n">ut</span><span class="o">.</span><span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
        
        <span class="k">for</span> <span class="n">tk</span> <span class="ow">in</span> <span class="n">task_json</span><span class="p">:</span>
            <span class="n">flow_task</span> <span class="o">=</span> <span class="n">from_dict</span><span class="p">(</span><span class="n">tk</span><span class="p">)</span>

            <span class="n">flow</span><span class="o">.</span><span class="n">add_function</span><span class="p">(</span><span class="n">flow_task</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">flow</span></div>
    

<div class="viewcode-block" id="replicate_workflow"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.replicate_workflow">[docs]</a><span class="k">def</span> <span class="nf">replicate_workflow</span><span class="p">(</span><span class="n">workflow</span><span class="p">,</span> <span class="n">data_object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Replicate a workflow to obtain the same result as previous experiment</span>

<span class="sd">    :param workflow: a flow of another object that contains step of its experiment</span>
<span class="sd">    :type workflow: Workflow</span>
<span class="sd">    :param data_object: a DataObject to which the same process will be applied through the Workflow</span>
<span class="sd">    :type data_object: DataObject</span>

<span class="sd">    :return: a DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">workflow</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data_object</span><span class="p">)</span></div>



<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">######################################################</span>

<span class="sd">                    VISUALIZE</span>

<span class="sd">######################################################</span>
<span class="sd">&quot;&quot;&quot;</span>


<div class="viewcode-block" id="show_figure"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.show_figure">[docs]</a><span class="k">def</span> <span class="nf">show_figure</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="n">xlabel</span> <span class="o">=</span> <span class="s2">&quot;x axis&quot;</span><span class="p">,</span> <span class="n">ylabel</span> <span class="o">=</span> <span class="s2">&quot;y axis&quot;</span><span class="p">,</span> <span class="n">x_ini_range</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">x_fin_range</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                    <span class="n">y_ini_range</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">y_fin_range</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">legend</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">legend_title</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> 
                    <span class="n">vheight</span> <span class="o">=</span> <span class="mi">600</span><span class="p">,</span> <span class="n">vwidth</span> <span class="o">=</span> <span class="mi">800</span><span class="p">,</span> <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Plot of the figure&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Show in the screen the figure plot with the specified figured parameters.</span>

<span class="sd">    :param fig: Figure to update.</span>
<span class="sd">    :type fig: Figure</span>
<span class="sd">    :param xlabel: Name of the X axis. Defaults to `x axis`.</span>
<span class="sd">    :type xlabel: str</span>
<span class="sd">    :param ylabel: Name of the Y axis. Defaults to `y axis`.</span>
<span class="sd">    :type ylabel: str</span>
<span class="sd">    :param x_ini_range: Initial value of X axis range. Defaults to None.</span>
<span class="sd">    :type x_ini_range: int</span>
<span class="sd">    :param x_fin_range: Final value of X axis range. Defaults to None.</span>
<span class="sd">    :type x_fin_range: int</span>
<span class="sd">    :param y_ini_range: Initial value of Y axis range. Defaults to None.</span>
<span class="sd">    :type y_ini_range: int</span>
<span class="sd">    :param y_fin_range: Final value of Y axis range. Defaults to None.</span>
<span class="sd">    :type y_fin_range: int</span>
<span class="sd">    :param legend: Indicates if the figure will have legend or not. Defaults to `True`.</span>
<span class="sd">    :type legend: bool</span>
<span class="sd">    :param legend_title: Name of the legend&#39;s label</span>
<span class="sd">    :type legend_title: str</span>
<span class="sd">    :param vheight: Height of the figure as image. Defaults to 600.</span>
<span class="sd">    :type vheight: int</span>
<span class="sd">    :param vwidth: Width of the figure as image. Defaults to 800.</span>
<span class="sd">    :type vwidth: int</span>
<span class="sd">    :param title: Title of the figure. Defaults to `Plot of the figure`.</span>
<span class="sd">    :type title: str</span>

<span class="sd">    :return: The figure with the layout&#39;s features updated.</span>
<span class="sd">    :rtype: Figure</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">vis</span><span class="o">.</span><span class="n">show_figure</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="n">xlabel</span> <span class="o">=</span> <span class="n">xlabel</span><span class="p">,</span> <span class="n">ylabel</span> <span class="o">=</span> <span class="n">ylabel</span><span class="p">,</span> <span class="n">x_ini_range</span> <span class="o">=</span> <span class="n">x_ini_range</span><span class="p">,</span> <span class="n">x_fin_range</span> <span class="o">=</span> <span class="n">x_fin_range</span><span class="p">,</span>
                    <span class="n">y_ini_range</span> <span class="o">=</span> <span class="n">y_ini_range</span><span class="p">,</span> <span class="n">y_fin_range</span> <span class="o">=</span> <span class="n">y_fin_range</span><span class="p">,</span> <span class="n">legend</span> <span class="o">=</span> <span class="n">legend</span><span class="p">,</span> <span class="n">legend_title</span> <span class="o">=</span> <span class="n">legend_title</span><span class="p">,</span>
                    <span class="n">vheight</span> <span class="o">=</span> <span class="n">vheight</span><span class="p">,</span> <span class="n">vwidth</span> <span class="o">=</span> <span class="n">vwidth</span><span class="p">,</span> <span class="n">title</span> <span class="o">=</span> <span class="n">title</span><span class="p">)</span></div>


<div class="viewcode-block" id="save_image"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.save_image">[docs]</a><span class="k">def</span> <span class="nf">save_image</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="n">fig_name</span> <span class="o">=</span> <span class="s2">&quot;fig&quot;</span><span class="p">,</span> <span class="n">img_format</span> <span class="o">=</span> <span class="s2">&quot;png&quot;</span><span class="p">,</span> <span class="n">os_path</span> <span class="o">=</span> <span class="s2">&quot;images&quot;</span><span class="p">,</span> 
                   <span class="n">xlabel</span> <span class="o">=</span> <span class="s2">&quot;x axis&quot;</span><span class="p">,</span> <span class="n">ylabel</span> <span class="o">=</span> <span class="s2">&quot;y axis&quot;</span><span class="p">,</span> <span class="n">x_ini_range</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">x_fin_range</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                   <span class="n">y_ini_range</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">y_fin_range</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">legend</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">legend_title</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">vheight</span> <span class="o">=</span> <span class="mi">600</span><span class="p">,</span> <span class="n">vwidth</span> <span class="o">=</span> <span class="mi">800</span><span class="p">,</span> 
                   <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Plot of the figure&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Save a figure as image on disk.</span>

<span class="sd">    :param fig: Figure to update.</span>
<span class="sd">    :type fig: Figure</span>
<span class="sd">    :param fig_name: Name of file.</span>
<span class="sd">    :type fig_name: str</span>
<span class="sd">    :param img_format: Format of the file saved. Available formats: &#39;png&#39;, &#39;jpeg&#39;, </span>
<span class="sd">        &#39;webp&#39;, &#39;svg&#39;, &#39;pdf&#39;.</span>
<span class="sd">    :type img_format: str</span>
<span class="sd">    :param os_path: Path to save the image.</span>
<span class="sd">    :type os_path: str</span>
<span class="sd">    :param xlabel: Name of the X axis. Defaults to `x axis`.</span>
<span class="sd">    :type xlabel: str</span>
<span class="sd">    :param ylabel: Name of the Y axis. Defaults to `y axis`.</span>
<span class="sd">    :type ylabel: str</span>
<span class="sd">    :param x_ini_range: Initial value of X axis range. Defaults to None.</span>
<span class="sd">    :type x_ini_range: int</span>
<span class="sd">    :param x_fin_range: Final value of X axis range. Defaults to None.</span>
<span class="sd">    :type x_fin_range: int</span>
<span class="sd">    :param y_ini_range: Initial value of Y axis range. Defaults to None.</span>
<span class="sd">    :type y_ini_range: int</span>
<span class="sd">    :param y_fin_range: Final value of Y axis range. Defaults to None.</span>
<span class="sd">    :type y_fin_range: int</span>
<span class="sd">    :param legend: Indicates if the figure will have legend or not. Defaults to `True`.</span>
<span class="sd">    :type legend: bool</span>
<span class="sd">    :param legend_title: Name of the legend&#39;s label</span>
<span class="sd">    :type legend_title: str</span>
<span class="sd">    :param vheight: Height of the figure as image. Defaults to 600.</span>
<span class="sd">    :type vheight: int</span>
<span class="sd">    :param vwidth: Width of the figure as image. Defaults to 800.</span>
<span class="sd">    :type vwidth: int</span>
<span class="sd">    :param title: Title of the figure. Defaults to `Plot of the figure`.</span>
<span class="sd">    :type title: str</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">vis</span><span class="o">.</span><span class="n">save_image</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="n">fig_name</span> <span class="o">=</span> <span class="n">fig_name</span><span class="p">,</span> <span class="n">img_format</span> <span class="o">=</span> <span class="n">img_format</span><span class="p">,</span> <span class="n">os_path</span> <span class="o">=</span> <span class="n">os_path</span><span class="p">,</span> 
                    <span class="n">xlabel</span> <span class="o">=</span> <span class="n">xlabel</span><span class="p">,</span> <span class="n">ylabel</span> <span class="o">=</span> <span class="n">ylabel</span><span class="p">,</span> <span class="n">x_ini_range</span> <span class="o">=</span> <span class="n">x_ini_range</span><span class="p">,</span> <span class="n">x_fin_range</span> <span class="o">=</span> <span class="n">x_fin_range</span><span class="p">,</span>
                    <span class="n">y_ini_range</span> <span class="o">=</span> <span class="n">y_ini_range</span><span class="p">,</span> <span class="n">y_fin_range</span> <span class="o">=</span> <span class="n">y_fin_range</span><span class="p">,</span> <span class="n">legend</span> <span class="o">=</span> <span class="n">legend</span><span class="p">,</span> <span class="n">legend_title</span> <span class="o">=</span> <span class="n">legend_title</span><span class="p">,</span>
                    <span class="n">vheight</span> <span class="o">=</span> <span class="n">vheight</span><span class="p">,</span> <span class="n">vwidth</span> <span class="o">=</span> <span class="n">vwidth</span><span class="p">,</span> <span class="n">title</span> <span class="o">=</span> <span class="n">title</span><span class="p">)</span></div>


<div class="viewcode-block" id="show_image_web"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.show_image_web">[docs]</a><span class="k">def</span> <span class="nf">show_image_web</span><span class="p">(</span><span class="n">fig</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Show an image on a web browser. If there is not an available web browser,</span>
<span class="sd">        it shows the image on the default application to visualize images.</span>

<span class="sd">    :param fig: Figure to dislay.</span>
<span class="sd">    :type fig: Figure</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">vis</span><span class="o">.</span><span class="n">show_image_web</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span></div>


<div class="viewcode-block" id="bar_plot"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.bar_plot">[docs]</a><span class="k">def</span> <span class="nf">bar_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">x</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Create a bar diagram with the data received as the input.</span>

<span class="sd">    :param data: Data values to create the diagram.</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param x: Name of X axis. Defaults to None.</span>
<span class="sd">    :type x: str</span>
<span class="sd">    :param y: Name of Y axis. Defaults to None.</span>
<span class="sd">    :type y: str</span>
<span class="sd">    </span>
<span class="sd">    :return: Figure that contains the bar plot</span>
<span class="sd">    :rtype: Figure</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">vis</span><span class="o">.</span><span class="n">bar_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">y</span><span class="p">)</span></div>


<div class="viewcode-block" id="kde_plot"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.kde_plot">[docs]</a><span class="k">def</span> <span class="nf">kde_plot</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Creates a Kernel Density Estimation (kde) diagram with the data received as the input.</span>
<span class="sd">        A kernel density estimate (KDE) plot is a method for visualizing the distribution </span>
<span class="sd">        of observations in a dataset, analagous to a histogram.</span>

<span class="sd">    :param data: Data values to create the diagram.</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    </span>
<span class="sd">    :return: Figure that contains the KDE plot.</span>
<span class="sd">    :rtype: Figure</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">vis</span><span class="o">.</span><span class="n">kde_plot</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="box_plot"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.box_plot">[docs]</a><span class="k">def</span> <span class="nf">box_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">x</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">median</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Creates a box plot with the data received as the input.</span>
<span class="sd">        If &quot;median&quot; parameter is true, the function will display a line with the median value.</span>
<span class="sd">        If x and y are not None, they must be name of row and column of the data.</span>

<span class="sd">    :param data: The data used to create the plot.</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param x: Name of X axis. Defaults to None.</span>
<span class="sd">    :type x: str</span>
<span class="sd">    :param y: Name of Y axis. Defaults to None.</span>
<span class="sd">    :type y: str</span>
<span class="sd">    :param median: Indicates if the diagram will show a median line. </span>
<span class="sd">        Defaults to `False`.</span>
<span class="sd">    :type median: bool</span>

<span class="sd">    :return: Figure that contains the box plot.</span>
<span class="sd">    :rtype: Figure</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">vis</span><span class="o">.</span><span class="n">box_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">y</span><span class="p">,</span> <span class="n">median</span> <span class="o">=</span> <span class="n">median</span><span class="p">)</span></div>



<div class="viewcode-block" id="mds_plot"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.mds_plot">[docs]</a><span class="k">def</span> <span class="nf">mds_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">color_list</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">clinical_info</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">symbols</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">text_plot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Creates a multidimensional diagram plot with the data received as the input.</span>
<span class="sd">    If &quot;color_list&quot; parameter is empty, colors will be formed automatically by function,</span>
<span class="sd">    otherwise, lenght of them must be the same as lenght of data columns</span>
<span class="sd">    If &quot;text_plot&quot; is true, the name of the columns will appear over the point its point on the plot,</span>
<span class="sd">    if False will only appear the point</span>

<span class="sd">    :param data: The data used to create the plot.</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param color_list: A list of colors to be used on plot</span>
<span class="sd">    :type color_list: list</span>
<span class="sd">    :param clinical_info: A classification of the samples to represent them by types</span>
<span class="sd">    :type clinical_info: list</span>
<span class="sd">    :param symbols: A classification of the samples to represent them by symbols</span>
<span class="sd">    :type symbols: list</span>
<span class="sd">    :param text_plot: Indicates if name of column should appear over its representation on plot</span>
<span class="sd">    :type text_plot: bool</span>

<span class="sd">    :return: Figure that contains the mds plot.</span>
<span class="sd">    :rtype: Figure</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">vis</span><span class="o">.</span><span class="n">mds_plot</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">,</span> <span class="n">color_list</span> <span class="o">=</span> <span class="n">color_list</span><span class="p">,</span> <span class="n">clinical_info</span> <span class="o">=</span> <span class="n">clinical_info</span><span class="p">,</span> <span class="n">symbols</span> <span class="o">=</span> <span class="n">symbols</span><span class="p">,</span> <span class="n">text_plot</span> <span class="o">=</span> <span class="n">text_plot</span><span class="p">)</span></div>



<div class="viewcode-block" id="volcano_plot"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.volcano_plot">[docs]</a><span class="k">def</span> <span class="nf">volcano_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">clinical_data</span><span class="p">,</span> <span class="n">grouped_by</span><span class="p">,</span> <span class="n">group_1</span><span class="p">,</span> <span class="n">group_2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Creates a volcano plot with the data received as the input.</span>
<span class="sd">        To do a volcano plot is necessary to indicates two types of information</span>
<span class="sd">        collected in clinical data.</span>
<span class="sd">        It will be neccesary calculated a stadistic information</span>

<span class="sd">    :param data: The data used to create the plot.</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param clinical_data: A dataframe with clinical information of the count matrix</span>
<span class="sd">    :type clinical_data: DataFrame</span>
<span class="sd">    :param grouped_by: Name of the clinical data column to organize the plot</span>
<span class="sd">    :type grouped_by: str</span>
<span class="sd">    :param group_1: A type of the clinical data column selected</span>
<span class="sd">    :type group_1: str</span>
<span class="sd">    :param group_2: A second type of the clinical data column selected</span>
<span class="sd">    :type group_2: str</span>

<span class="sd">    :return: Figure that contains the volcano plot.</span>
<span class="sd">    :rtype: Figure</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">vis</span><span class="o">.</span><span class="n">volcano_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">clinical_data</span><span class="p">,</span> <span class="n">grouped_by</span><span class="p">,</span> <span class="n">group_1</span><span class="p">,</span> <span class="n">group_2</span><span class="p">)</span></div>


<div class="viewcode-block" id="heatmap"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.heatmap">[docs]</a><span class="k">def</span> <span class="nf">heatmap</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">z</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s1">&#39;YlOrRd&#39;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Create a heatmap with the data received as the input.</span>

<span class="sd">    :param x: The data to create the heatmap.</span>
<span class="sd">    :type x: list</span>
<span class="sd">    :param y: The data to create the heatmap.</span>
<span class="sd">    :type y: list</span>
<span class="sd">    :param z: The data to create the heatmap.</span>
<span class="sd">    :type z: list</span>
<span class="sd">    :param color: color spectrum used on heatmap</span>
<span class="sd">    :type color: str</span>
<span class="sd">    </span>
<span class="sd">    :return: Figure that contains the heatmap.</span>
<span class="sd">    :rtype: Figure</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">vis</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">z</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="n">color</span><span class="p">)</span></div>


<div class="viewcode-block" id="clustermap"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.clustermap">[docs]</a><span class="k">def</span> <span class="nf">clustermap</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s1">&#39;YlOrRd&#39;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Create a clustermap with the data received as the input.</span>
<span class="sd">        A clustermap order data by similarity. </span>
<span class="sd">        This reorganizes the data for the rows and columns and displays </span>
<span class="sd">        similar content next to one another for even more </span>
<span class="sd">        depth of understanding the data.</span>

<span class="sd">    :param data: The data to create the clustermap.</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param color: color spectrum used on heatmap of clustermap</span>
<span class="sd">    :type color: str</span>
<span class="sd">    </span>
<span class="sd">    :return: Figure that contains the clustermap.</span>
<span class="sd">    :rtype: Figure</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">vis</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="duplicated_corr_plot"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.duplicated_corr_plot">[docs]</a><span class="k">def</span> <span class="nf">duplicated_corr_plot</span><span class="p">(</span><span class="n">data_duplicated</span><span class="p">,</span> <span class="n">ini_row</span> <span class="o">=</span> <span class="mi">1001</span><span class="p">,</span> <span class="n">fin_row</span> <span class="o">=</span> <span class="mi">2000</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Create a diagram to display the correlation between duplicate variables.</span>
<span class="sd">        In case the original dataFrame is set as the input, the method will be check</span>
<span class="sd">        for duplicates. If the DataFrame contains only the duplicates, this checking</span>
<span class="sd">        will not affect the result.</span>
<span class="sd">        The layout wil be created the most squared and symetric as possible.</span>

<span class="sd">    :param data_duplicated: The duplicated data used to create the diagram</span>
<span class="sd">    :type data_duplicated: DataFrame</span>
<span class="sd">    :param ini_row: First row to select the data will be used. Defaults to 1001.</span>
<span class="sd">    :type ini_row: int</span>
<span class="sd">    :param fin_row: Final row to select the data will be used. Dafults to 2000.</span>
<span class="sd">    :type fin_row: int</span>
<span class="sd">    </span>
<span class="sd">    :return: figure that contains a diagram with duplicated data of a correlation</span>
<span class="sd">    :rtype: figure</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">vis</span><span class="o">.</span><span class="n">duplicated_corr_plot</span><span class="p">(</span><span class="n">data_duplicated</span><span class="p">,</span> <span class="n">ini_row</span> <span class="o">=</span> <span class="n">ini_row</span><span class="p">,</span> <span class="n">fin_row</span> <span class="o">=</span> <span class="n">fin_row</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                      MACHINE LEARNING VISUALIZATIONS</span>
<span class="c1"># =============================================================================  </span>



<div class="viewcode-block" id="plot_prec_recall_vs_thresh"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.plot_prec_recall_vs_thresh">[docs]</a><span class="k">def</span> <span class="nf">plot_prec_recall_vs_thresh</span><span class="p">(</span><span class="n">testy</span><span class="p">,</span> <span class="n">predictions</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Creates the curve between Precision-Real and Thresholds.</span>

<span class="sd">    :param testy: Real value of a dataset if  a ML algorithm has been applied.</span>
<span class="sd">    :type testy: list</span>
<span class="sd">    :param predictions: Result of apply ML on the dataset.</span>
<span class="sd">    :type predictions: list</span>

<span class="sd">    :return: Figure that contains the plot showing the precision recall curve over thresholds.</span>
<span class="sd">    :rtype: Figure</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">vis</span><span class="o">.</span><span class="n">plot_prec_recall_vs_thresh</span><span class="p">(</span><span class="n">testy</span><span class="p">,</span> <span class="n">predictions</span><span class="p">)</span></div>


<div class="viewcode-block" id="plot_roc"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.plot_roc">[docs]</a><span class="k">def</span> <span class="nf">plot_roc</span><span class="p">(</span><span class="n">testy</span><span class="p">,</span> <span class="n">predictions</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Creates a ROC Curve.</span>

<span class="sd">    :param testy: Real value of a dataset if  a ML algorithm has been applied.</span>
<span class="sd">    :type testy: list</span>
<span class="sd">    :param predictions: Result of apply ML on the dataset.</span>
<span class="sd">    :type predictions: list</span>

<span class="sd">    :return: Figure that contains the ROC Curve of the Real Value and Predictions.</span>
<span class="sd">    :rtype: Figure</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">vis</span><span class="o">.</span><span class="n">plot_roc</span><span class="p">(</span><span class="n">testy</span><span class="p">,</span> <span class="n">predictions</span><span class="p">)</span></div>


<div class="viewcode-block" id="plot_prc"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.plot_prc">[docs]</a><span class="k">def</span> <span class="nf">plot_prc</span><span class="p">(</span><span class="n">testy</span><span class="p">,</span> <span class="n">predictions</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Creates a Precision-Recall (PR) Curve.</span>

<span class="sd">    :param testy: Real value of a dataset if  a ML algorithm has been applied.</span>
<span class="sd">    :type testy: list</span>
<span class="sd">    :param predictions: Result of apply ML on the dataset.</span>
<span class="sd">    :type predictions: list</span>
<span class="sd">    </span>
<span class="sd">    :return: Figure that contains the PR Curve of the Real Value and Predictions.</span>
<span class="sd">    :rtype: Figure</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">vis</span><span class="o">.</span><span class="n">plot_prc</span><span class="p">(</span><span class="n">testy</span><span class="p">,</span> <span class="n">predictions</span><span class="p">)</span></div>


<div class="viewcode-block" id="plot_confusion_matrix"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.plot_confusion_matrix">[docs]</a><span class="k">def</span> <span class="nf">plot_confusion_matrix</span><span class="p">(</span><span class="n">matrix</span><span class="p">,</span> <span class="n">label_list</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s2">&quot;Viridis&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Create a figure with the Confusion Matriz as Heatmap.</span>

<span class="sd">    :param matrix: Contains the elements and data of a confusion matrix.</span>
<span class="sd">    :type matrix: list[list]</span>
<span class="sd">    :param label_list: Used as value of axis. Defaults to None.</span>
<span class="sd">    :type label_list: list</span>
<span class="sd">    :param color: color spectrum used on heatmap. Defaults to `Viridis`.</span>
<span class="sd">    :type color: list</span>
<span class="sd">    </span>
<span class="sd">    :return: Figure that contains the heatmap showing the colored confusion matrix.</span>
<span class="sd">    :rtype: Figure</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">vis</span><span class="o">.</span><span class="n">plot_confusion_matrix</span><span class="p">(</span><span class="n">matrix</span><span class="p">,</span> <span class="n">label_list</span> <span class="o">=</span> <span class="n">label_list</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="n">color</span><span class="p">)</span></div>



<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">######################################################</span>

<span class="sd">                    PROCESSING</span>

<span class="sd">######################################################</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="c1"># =============================================================================</span>
<span class="c1">#                               ACCESS</span>
<span class="c1"># =============================================================================</span>


<div class="viewcode-block" id="columnames"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.columnames">[docs]</a><span class="k">def</span> <span class="nf">columnames</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return on a list the name of the columns of the DataFrame</span>

<span class="sd">    :param data: a dataframe to obtain the name of its columns</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    </span>
<span class="sd">    :return: a list with the column names of the data</span>
<span class="sd">    :rtype: list</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DataColumnames</span><span class="p">()</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="rownames"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.rownames">[docs]</a><span class="k">def</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return on a list the name of the rows of the DataFrame</span>

<span class="sd">    :param data: a dataframe to obtain the name of its rows</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    </span>
<span class="sd">    :return: a list with the row names of the data</span>
<span class="sd">    :rtype: list</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DataRownames</span><span class="p">()</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="explain_variable"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.explain_variable">[docs]</a><span class="k">def</span> <span class="nf">explain_variable</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="n">var_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return information about a specific variable of a DataFrame</span>

<span class="sd">    :param samples: a dataframe with information of variables</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param var_id: specific identificator of a variable</span>
<span class="sd">    :type var_id: str, int</span>
<span class="sd">    </span>
<span class="sd">    :return: specific information of a variable</span>
<span class="sd">    :rtype: series</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DataExplainVariable</span><span class="p">(</span><span class="n">var_id</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">samples</span><span class="p">)</span></div>


<div class="viewcode-block" id="explain_variable_colname"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.explain_variable_colname">[docs]</a><span class="k">def</span> <span class="nf">explain_variable_colname</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="n">var_id</span><span class="p">,</span> <span class="n">colnm</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return specific field about a specific variable of a DataFrame</span>

<span class="sd">    :param samples: a dataframe with information of variables</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param var_id: specific identificator of a variable</span>
<span class="sd">    :type var_id: str, int</span>
<span class="sd">    :param colnm: specific column name of the DataFrame (field)</span>
<span class="sd">    :type colnm: str, int</span>
<span class="sd">    </span>
<span class="sd">    :return: specific field of information of a variable</span>
<span class="sd">    :rtype: series</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DataExplainVariableColname</span><span class="p">(</span><span class="n">var_id</span><span class="p">,</span> <span class="n">colnm</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">samples</span><span class="p">)</span></div>


<div class="viewcode-block" id="explain_all_variable_colname"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.explain_all_variable_colname">[docs]</a><span class="k">def</span> <span class="nf">explain_all_variable_colname</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="n">colnm</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return specific field about all variables of a DataFrame</span>

<span class="sd">    :param samples: a dataframe with information of variables</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param colnm: specific column name of the DataFrame (field)</span>
<span class="sd">    :type colnm: str, int</span>
<span class="sd">    </span>
<span class="sd">    :return: a dataframe with specific field of information about all variables</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DataExplainAllVariableColname</span><span class="p">(</span><span class="n">colnm</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">samples</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                               CHECKING</span>
<span class="c1"># =============================================================================</span>


<div class="viewcode-block" id="check_element"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.check_element">[docs]</a><span class="k">def</span> <span class="nf">check_element</span><span class="p">(</span><span class="n">list_</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">element</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Check if a given element is on a list or array</span>
<span class="sd">        </span>
<span class="sd">    :param list_: The current list to be process</span>
<span class="sd">    :type list_: list</span>
<span class="sd">    :param element: element to find on a list</span>
<span class="sd">    :type element: int, float, str, bool</span>

<span class="sd">    :return: True if element is on a list, False otherwise</span>
<span class="sd">    :rtype: bool</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">CheckElement</span><span class="p">(</span><span class="n">element</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">list_</span><span class="p">)</span></div>


<div class="viewcode-block" id="check_all_element"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.check_all_element">[docs]</a><span class="k">def</span> <span class="nf">check_all_element</span><span class="p">(</span><span class="n">list_</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">sublist</span> <span class="o">=</span> <span class="p">[]):</span>
    <span class="sd">&quot;&quot;&quot;Check if a given list of elements are on a list or array</span>
<span class="sd">        </span>
<span class="sd">    :param list_: The current list to be process</span>
<span class="sd">    :type list_: list</span>
<span class="sd">    :param sublist: list of element to find</span>
<span class="sd">    :type sublist: list, ndarray</span>

<span class="sd">    :return: True if all elements is on a list, False otherwise</span>
<span class="sd">    :rtype: bool</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">CheckAllElement</span><span class="p">(</span><span class="n">sublist</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">list_</span><span class="p">)</span></div>


<div class="viewcode-block" id="check_sub_element"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.check_sub_element">[docs]</a><span class="k">def</span> <span class="nf">check_sub_element</span><span class="p">(</span><span class="n">list_</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">partial_sublist</span> <span class="o">=</span> <span class="p">[]):</span>
    <span class="sd">&quot;&quot;&quot;It receives a list with elements where it is only needed to find </span>
<span class="sd">        a SUBSTRING of the name to be valid.</span>
<span class="sd">    </span>
<span class="sd">    :param list_: The current list to be process</span>
<span class="sd">    :type list_: list</span>
<span class="sd">    :param partial_sublist: list of element to find without need to match all name</span>
<span class="sd">    :type partial_sublist: list, ndarray</span>

<span class="sd">    :return: a list of coincidence with the large name</span>
<span class="sd">    :rtype: list, ndarray</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">CheckSubElement</span><span class="p">(</span><span class="n">partial_sublist</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">list_</span><span class="p">)</span></div>


<div class="viewcode-block" id="check_sub_element_short"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.check_sub_element_short">[docs]</a><span class="k">def</span> <span class="nf">check_sub_element_short</span><span class="p">(</span><span class="n">list_</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">partial_sublist</span> <span class="o">=</span> <span class="p">[]):</span>
    <span class="sd">&quot;&quot;&quot;It receives a list with elements where it is only needed to find </span>
<span class="sd">        a SUBSTRING of the name to be valid, but returns the short name founded</span>
<span class="sd">    </span>
<span class="sd">    :param list_: The current list to be process</span>
<span class="sd">    :type list_: list</span>
<span class="sd">    :param partial_sublist: list of element to find without need to match all name</span>
<span class="sd">    :type partial_sublist: list, ndarray</span>

<span class="sd">    :return: a list of coincidence with the short name</span>
<span class="sd">    :rtype: list, ndarray</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">CheckSubElementShort</span><span class="p">(</span><span class="n">partial_sublist</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">list_</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                               EXTRACTION</span>
<span class="c1"># =============================================================================</span>


<span class="c1"># Projection means choosing which columns (or expressions) the query shall return.</span>

<div class="viewcode-block" id="data_projection_list"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.data_projection_list">[docs]</a><span class="k">def</span> <span class="nf">data_projection_list</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">column_list</span><span class="o">=</span><span class="p">[]):</span>
    <span class="sd">&quot;&quot;&quot;Extract a Sub-DataFrame selecting a specific column list.</span>
<span class="sd">        </span>
<span class="sd">    :param data: The current DataFrane to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param column_list: A list with the names of the columns to keep</span>
<span class="sd">    :type column_list: list</span>
<span class="sd">    :return: a Sub-DataFrame with selected columns</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DataProjectionList</span><span class="p">(</span><span class="n">column_list</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="data_projection_index"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.data_projection_index">[docs]</a><span class="k">def</span> <span class="nf">data_projection_index</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">indx</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract a Sub-DataFrame selecting a specific column by index.</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrane to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param indx: a number of a column. Must be on range of lenght columns</span>
<span class="sd">    :type indx: int</span>

<span class="sd">    :return: a Sub-DataFrame with selected column</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DataProjectionIndex</span><span class="p">(</span><span class="n">indx</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="data_projection_name"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.data_projection_name">[docs]</a><span class="k">def</span> <span class="nf">data_projection_name</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract a Sub-DataFrame selecting a specific column by name.</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrane to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param name: a name of a column. Must exists</span>
<span class="sd">    :type name: str</span>

<span class="sd">    :return: a Sub-DataFrame with selected column</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DataProjectionName</span><span class="p">(</span><span class="n">name</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="data_projection_range"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.data_projection_range">[docs]</a><span class="k">def</span> <span class="nf">data_projection_range</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ini_column</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">fini_column</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract a Sub-DataFrame selecting a range of columns</span>
<span class="sd">        </span>
<span class="sd">    :param data: The current DataFrane to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param ini_column: a initial position to project. Must be on range of lenght columns</span>
<span class="sd">    :type ini_column: int</span>
<span class="sd">    :param fini_column: a final position to project. Must be on range of lenght columns</span>
<span class="sd">    :type fini_column: int</span>

<span class="sd">    :return: a Sub-DataFrame with the selected columns in an interval</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DataProjectionRange</span><span class="p">(</span><span class="n">ini_column</span><span class="p">,</span> <span class="n">fini_column</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="data_projection_filter"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.data_projection_filter">[docs]</a><span class="k">def</span> <span class="nf">data_projection_filter</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">filter_</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract a Sub-DataFrame selecting columns by filter.</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrane to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param filter_: a filter to do the projection by columns</span>
<span class="sd">    :type filter_: pd.Series, pd.DataFrame</span>

<span class="sd">    :return: a Sub-DataFrame after apply a filter by columns</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DataProjectionFilter</span><span class="p">(</span><span class="n">filter_</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="data_projection_substring"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.data_projection_substring">[docs]</a><span class="k">def</span> <span class="nf">data_projection_substring</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">column_list</span><span class="o">=</span><span class="p">[],</span> <span class="n">rename</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract a Sub-DataFrame selecting a specific column list no matter if name doesn&#39;t match complete.</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrane to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param column_list: A list with the names of the columns to keep</span>
<span class="sd">    :type column_list: list</span>
<span class="sd">    :param rename: check if should rename the column to the short name. True, then rename, False, not</span>
<span class="sd">    :type rename: bool</span>

<span class="sd">    :return: a Sub-DataFrame with selected columns</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DataProjectionSubstring</span><span class="p">(</span><span class="n">column_list</span><span class="p">,</span> <span class="n">rename</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<span class="c1"># Selection means which rows are to be returned.</span>

<div class="viewcode-block" id="data_selection_list"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.data_selection_list">[docs]</a><span class="k">def</span> <span class="nf">data_selection_list</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">row_list</span><span class="o">=</span><span class="p">[]):</span>
    <span class="sd">&quot;&quot;&quot;Extract a Sub-DataFrame selecting a specific row list.</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrane to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param row_list: A list with the names of the rows to keep</span>
<span class="sd">    :type row_list: list</span>

<span class="sd">    :return: a Sub-DataFrame with selected rows</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DataSelectionList</span><span class="p">(</span><span class="n">row_list</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="data_selection_index"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.data_selection_index">[docs]</a><span class="k">def</span> <span class="nf">data_selection_index</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">indx</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract a Sub-DataFrame selecting a specific row by index.</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrane to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param indx: a number of a row. Must be on range of lenght row</span>
<span class="sd">    :type indx: int</span>

<span class="sd">    :return: a Sub-DataFrame with selected row</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DataSelectionIndex</span><span class="p">(</span><span class="n">indx</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="data_selection_name"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.data_selection_name">[docs]</a><span class="k">def</span> <span class="nf">data_selection_name</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract a Sub-DataFrame selecting a specific row by name.</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrane to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param name: a name of a row. Must exists</span>
<span class="sd">    :type name: str</span>

<span class="sd">    :return: a Sub-DataFrame with selected row</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DataSelectionName</span><span class="p">(</span><span class="n">name</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="data_selection_range"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.data_selection_range">[docs]</a><span class="k">def</span> <span class="nf">data_selection_range</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ini_row</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">fini_row</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract a Sub-DataFrame selecting a range of rows</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrane to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param ini_row: a initial position to select. Must be on range of lenght rows</span>
<span class="sd">    :type ini_row: int</span>
<span class="sd">    :param fini_row: a final position to select. Must be on range of lenght rows</span>
<span class="sd">    :type fini_row: int</span>

<span class="sd">    :return: a Sub-DataFrame with the selected rows in an interval</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DataSelectionRange</span><span class="p">(</span><span class="n">ini_row</span><span class="p">,</span> <span class="n">fini_row</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="data_selection_filter"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.data_selection_filter">[docs]</a><span class="k">def</span> <span class="nf">data_selection_filter</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">filter_</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract a Sub-DataFrame selecting rows by filter.</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrane to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param filter_: a filter to do the selection by rows</span>
<span class="sd">    :type filter_: pd.Series, pd.DataFrame</span>

<span class="sd">    :return: a Sub-DataFrame after apply a filter by rows</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DataSelectionFilter</span><span class="p">(</span><span class="n">filter_</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="data_selection_series"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.data_selection_series">[docs]</a><span class="k">def</span> <span class="nf">data_selection_series</span><span class="p">(</span><span class="n">series</span> <span class="p">:</span> <span class="n">ut</span><span class="o">.</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract a Sub-DataFrame selecting a specific row by series.</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrane to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param series: A series with a structure to keep</span>
<span class="sd">    :type series: list</span>

<span class="sd">    :return: a Sub-DataFrame with selected rows</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DataSelectionSeries</span><span class="p">(</span><span class="n">series</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                               FILTERING</span>
<span class="c1"># =============================================================================</span>


<div class="viewcode-block" id="filter_by_index"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.filter_by_index">[docs]</a><span class="k">def</span> <span class="nf">filter_by_index</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">index_cond</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract a Sub-DataFrame selected by a filter of index condition.</span>
<span class="sd">        For multi-index cases:</span>
<span class="sd">        index_cond must have index type (it is returned after  do dataframe.index)</span>
<span class="sd">        It keeps only the rows that are in index_cond</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrane to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param index_cond: a condition as filter</span>
<span class="sd">    :type index_cond: str</span>

<span class="sd">    :return: a Sub-DataFrame that keeps only the rows that are in index_cond</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">FilterByIndex</span><span class="p">(</span><span class="n">index_cond</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="filter_dictionary"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.filter_dictionary">[docs]</a><span class="k">def</span> <span class="nf">filter_dictionary</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">list_var</span> <span class="o">=</span> <span class="p">[]):</span>
    <span class="sd">&quot;&quot;&quot;Given a Dictionary, select only elements on a passed list</span>
<span class="sd">        </span>
<span class="sd">    :param dict_: The current Dictionary to be process</span>
<span class="sd">    :type dict_: dict</span>
<span class="sd">    :param list_var: new Count Matrix</span>
<span class="sd">    :type list_var: list</span>

<span class="sd">    :return: A new dictionary with filtered elements</span>
<span class="sd">    :rtype: dict</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">FilterDictionary</span><span class="p">(</span><span class="n">list_var</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                             INTERSECTION</span>
<span class="c1"># =============================================================================</span>


<div class="viewcode-block" id="list_intersection"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.list_intersection">[docs]</a><span class="k">def</span> <span class="nf">list_intersection</span><span class="p">(</span><span class="n">list_</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">sublist</span> <span class="o">=</span> <span class="p">[]):</span>
    <span class="sd">&quot;&quot;&quot;Obtains a new list with commom elements of two list or array</span>
<span class="sd">    </span>
<span class="sd">    :param list_: The primary list to be process</span>
<span class="sd">    :type list_: list</span>
<span class="sd">    :param sublist: secondary list to find commom elements</span>
<span class="sd">    :type sublist: list</span>

<span class="sd">    :return: A list which is the intersection between both input lists</span>
<span class="sd">    :rtype: list</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">ListIntersection</span><span class="p">(</span><span class="n">sublist</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">list_</span><span class="p">)</span></div>


<div class="viewcode-block" id="list_intersection_substring"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.list_intersection_substring">[docs]</a><span class="k">def</span> <span class="nf">list_intersection_substring</span><span class="p">(</span><span class="n">list_</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">substrings</span> <span class="o">=</span> <span class="p">[]):</span>
    <span class="sd">&quot;&quot;&quot;Obtains a new list with commom elements of two lists or array</span>
<span class="sd">        It is enough if only match partial name of element</span>
<span class="sd">    </span>
<span class="sd">    :param list_: The primary list to be process</span>
<span class="sd">    :type list_: list</span>
<span class="sd">    :param substrings: secondary list to find commom elements</span>
<span class="sd">    :type substrings: list</span>
<span class="sd">    :return: A list with founded common substring between two list</span>
<span class="sd">    :rtype: list</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">ListIntersectionSubString</span><span class="p">(</span><span class="n">substrings</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">list_</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                                ZIP</span>
<span class="c1"># =============================================================================</span>



<div class="viewcode-block" id="list_zip"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.list_zip">[docs]</a><span class="k">def</span> <span class="nf">list_zip</span><span class="p">(</span><span class="n">list_</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">sublist</span> <span class="o">=</span> <span class="p">[]):</span>
    <span class="sd">&quot;&quot;&quot;Creates a new dictionary using one list</span>
<span class="sd">        for the keys, and other to its values</span>
<span class="sd">    </span>
<span class="sd">    :param list_: The current list to be process</span>
<span class="sd">    :type list_: list</span>
<span class="sd">    :param sublist: a list with values of the dictionary</span>
<span class="sd">    :type sublist: list</span>

<span class="sd">    :return: with the current list used as keys, creates and returns a dictionary with second list as values</span>
<span class="sd">    :rtype: dict</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">ListZip</span><span class="p">(</span><span class="n">sublist</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">list_</span><span class="p">)</span></div>


<div class="viewcode-block" id="data_zip"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.data_zip">[docs]</a><span class="k">def</span> <span class="nf">data_zip</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">key_column</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">values_column</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Creates a dictionary using information about two columns</span>
<span class="sd">        of a DataFrame</span>
<span class="sd">        </span>
<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param key_column: column name of a DataFrame used for the dictionary keys</span>
<span class="sd">    :type key_column: str</span>
<span class="sd">    :param values_column: column name of a DataFrame used for the dictionary values</span>
<span class="sd">    :type values_column: str</span>

<span class="sd">    :return: with both columns indicates, create and returns a dictionary</span>
<span class="sd">    :rtype: dict</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DataZip</span><span class="p">(</span><span class="n">key_column</span><span class="p">,</span> <span class="n">values_column</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                                RENAME</span>
<span class="c1"># =============================================================================</span>



<div class="viewcode-block" id="rename_columname"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.rename_columname">[docs]</a><span class="k">def</span> <span class="nf">rename_columname</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pos_ini</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">pos_fin</span> <span class="o">=</span> <span class="mi">0</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Rename all names of the columns triping the name</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param pos_ini: position of the first character</span>
<span class="sd">    :type pos_ini: int</span>
<span class="sd">    :param pos_fin: position of the last character</span>
<span class="sd">    :type pos_fin: int</span>

<span class="sd">    :return: A DataFrame with the column renamed with a substring of the inner column name</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">RenameColumname</span><span class="p">(</span><span class="n">pos_ini</span><span class="p">,</span> <span class="n">pos_fin</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="rename_index"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.rename_index">[docs]</a><span class="k">def</span> <span class="nf">rename_index</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">new_name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Rename the name of the current index of DataFrame</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param new_name: name that will replace current one</span>
<span class="sd">    :type new_name: str</span>

<span class="sd">    :return: The same DataFrame with the index changed</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">RenameIndex</span><span class="p">(</span><span class="n">new_name</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                                DUPLICATES</span>
<span class="c1"># =============================================================================</span>


<div class="viewcode-block" id="duplicates"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.duplicates">[docs]</a><span class="k">def</span> <span class="nf">duplicates</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">keep</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns a DataFrame with the duplicates rows or columns of a input DataFrame</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param axis: axis 0 means by rows, 1 by columns</span>
<span class="sd">    :type axis: int</span>
<span class="sd">    :param keep: Indicates which duplicates will keep</span>
<span class="sd">        \&quot;first\&quot; keeps only the first duplicated column/row; \&quot;last\&quot; keeps only the last duplicated column/row; False keeps both duplicated columns/rows</span>
<span class="sd">    :type keep: str, bool</span>

<span class="sd">    :return: A dataframe that only contains duplicated columns or rows</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DataDuplicates</span><span class="p">(</span><span class="n">axis</span><span class="p">,</span> <span class="n">keep</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                              INFORMATION</span>
<span class="c1"># =============================================================================</span>


<div class="viewcode-block" id="describe"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.describe">[docs]</a><span class="k">def</span> <span class="nf">describe</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">perc</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Show a statistic description about a DataFrame</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param perc: The percentiles to include in the output</span>
<span class="sd">    :type perc: list, float</span>

<span class="sd">    :return: descriptive statistics that summarize the central tendency, dispersion and shape </span>
<span class="sd">        of the dataset‚Äôs distribution, excluding NaN values</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">Describe</span><span class="p">(</span><span class="n">perc</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="count_types"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.count_types">[docs]</a><span class="k">def</span> <span class="nf">count_types</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">col_name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Shows a Series containing counts of unique values of a specific column.</span>
<span class="sd">        It will be in descending order so that the first element is the most frequently-occurring element</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param col_name: name of the column to count types</span>
<span class="sd">    :type col_name: str</span>
<span class="sd">    :return: information about types of a column</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">CountTypes</span><span class="p">(</span><span class="n">col_name</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                               REPLACE</span>
<span class="c1"># =============================================================================</span>


<div class="viewcode-block" id="replace"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.replace">[docs]</a><span class="k">def</span> <span class="nf">replace</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">to_replace</span> <span class="o">=</span> <span class="s2">&quot;Unknown&quot;</span><span class="p">,</span> <span class="n">replaced_by</span> <span class="o">=</span> <span class="n">ut</span><span class="o">.</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Replace DataFrame&#39;s selected values with another.</span>

<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param to_replace: values to be replaced</span>
<span class="sd">    :type to_replace: str, float, int, bool, list, ndarray</span>
<span class="sd">    :param replaced_by: new value to replaced the value before</span>
<span class="sd">    :type replaced_by: str, float, int, bool</span>

<span class="sd">    :return: a DataFrame with all values indicates replaced by other value</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">Replace</span><span class="p">(</span><span class="n">to_replace</span><span class="p">,</span> <span class="n">replaced_by</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="fill_nan"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.fill_nan">[docs]</a><span class="k">def</span> <span class="nf">fill_nan</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">to_replace</span> <span class="o">=</span> <span class="s2">&quot;Unknown&quot;</span><span class="p">,</span> <span class="n">replaced_by</span> <span class="o">=</span> <span class="n">ut</span><span class="o">.</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Replace nan values to mean of the row of DataFrame</span>
<span class="sd">        Only modifies values of DataFrame.</span>

<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>

<span class="sd">    :return: A DataFrame with all nan values replace to mean</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">FillNan</span><span class="p">(</span><span class="n">to_replace</span><span class="p">,</span> <span class="n">replaced_by</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                              TRANSPOSE</span>
<span class="c1"># =============================================================================</span>


<div class="viewcode-block" id="transpose"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.transpose">[docs]</a><span class="k">def</span> <span class="nf">transpose</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">index_name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Do transpose to a DataFrame keeping the name of columns and index</span>
<span class="sd">        without multi-index</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param index_name: the name of the index before transpose to keep it</span>
<span class="sd">    :type index_name: str, int</span>

<span class="sd">    :return: a DataFrame with transpose of the input taking into account the extra row that it is created after the transpose</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">Transpose</span><span class="p">(</span><span class="n">index_name</span> <span class="o">=</span> <span class="n">index_name</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                                ADD</span>
<span class="c1"># =============================================================================</span>


<div class="viewcode-block" id="add_column"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.add_column">[docs]</a><span class="k">def</span> <span class="nf">add_column</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">name_column</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">values</span> <span class="o">=</span> <span class="p">[]):</span>
    <span class="sd">&quot;&quot;&quot;Add new column to a DataFrame</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param name_column: Name of the new column</span>
<span class="sd">    :type name_column: str</span>
<span class="sd">    :param values: values of the new column. Must have the same lenght as number of rows of DataFrame</span>
<span class="sd">    :type values: list, ndarray</span>

<span class="sd">    :return: the input DataFrame with a new column with its values</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">AddColumn</span><span class="p">(</span><span class="n">name_column</span><span class="p">,</span> <span class="n">values</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="add_label"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.add_label">[docs]</a><span class="k">def</span> <span class="nf">add_label</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">list_0s</span><span class="p">,</span> <span class="n">column_observed</span><span class="p">,</span> <span class="n">name_label</span> <span class="o">=</span> <span class="s2">&quot;label&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Add a special column to a DataFrame.</span>
<span class="sd">        It will be used on Machine Learning as the Target Column of Classification</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param list_0s: List of values that consider as 0 on Label value</span>
<span class="sd">    :type list_0s: list, ndarray</span>
<span class="sd">    :param column_observed: Column of the DataFrame used to decide when value of target is 0 or 1</span>
<span class="sd">    :type column_observed: str</span>
<span class="sd">    :param name_label: values of the new column. Must have the same lenght as number of rows of DataFrame</span>
<span class="sd">    :type name_label: list, ndarray</span>

<span class="sd">    :return: a DataFrame with a new column that represents the labels of data to classify</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">AddLabel</span><span class="p">(</span><span class="n">list_0s</span><span class="p">,</span> <span class="n">column_observed</span><span class="p">,</span> <span class="n">name_label</span><span class="o">=</span><span class="n">name_label</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                                DELETE</span>
<span class="c1"># =============================================================================</span>


<div class="viewcode-block" id="drop_duplicated"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.drop_duplicated">[docs]</a><span class="k">def</span> <span class="nf">drop_duplicated</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">keep</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">by_name</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Drop data duplicates of a DataFrame</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param axis: axis 0 means by rows, 1 by columns</span>
<span class="sd">    :type axis: int</span>
<span class="sd">    :param keep: Indicates which duplicates will keep</span>
<span class="sd">        \&quot;first\&quot; keeps only the first duplicated column/row; \&quot;last\&quot; keeps only the last duplicated column/row; False keeps both duplicated columns/rows</span>
<span class="sd">    :type keep: str, bool</span>
<span class="sd">    :param by_name: True if only evaluate the name of the column/row, False if evaluate the content of the column/row to decide if is duplicated or not</span>
<span class="sd">    :type by_name: bool</span>

<span class="sd">    :return: A new dataframe without duplicates rows or columns</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DropDuplicates</span><span class="p">(</span><span class="n">axis</span><span class="p">,</span> <span class="n">keep</span><span class="p">,</span> <span class="n">by_name</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="drop_values"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.drop_values">[docs]</a><span class="k">def</span> <span class="nf">drop_values</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">to_delete</span> <span class="o">=</span> <span class="n">ut</span><span class="o">.</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Drop values of a DataFrame</span>

<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param to_delete: value to be delete.</span>
<span class="sd">    :type to_delete: str, float, int, bool, list, ndarray, </span>
<span class="sd">    :param axis: Indicates 0 to row and 1 to columns</span>
<span class="sd">    :type axis: int</span>
<span class="sd">    :param method: all to check all values are the same, any if a partial value match</span>
<span class="sd">    :type method: str</span>

<span class="sd">    :return: A new dataframe within this values</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DropValues</span><span class="p">(</span><span class="n">to_delete</span><span class="p">,</span> <span class="n">axis</span><span class="p">,</span> <span class="n">method</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="drop_nan_by_thresh"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.drop_nan_by_thresh">[docs]</a><span class="k">def</span> <span class="nf">drop_nan_by_thresh</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">thresh</span> <span class="o">=</span> <span class="mi">0</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Drop DataFrame&#39;s values limited by Threshold.</span>

<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param axis: Indicates 0 to row and 1 to columns</span>
<span class="sd">    :type axis: int</span>
<span class="sd">    :param thresh: threshold to evaluate element. If value less than threshold, that will be selected to drop</span>
<span class="sd">    :type thresh: float, int</span>

<span class="sd">    :return: A dataframe with drop element checking a threshold</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DropNanByThresh</span><span class="p">(</span><span class="n">axis</span><span class="p">,</span> <span class="n">thresh</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="drop_rows_by_colname"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.drop_rows_by_colname">[docs]</a><span class="k">def</span> <span class="nf">drop_rows_by_colname</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">col_name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">to_delete</span> <span class="o">=</span> <span class="n">ut</span><span class="o">.</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Drop DataFrame&#39;s rows searching a value on a specific column</span>

<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param col_name: name of the column to search</span>
<span class="sd">    :type col_name: str</span>
<span class="sd">    :param to_delete: value to find on a column to decide with row delete is exists</span>
<span class="sd">    :type to_delete: str, float, int, bool, list, ndarray</span>

<span class="sd">    :return: A dataframe with dropped rows search a value on a specific column</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">DropRowsByColname</span><span class="p">(</span><span class="n">col_name</span><span class="p">,</span> <span class="n">to_delete</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>



<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">######################################################</span>

<span class="sd">                QUANTITATIVE ANALYSIS</span>

<span class="sd">######################################################</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="c1"># =============================================================================</span>
<span class="c1">#                              CORRELATION</span>
<span class="c1"># =============================================================================</span>


<div class="viewcode-block" id="var_correlation"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.var_correlation">[docs]</a><span class="k">def</span> <span class="nf">var_correlation</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Calculates correlation between variables of a DataFrame</span>

<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>

<span class="sd">    :return: A dataframe with the correlation between the variables</span>
<span class="sd">        with values from 0.0 to 1.0, being 1.0 the highest value of correlation</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">VarCorrelation</span><span class="p">()</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="remove_correlation"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.remove_correlation">[docs]</a><span class="k">def</span> <span class="nf">remove_correlation</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">thresh</span> <span class="o">=</span> <span class="mf">0.0</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Remove the elements that has high correlation</span>

<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>

<span class="sd">    :return: A dataframe with high correlation removed</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">RemoveCorrelation</span><span class="p">(</span><span class="n">thresh</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                               VARIANCE</span>
<span class="c1"># =============================================================================</span>


<div class="viewcode-block" id="variance"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.variance">[docs]</a><span class="k">def</span> <span class="nf">variance</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Calculates the variance of a matrix</span>

<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>

<span class="sd">    :return: A dictionary with the variance of each variable</span>
<span class="sd">    :rtype: dict</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">Variance</span><span class="p">()</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                              SELECTION</span>
<span class="c1"># =============================================================================</span>


<div class="viewcode-block" id="top_variables"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.top_variables">[docs]</a><span class="k">def</span> <span class="nf">top_variables</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">n_var</span> <span class="o">=</span> <span class="mi">0</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Calculates the best n variables of a DataFrame</span>

<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param n_var: Number of variables selected as the best</span>
<span class="sd">    :type n_var: int</span>

<span class="sd">    :return: A dataframe with n_var rows that are the best</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">TopVariables</span><span class="p">(</span><span class="n">n_var</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="remove_low_reads"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.remove_low_reads">[docs]</a><span class="k">def</span> <span class="nf">remove_low_reads</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">thresh</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">al_least</span> <span class="o">=</span> <span class="mi">2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Calculates the best n variables of a DataFrame</span>

<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param thresh: threshold that indicates the limit to remove feature less than it</span>
<span class="sd">    :type thresh: float</span>
<span class="sd">    :param at_least: Indicates the minimum match to pass the selection</span>
<span class="sd">    :type at_least: int</span>

<span class="sd">    :return: A dataframe with removed features which have less then ‚Äòthreshold‚Äô reads all columns</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">RemoveLowReads</span><span class="p">(</span><span class="n">thresh</span><span class="p">,</span> <span class="n">al_least</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="thresh_by_perc"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.thresh_by_perc">[docs]</a><span class="k">def</span> <span class="nf">thresh_by_perc</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">perc</span> <span class="o">=</span> <span class="mf">10.0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Calculate a limit by percentage by number of rows or columns</span>

<span class="sd">    :param data: the dataframe to calculate that limit according to number of rows or columns.</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param perc: portion of total elements (0.00 to 100.0).</span>
<span class="sd">    :type perc: float</span>
<span class="sd">    :param axis: Axis 0 means by rows, 1 by columns.</span>
<span class="sd">    :type axis: int</span>

<span class="sd">    :return: the mininum number of elements that consider a limit.</span>
<span class="sd">    :rtype: int</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ut</span><span class="o">.</span><span class="n">thresh_by_perc</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">perc</span><span class="p">,</span> <span class="n">axis</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                              NORMALIZATION</span>
<span class="c1"># =============================================================================</span>


<div class="viewcode-block" id="cpm"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.cpm">[docs]</a><span class="k">def</span> <span class="nf">cpm</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">log_method</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">prior_count</span> <span class="o">=</span> <span class="mi">2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Normalize values of DataFrame using CPM method</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param log_method: True if CPM will do with logaritmic alghoritm, False otherwise</span>
<span class="sd">    :type log_method: bool</span>
<span class="sd">    :param prior_count: If log is True, ends up getting scaled by the ratio of a library size to the average library size and then multiplied by its value before getting added to each library size</span>
<span class="sd">    :type prior_count: int</span>

<span class="sd">    :return: a DataFrame to computed one of the CPM method</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">CPM</span><span class="p">(</span><span class="n">log_method</span> <span class="o">=</span> <span class="n">log_method</span><span class="p">,</span> <span class="n">prior_count</span> <span class="o">=</span> <span class="n">prior_count</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>



<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">######################################################</span>

<span class="sd">                    RNA PROCESS</span>

<span class="sd">######################################################</span>
<span class="sd">&quot;&quot;&quot;</span>


<div class="viewcode-block" id="rna_sample_condition"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.rna_sample_condition">[docs]</a><span class="k">def</span> <span class="nf">rna_sample_condition</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Classify a given sample analyze its barcode</span>

<span class="sd">    :param data: The current str to be process</span>
<span class="sd">    :type brc_string: str</span>
<span class="sd">    :param sep: indicates with character is used to separate values</span>
<span class="sd">    :type sep: str</span>

<span class="sd">    :return: classification of the sample by its barcode passed on input</span>
<span class="sd">    :rtype: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">SampleCondition</span><span class="p">(</span><span class="n">sep</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="rna_stat_test"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.rna_stat_test">[docs]</a><span class="k">def</span> <span class="nf">rna_stat_test</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">grouped_by</span><span class="p">,</span> <span class="n">group_1</span><span class="p">,</span> <span class="n">group_2</span><span class="p">,</span> <span class="n">clinical_data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Generate a DataFrame with several statistic information like FoldChange, T, P-Value and a DiffExpr</span>

<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param grouped_by: name of the column of clinical data to organize data</span>
<span class="sd">    :type grouped_by: str</span>
<span class="sd">    :param group_1: one of the type to create a group</span>
<span class="sd">    :type group_1: str</span>
<span class="sd">    :param group_2: another type to create a group</span>
<span class="sd">    :type group_2: str</span>
<span class="sd">    :param clinical_data: a dataframe with clinical information of the data</span>
<span class="sd">    :type clinical_data: DataFrame</span>

<span class="sd">    :return: A dataframe with several stadistic information: FoldChange, T, P-Value and a DiffExpr</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">StatTest</span><span class="p">(</span><span class="n">grouped_by</span><span class="p">,</span> <span class="n">group_1</span><span class="p">,</span> <span class="n">group_2</span><span class="p">,</span> <span class="n">clinical_data</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="rna_genes_lenght"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.rna_genes_lenght">[docs]</a><span class="k">def</span> <span class="nf">rna_genes_lenght</span><span class="p">(</span><span class="n">genemodel</span><span class="p">,</span> <span class="n">colnm_lenght</span> <span class="o">=</span> <span class="s2">&quot;width&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Given a DataFrame with the genemodel information, return all lenght of the actual studied genes</span>
<span class="sd">    </span>
<span class="sd">    :param genemodel: The current DataFrame to be process</span>
<span class="sd">    :type genemodel: DataFrame</span>
<span class="sd">    :param colnm_lenght: name of the column that contains information about their lenght</span>
<span class="sd">    :type colnm_lenght: str</span>

<span class="sd">    :return: a dataframe with lenghts of the genes</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">GenesLenght</span><span class="p">(</span><span class="n">colnm_lenght</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">genemodel</span><span class="p">)</span></div>


<div class="viewcode-block" id="rna_gen_length"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.rna_gen_length">[docs]</a><span class="k">def</span> <span class="nf">rna_gen_length</span><span class="p">(</span><span class="n">genemodel</span><span class="p">,</span> <span class="n">gen</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns lenght of a specific gen indicates</span>
<span class="sd">    </span>
<span class="sd">    :param genemodel: a DataFrame with genemodel information</span>
<span class="sd">    :type genemodel: DataFrame</span>
<span class="sd">    :param gen_id: identifier of the gene on a genemodel</span>
<span class="sd">    :type gen_id: str</span>

<span class="sd">    :return: the lenght of a specific gen selected</span>
<span class="sd">    :rtype: int</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">GeneIDLenght</span><span class="p">(</span><span class="n">gen_id</span><span class="o">=</span><span class="n">gen</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">genemodel</span><span class="p">)</span></div>


<div class="viewcode-block" id="rna_rpk"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.rna_rpk">[docs]</a><span class="k">def</span> <span class="nf">rna_rpk</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">genemodel</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Normalize values of DataFrame using RPK method</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param genemodel: a DataFrame with gene information, like gene lenght</span>
<span class="sd">    :type genemodel: DataFrame</span>

<span class="sd">    :return: a DataFrame with computed RPK</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">RPK</span><span class="p">(</span><span class="n">genemodel</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="rna_tpm"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.rna_tpm">[docs]</a><span class="k">def</span> <span class="nf">rna_tpm</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">genemodel</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Normalize values of DataFrame using TPM method</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param genemodel: a DataFrame with gene information, like gene lenght</span>
<span class="sd">    :type genemodel: DataFrame</span>

<span class="sd">    :return: a DataFrame with computed TPM</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">TPM</span><span class="p">(</span><span class="n">genemodel</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="rna_rpkm"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.rna_rpkm">[docs]</a><span class="k">def</span> <span class="nf">rna_rpkm</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">genemodel</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Normalize values of DataFrame using RPKM method</span>
<span class="sd">    </span>
<span class="sd">    :param data: The current DataFrame to be process</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param genemodel: a DataFrame with gene information, like gene lenght</span>
<span class="sd">    :type genemodel: DataFrame</span>

<span class="sd">    :return: a DataFrame with computed RPKM</span>
<span class="sd">    :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pro</span><span class="o">.</span><span class="n">RPKM</span><span class="p">(</span><span class="n">genemodel</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>



<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">######################################################</span>

<span class="sd">                    DATA OBJECT</span>

<span class="sd">######################################################</span>
<span class="sd">&quot;&quot;&quot;</span>


<div class="viewcode-block" id="create_data_object"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.create_data_object">[docs]</a><span class="k">def</span> <span class="nf">create_data_object</span><span class="p">(</span><span class="n">counts_dframe</span> <span class="p">:</span> <span class="n">ut</span><span class="o">.</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">obs_</span> <span class="o">=</span> <span class="n">ut</span><span class="o">.</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(),</span> <span class="n">var_</span> <span class="o">=</span> <span class="n">ut</span><span class="o">.</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(),</span> <span class="n">log</span> <span class="o">=</span> <span class="n">Workflow</span><span class="p">(),</span> <span class="n">unsdict</span> <span class="o">=</span> <span class="p">{}</span> <span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Creates a Data Object to save information about a project downloaded or read</span>

<span class="sd">    :param counts_dframe: That will be the count matrix of the DataObject</span>
<span class="sd">    :type counts_dframe: DataFrame </span>
<span class="sd">    :param obs_: That will be the clinical information associated to count matrix</span>
<span class="sd">        on DataObject</span>
<span class="sd">    :type obs_: DataFrame</span>
<span class="sd">    :param var_: That will be the gene information associated to count matrix</span>
<span class="sd">        on DataObject</span>
<span class="sd">    :type var_: DataFrame</span>
<span class="sd">    :param log: That will contains a flow that save step by step the changes of</span>
<span class="sd">        the DataObject. Useful to replicate an experiment with another information</span>
<span class="sd">    :type log: Workflow</span>
<span class="sd">    :param unsdict: It is a variable used to save extra data with an unclassified distribution</span>
<span class="sd">        on DataObject.</span>
<span class="sd">    :type unsdict: dict</span>

<span class="sd">    :return: A DataObject with all information organize</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">DataObject</span><span class="p">(</span><span class="n">counts_dframe</span> <span class="o">=</span> <span class="n">counts_dframe</span><span class="p">,</span> <span class="n">obs_</span> <span class="o">=</span> <span class="n">obs_</span><span class="p">,</span> <span class="n">var_</span> <span class="o">=</span> <span class="n">var_</span><span class="p">,</span> <span class="n">log</span> <span class="o">=</span> <span class="n">log</span><span class="p">,</span> <span class="n">unsdict</span> <span class="o">=</span> <span class="n">unsdict</span> <span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                        BASIC FUNCTIONS FOR COUNTS</span>
<span class="c1"># =============================================================================</span>


<div class="viewcode-block" id="set_counts"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.set_counts">[docs]</a><span class="k">def</span> <span class="nf">set_counts</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">new_counts</span> <span class="o">=</span> <span class="n">ut</span><span class="o">.</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()):</span>
    <span class="sd">&quot;&quot;&quot;Set DataObject&#39;s Count Matrix to new one.</span>
<span class="sd">        Control that the rest of the data match with new Count Matrix</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>
<span class="sd">    </span>
<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param new_counts: new Count Matrix</span>
<span class="sd">    :type new_counts: DataFrame</span>
<span class="sd">    </span>
<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">SetCounts</span><span class="p">(</span><span class="n">new_counts</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>


<div class="viewcode-block" id="update_counts"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.update_counts">[docs]</a><span class="k">def</span> <span class="nf">update_counts</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">new_counts</span> <span class="o">=</span> <span class="n">ut</span><span class="o">.</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()):</span>
    <span class="sd">&quot;&quot;&quot;Set DataObject&#39;s Count Matrix to new one but only modify its values.</span>
<span class="sd">        In this case, the function not control that the rest of the data match with new Count Matrix</span>
<span class="sd">        because any row or column will be modify. Will have the same shape</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>
<span class="sd">    </span>
<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param new_counts: new Count Matrix</span>
<span class="sd">    :type new_counts: DataFrame</span>

<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">UpdateCounts</span><span class="p">(</span><span class="n">new_counts</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                       ACCESSING COUNTS DATAFRAME </span>
<span class="c1"># =============================================================================</span>


<div class="viewcode-block" id="counts_selection"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.counts_selection">[docs]</a><span class="k">def</span> <span class="nf">counts_selection</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">row_list</span> <span class="o">=</span> <span class="p">[]):</span>
    <span class="sd">&quot;&quot;&quot;Set DataObject&#39;s Count Matrix selecting a specific row list.</span>
<span class="sd">        Control that the rest of the data match with new Count Matrix</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>
<span class="sd">    </span>
<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param row_list: A list with the names of the rows to keep</span>
<span class="sd">    :type row_list: list</span>

<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">CountsSelection</span><span class="p">(</span><span class="n">row_list</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>


<div class="viewcode-block" id="counts_projection"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.counts_projection">[docs]</a><span class="k">def</span> <span class="nf">counts_projection</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">col_list</span> <span class="o">=</span> <span class="p">[]):</span>
    <span class="sd">&quot;&quot;&quot;Set DataObject&#39;s Count Matrix selecting a specific column list.</span>
<span class="sd">        Control that the rest of the data match with new Count Matrix</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>
<span class="sd">    </span>
<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param col_list: A list with the names of the columns to keep</span>
<span class="sd">    :type col_list: list</span>

<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">CountsProjection</span><span class="p">(</span><span class="n">col_list</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>


<div class="viewcode-block" id="counts_projection_substring"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.counts_projection_substring">[docs]</a><span class="k">def</span> <span class="nf">counts_projection_substring</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">col_list</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">rename</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Set DataObject&#39;s Count Matrix selecting a specific column list.</span>
<span class="sd">        In that case, is not neccesary that name of the column were exactly</span>
<span class="sd">        Control that the rest of the data match with new Count Matrix</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>
<span class="sd">    </span>
<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param col_list: A list with the names of the columns to keep</span>
<span class="sd">    :type col_list: list</span>
<span class="sd">    :param rename: True if the column names of Count Matrix need to be renamed, False otherwise</span>
<span class="sd">    :type rename: bool</span>

<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">CountsProjectionSubstring</span><span class="p">(</span><span class="n">col_list</span><span class="p">,</span> <span class="n">rename</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>


<div class="viewcode-block" id="counts_replace_nan"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.counts_replace_nan">[docs]</a><span class="k">def</span> <span class="nf">counts_replace_nan</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Replace nan values to mean of the row of Count Matrix</span>
<span class="sd">        Only modifies values of Count Matrix.</span>
<span class="sd">    </span>
<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">CountsReplaceNan</span><span class="p">()</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>


<div class="viewcode-block" id="counts_drop_duplicates"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.counts_drop_duplicates">[docs]</a><span class="k">def</span> <span class="nf">counts_drop_duplicates</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">keep</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">by_name</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Drop DataObject&#39;s Count Matrix duplicates.</span>
<span class="sd">        Drops duplicates in DataObject following a &#39;keep&#39; criteria.</span>
<span class="sd">        Control that the rest of the data match with new Count Matrix</span>
<span class="sd">        </span>
<span class="sd">    </span>
<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param axis: axis 0 means by rows, 1 by columns</span>
<span class="sd">    :type axis: int</span>
<span class="sd">    :param keep: Indicates which duplicates will keep on Count Matrix.</span>
<span class="sd">        \&quot;first\&quot; keeps only the first duplicated column/row; \&quot;last\&quot; keeps only the last duplicated column/row; False keeps both duplicated columns/rows</span>
<span class="sd">    :type keep: str, bool</span>
<span class="sd">    :param by_name: True if only evaluate the name of the column/row, False if evaluate the content of the column/row to decide if is duplicated or not</span>
<span class="sd">    :type by_name: bool</span>
<span class="sd">   </span>
<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">CountsDropDuplicates</span><span class="p">(</span><span class="n">axis</span><span class="p">,</span> <span class="n">keep</span><span class="p">,</span> <span class="n">by_name</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>


<div class="viewcode-block" id="counts_drop_values"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.counts_drop_values">[docs]</a><span class="k">def</span> <span class="nf">counts_drop_values</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">to_delete</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Drop DataObject&#39;s Count Matrix values.</span>
<span class="sd">        Control that the rest of the data match with new Count Matrix</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>

<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param to_delete: value to be delete.</span>
<span class="sd">    :type to_delete: str, float, int, bool, list, ndarray, </span>
<span class="sd">    :param axis: Indicates 0 to row and 1 to columns</span>
<span class="sd">    :type axis: int</span>
<span class="sd">    :param method: all to check all values are the same, any if a partial value match</span>
<span class="sd">    :type method: str</span>

<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">CountsDropValues</span><span class="p">(</span><span class="n">to_delete</span> <span class="o">=</span> <span class="n">to_delete</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">axis</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="n">method</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>


<div class="viewcode-block" id="counts_drop_nan_by_thresh"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.counts_drop_nan_by_thresh">[docs]</a><span class="k">def</span> <span class="nf">counts_drop_nan_by_thresh</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">thresh</span> <span class="o">=</span> <span class="mi">0</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Drop DataObject&#39;s Count Matrix values limited by Threshold.</span>
<span class="sd">        Control that the rest of the data match with new Count Matrix</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>

<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param axis: Indicates 0 to row and 1 to columns</span>
<span class="sd">    :type axis: int</span>
<span class="sd">    :param thresh: threshold to evaluate element. If value less than threshold, that will be selected to drop</span>
<span class="sd">    :type thresh: float, int</span>

<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">CountsDropNanByThresh</span><span class="p">(</span><span class="n">axis</span><span class="p">,</span> <span class="n">thresh</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>


<div class="viewcode-block" id="counts_replace"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.counts_replace">[docs]</a><span class="k">def</span> <span class="nf">counts_replace</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">to_replace</span> <span class="o">=</span> <span class="s2">&quot;Unknown&quot;</span><span class="p">,</span> <span class="n">replaced_by</span> <span class="o">=</span> <span class="n">ut</span><span class="o">.</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Replace DataObject&#39;s Count Matrix values with another.</span>
<span class="sd">        Control that the rest of the data match with new Count Matrix</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>

<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param to_replace: values to be replaced</span>
<span class="sd">    :type to_replace: str, float, int, bool, list, ndarray</span>
<span class="sd">    :param replaced_by: new value to replaced the value before</span>
<span class="sd">    :type replaced_by: str, float, int, bool</span>

<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">CountsReplace</span><span class="p">(</span><span class="n">to_replace</span><span class="p">,</span> <span class="n">replaced_by</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>


<div class="viewcode-block" id="counts_transpose"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.counts_transpose">[docs]</a><span class="k">def</span> <span class="nf">counts_transpose</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">index_name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Do transpose to DataObject&#39;s Count Matrix</span>
<span class="sd">        Control that the rest of the data match with new Count Matrix</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>
<span class="sd">    </span>
<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param index_name: the name of the index before transpose to keep it</span>
<span class="sd">    :type index_name: str, int</span>

<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">CountsTranspose</span><span class="p">(</span><span class="n">index_name</span> <span class="o">=</span> <span class="n">index_name</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>


<div class="viewcode-block" id="counts_cpm"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.counts_cpm">[docs]</a><span class="k">def</span> <span class="nf">counts_cpm</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">log_method</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">prior_count</span> <span class="o">=</span> <span class="mi">2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Normalize DataObject&#39;s Count Matrix using CPM method</span>
<span class="sd">        It only modify values inside Count Matrix.</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>
<span class="sd">        </span>
<span class="sd">    :param dobject: The DataObject to apply normalization.</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param log_method: True if CPM will do with logaritmic alghoritm, False otherwise</span>
<span class="sd">    :type log_method: bool</span>
<span class="sd">    :param prior_count: If log is True, ends up getting scaled by the ratio of a library size to the average library size and then multiplied by its value before getting added to each library size</span>
<span class="sd">    :type prior_count: int</span>

<span class="sd">    :return: The DataObject with CPM&#39;s values.</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">CountsCPM</span><span class="p">(</span><span class="n">log_method</span> <span class="o">=</span> <span class="n">log_method</span><span class="p">,</span> <span class="n">prior_count</span> <span class="o">=</span> <span class="n">prior_count</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>



<span class="sd">&quot;&quot;&quot;</span>

<span class="sd">OBSERVATIONS</span>

<span class="sd">&quot;&quot;&quot;</span>


<div class="viewcode-block" id="set_obs"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.set_obs">[docs]</a><span class="k">def</span> <span class="nf">set_obs</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">new_obs</span> <span class="o">=</span> <span class="n">ut</span><span class="o">.</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()):</span>
    <span class="sd">&quot;&quot;&quot;Set DataObject&#39;s Clinical Information to new one.</span>
<span class="sd">        Control that the rest of the data match with new Observations</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>
<span class="sd">    </span>
<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param new_obs: new Clinical Information</span>
<span class="sd">    :type new_obs: DataFrame</span>

<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">SetObs</span><span class="p">(</span><span class="n">new_obs</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                           ACCESSING OBS DATAFRAME </span>
<span class="c1"># =============================================================================</span>


<div class="viewcode-block" id="obs_selection"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.obs_selection">[docs]</a><span class="k">def</span> <span class="nf">obs_selection</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">row_list</span> <span class="o">=</span> <span class="p">[]):</span>
    <span class="sd">&quot;&quot;&quot;Set DataObject&#39;s Observations selecting a specific row list.</span>
<span class="sd">        Control that the rest of the data match with new Observations</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>
<span class="sd">    </span>
<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param row_list: A list with the names of the rows to keep</span>
<span class="sd">    :type row_list: list</span>

<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">ObsSelection</span><span class="p">(</span><span class="n">row_list</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>


<div class="viewcode-block" id="obs_projection"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.obs_projection">[docs]</a><span class="k">def</span> <span class="nf">obs_projection</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">col_list</span> <span class="o">=</span> <span class="p">[]):</span>
    <span class="sd">&quot;&quot;&quot;Set DataObject&#39;s Observations selecting a specific column list.</span>
<span class="sd">        Control that the rest of the data match with new Observations</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>
<span class="sd">    </span>
<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param col_list: A list with the names of the columns to keep</span>
<span class="sd">    :type col_list: list</span>

<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">ObsProjection</span><span class="p">(</span><span class="n">col_list</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>


<div class="viewcode-block" id="obs_add_column"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.obs_add_column">[docs]</a><span class="k">def</span> <span class="nf">obs_add_column</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">name_column</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">values</span> <span class="o">=</span> <span class="p">[]):</span>
    <span class="sd">&quot;&quot;&quot;Add to DataObject&#39;s Observations a new column</span>
<span class="sd">        It is not neccesary control that the rest of the data match with new Observations.</span>
<span class="sd">        Add new column doesn&#39;t affect the rest of the DataObject</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>
<span class="sd">    </span>
<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param name_column: Name of the new column</span>
<span class="sd">    :type name_column: str</span>
<span class="sd">    :param values: values of the new column. Must have the same lenght as number of rows of Observations</span>
<span class="sd">    :type values: list, ndarray</span>

<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">ObsAddColumn</span><span class="p">(</span><span class="n">name_column</span><span class="p">,</span> <span class="n">values</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>


<div class="viewcode-block" id="obs_replace"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.obs_replace">[docs]</a><span class="k">def</span> <span class="nf">obs_replace</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">to_replace</span> <span class="o">=</span> <span class="s2">&quot;Unknown&quot;</span><span class="p">,</span> <span class="n">replaced_by</span> <span class="o">=</span> <span class="n">ut</span><span class="o">.</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Replace DataObject&#39;s Observations values with another.</span>
<span class="sd">        Control that the rest of the data match with new Observations</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>

<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param to_replace: values to be replaced</span>
<span class="sd">    :type to_replace: str, float, int, bool, list, ndarray</span>
<span class="sd">    :param replaced_by: new value to replaced the value before</span>
<span class="sd">    :type replaced_by: str, float, int, bool</span>

<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">ObsReplace</span><span class="p">(</span><span class="n">to_replace</span><span class="p">,</span> <span class="n">replaced_by</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>


<div class="viewcode-block" id="obs_drop_duplicates"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.obs_drop_duplicates">[docs]</a><span class="k">def</span> <span class="nf">obs_drop_duplicates</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">keep</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">by_name</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Drop DataObject&#39;s Observations duplicates.</span>
<span class="sd">        Control that the rest of the data match with new Observations</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>
<span class="sd">    </span>
<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param axis: axis 0 means by rows, 1 by columns</span>
<span class="sd">    :type axis: int</span>
<span class="sd">    :param keep: Indicates which duplicates will keep on Observations.</span>
<span class="sd">        \&quot;first\&quot; keeps only the first duplicated column/row; \&quot;last\&quot; keeps only the last duplicated column/row; False keeps both duplicated columns/rows</span>
<span class="sd">    :type keep: str, bool</span>
<span class="sd">    :param by_name: True if only evaluate the name of the column/row, False if evaluate the content of the column/row to decide if is duplicated or not</span>
<span class="sd">    :type by_name: bool</span>

<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">ObsDropDuplicates</span><span class="p">(</span><span class="n">axis</span><span class="p">,</span> <span class="n">keep</span><span class="p">,</span> <span class="n">by_name</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>



<span class="sd">&quot;&quot;&quot;</span>

<span class="sd">VARIABLES</span>

<span class="sd">&quot;&quot;&quot;</span>


<div class="viewcode-block" id="set_vars"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.set_vars">[docs]</a><span class="k">def</span> <span class="nf">set_vars</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">new_var</span> <span class="o">=</span> <span class="n">ut</span><span class="o">.</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()):</span>
    <span class="sd">&quot;&quot;&quot;Set DataObject&#39;s Variables to new one.</span>
<span class="sd">        Control that the rest of the data match with new Variables</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>
<span class="sd">    </span>
<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param new_var: new Gene Information</span>
<span class="sd">    :type new_var: DataFrame</span>
<span class="sd">    </span>
<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">SetVars</span><span class="p">(</span><span class="n">new_var</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>
    



<span class="c1"># =============================================================================</span>
<span class="c1">#                           ACCESSING VAR DATAFRAME </span>
<span class="c1"># =============================================================================</span>


<div class="viewcode-block" id="vars_selection"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.vars_selection">[docs]</a><span class="k">def</span> <span class="nf">vars_selection</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">row_list</span> <span class="o">=</span> <span class="p">[]):</span>
    <span class="sd">&quot;&quot;&quot;Set DataObject&#39;s Variables selecting a specific row list.</span>
<span class="sd">        Control that the rest of the data match with new Variables</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>
<span class="sd">    </span>
<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param row_list: A list with the names of the rows to keep</span>
<span class="sd">    :type row_list: list</span>

<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">ObsSelection</span><span class="p">(</span><span class="n">row_list</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>


<div class="viewcode-block" id="vars_projection"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.vars_projection">[docs]</a><span class="k">def</span> <span class="nf">vars_projection</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">col_list</span> <span class="o">=</span> <span class="p">[]):</span>
    <span class="sd">&quot;&quot;&quot;Set DataObject&#39;s Variables selecting a specific column list.</span>
<span class="sd">        Control that the rest of the data match with new Variables</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>
<span class="sd">    </span>
<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param col_list: A list with the names of the columns to keep</span>
<span class="sd">    :type col_list: list</span>

<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">VarsProjection</span><span class="p">(</span><span class="n">col_list</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>


<div class="viewcode-block" id="vars_add_column"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.vars_add_column">[docs]</a><span class="k">def</span> <span class="nf">vars_add_column</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">name_column</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">values</span> <span class="o">=</span> <span class="p">[]):</span>
    <span class="sd">&quot;&quot;&quot;Add to DataObject&#39;s Variables a new column</span>
<span class="sd">        It is not neccesary control that the rest of the data match with new Variables.</span>
<span class="sd">        Add new column doesn&#39;t affect the rest of the DataObject</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>
<span class="sd">    </span>
<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param name_column: Name of the new column</span>
<span class="sd">    :type name_column: str</span>
<span class="sd">    :param values: values of the new column. Must have the same lenght as number of rows of Variables</span>
<span class="sd">    :type values: list, ndarray</span>

<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">VarsAddColumn</span><span class="p">(</span><span class="n">name_column</span><span class="p">,</span> <span class="n">values</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>


<div class="viewcode-block" id="vars_replace"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.vars_replace">[docs]</a><span class="k">def</span> <span class="nf">vars_replace</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">to_replace</span> <span class="o">=</span> <span class="s2">&quot;Unknown&quot;</span><span class="p">,</span> <span class="n">replaced_by</span> <span class="o">=</span> <span class="n">ut</span><span class="o">.</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Replace DataObject&#39;s Variables values with another.</span>
<span class="sd">        Control that the rest of the data match with new Variables</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>

<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param to_replace: values to be replaced</span>
<span class="sd">    :type to_replace: str, float, int, bool, list, ndarray</span>
<span class="sd">    :param replaced_by: new value to replaced the value before</span>
<span class="sd">    :type replaced_by: str, float, int, bool</span>

<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">VarsReplace</span><span class="p">(</span><span class="n">to_replace</span><span class="p">,</span> <span class="n">replaced_by</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>


<div class="viewcode-block" id="vars_drop_duplicates"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.vars_drop_duplicates">[docs]</a><span class="k">def</span> <span class="nf">vars_drop_duplicates</span><span class="p">(</span><span class="n">dobject</span> <span class="p">:</span> <span class="n">DataObject</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">keep</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">by_name</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Drop DataObject&#39;s Variables duplicates.</span>
<span class="sd">        Control that the rest of the data match with new Variables</span>
<span class="sd">        This class is a Task which allows generate a Workflow</span>
<span class="sd">        When this class is called, save his instance on DataObject&#39;s Log</span>
<span class="sd">    </span>
<span class="sd">    :param dobject: The current DataObject to be process</span>
<span class="sd">    :type dobject: DataObject</span>
<span class="sd">    :param axis: axis 0 means by rows, 1 by columns</span>
<span class="sd">    :type axis: int</span>
<span class="sd">    :param keep: Indicates which duplicates will keep on Observations.</span>
<span class="sd">        \&quot;first\&quot; keeps only the first duplicated column/row; \&quot;last\&quot; keeps only the last duplicated column/row; False keeps both duplicated columns/rows</span>
<span class="sd">    :type keep: str, bool</span>
<span class="sd">    :param by_name: True if only evaluate the name of the column/row, False if evaluate the content of the column/row to decide if is duplicated or not</span>
<span class="sd">    :type by_name: bool</span>

<span class="sd">    :return: The DataObject passed with the Task applied</span>
<span class="sd">    :rtype: DataObject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dobj</span><span class="o">.</span><span class="n">VarsDropDuplicates</span><span class="p">(</span><span class="n">axis</span><span class="p">,</span> <span class="n">keep</span><span class="p">,</span> <span class="n">by_name</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>



<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">######################################################</span>

<span class="sd">                        MODEL</span>

<span class="sd">######################################################</span>
<span class="sd">&quot;&quot;&quot;</span>


<div class="viewcode-block" id="create_model_selection"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.create_model_selection">[docs]</a><span class="k">def</span> <span class="nf">create_model_selection</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">source</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">target</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;This class performs different techniques to the data such as: splitting data,</span>
<span class="sd">        over-sampling and under-sampling, normalization, computes confusion matrix and AUC, etc.,</span>
<span class="sd">        to the data in order to improve Machine Learning algorithm&#39;s performance.</span>
<span class="sd">    </span>
<span class="sd">    :param data: The data to apply Machine Learning methods. Defaults to None.</span>
<span class="sd">        If it is not specified, user must provide the `source` and `target` values.</span>
<span class="sd">    :type data: DataFrame</span>
<span class="sd">    :param source: The features that describe the samples. Defaults to None.</span>
<span class="sd">        If it is noy specified, user must provide `data` value. </span>
<span class="sd">        Otherwise, if it is different to None, `target` values must be provided.</span>
<span class="sd">    :type source: DataFrame</span>
<span class="sd">    :param target: The labels to be predicted. Defaults to None.</span>
<span class="sd">        If it is noy specified, user must provide `data` value. </span>
<span class="sd">        Otherwise, if it is different to None, `source` values must be provided.</span>
<span class="sd">    :type target: Series</span>
<span class="sd">    :param X_train: All the observations that will be used to train the model.</span>
<span class="sd">        Defaults to None. Only can be updated after use `get_train_test_sample` method.</span>
<span class="sd">    :type X_train: DataFrame</span>
<span class="sd">    :param Y_train: The dependent variable wich need to be predicted by the model.</span>
<span class="sd">        Defaults to None. Only can be updated after use `get_train_test_sample` method.</span>
<span class="sd">    :type Y_train: Series</span>
<span class="sd">    :param X_test:  The remaining portion of the independent  variables which </span>
<span class="sd">        will not be used in the training phase and will be used</span>
<span class="sd">        to make predictions to test the accuracy of the model.</span>
<span class="sd">        Defaults to None. Only can be updated after use `get_train_test_sample` method.</span>
<span class="sd">    :type X_test: DataFrame</span>
<span class="sd">    :param Y_test: The labels of the test data. These labels will be used to test </span>
<span class="sd">        the accuracy between actual and predicted categories. Defaults to None.</span>
<span class="sd">        Only can be updated after use `get_train_test_sample` method.</span>
<span class="sd">    :type Y_test: Series</span>

<span class="sd">    :return: a Model Selection with a data used on Machine Learning</span>
<span class="sd">    :rtype: ModelSelection</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ModelSelection</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">,</span> <span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span> <span class="n">target</span> <span class="o">=</span> <span class="n">target</span><span class="p">)</span></div>


<div class="viewcode-block" id="model_logistic_regression"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.model_logistic_regression">[docs]</a><span class="k">def</span> <span class="nf">model_logistic_regression</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">penalty</span> <span class="o">=</span> <span class="s1">&#39;l2&#39;</span><span class="p">,</span> <span class="n">dual</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">tol</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">pow</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="o">-</span><span class="mi">4</span><span class="p">),</span>
                 <span class="n">C</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span> <span class="n">fit_intercept</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">intercept_scaling</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">class_weight</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">random_state</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                 <span class="n">solver</span> <span class="o">=</span> <span class="s1">&#39;lbfgs&#39;</span><span class="p">,</span> <span class="n">max_iter</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="n">multi_class</span> <span class="o">=</span> <span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">warm_start</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> 
                 <span class="n">n_jobs</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">l1_ratio</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Generate a Logistic Regression algorithm from scikit-learn library.</span>

<span class="sd">    :param name: Name of the model.</span>
<span class="sd">    :type name: str</span>
<span class="sd">    :param penalty: Specify the norm of the penalty, where `none` means no penalty is added,</span>
<span class="sd">        `l2` uses Ridge regularization, `l1` uses Lasso regularization, and `elasticnet`</span>
<span class="sd">        means that both `l1` and `l2` penalty terms are added. Defaults to `l2`.</span>
<span class="sd">    :type penalty: `l1`,`l2`,`elasticnet`, `none`</span>
<span class="sd">    :param dual: Dual or primal formulation. Dual formulation is only implemented </span>
<span class="sd">        for l2 penalty with liblinear solver. Defauls to `False`.</span>
<span class="sd">    :type dual: bool</span>
<span class="sd">    :param tol: Tolerance for stopping criteria. Defaults to `1e-4`.</span>
<span class="sd">    :type tol: float</span>
<span class="sd">    :param C: Inverse of regularization strenght, must be a positive float. </span>
<span class="sd">        Smaller values specify stronger regularization. Defaults to `1.0`.</span>
<span class="sd">    :type C: float</span>
<span class="sd">    :param fit_intercept: Specifies if a constant (bias or intercept) should be added</span>
<span class="sd">        to the decision function. Defaults to `True`.</span>
<span class="sd">    :type fit_intercept: bool</span>
<span class="sd">    :param intercept_scaling: Useful only when the solver `liblinear` is used and</span>
<span class="sd">        `fit_intercept` True. In this case, x becomes [x, self.intercept_scaling].</span>
<span class="sd">        Defaults to `1`.</span>
<span class="sd">    :type intercept_scaling: float</span>
<span class="sd">    :param class_weight: Weights associated with classes in the form {class_label: weight}.</span>
<span class="sd">        Defaults to None.</span>
<span class="sd">    :type class_weight: dict, `balanced`, or None</span>
<span class="sd">    :param random_state: Used when solver == `sag`, `saga` or `liblinear` to shuffle the data. </span>
<span class="sd">        Defaults to None.</span>
<span class="sd">    :type random_state: int, RandomState instance, or None</span>
<span class="sd">    :param solver: Algorith to use in the optimization problem. Defaults to `lbfgs`.</span>
<span class="sd">    :type solver: {`newton-cg`, `lbfgs`, `liblinear`, `sag`, `saga`}</span>
<span class="sd">    :param max_iter: Maximum number of iterations taken for the solvers to converge.</span>
<span class="sd">        Defaults to `100`.</span>
<span class="sd">    :type max_iter: int</span>
<span class="sd">    :param multi_class: If the option chosen is `ovr`, then a binary problem is fit </span>
<span class="sd">        for each label. For `multinomial` the loss minimised is the multinomial loss fit</span>
<span class="sd">        across the entire probability distribution, even when the data is binary.</span>
<span class="sd">        Defaults to `auto`.</span>
<span class="sd">    :type multi_class: {`auto` , `ovr` , `multinomial`}</span>
<span class="sd">    :param verbose: For the liblinear and lbfgs solvers set verbose to any positive</span>
<span class="sd">        number for verbosity. Defaults to ``0`.</span>
<span class="sd">    :type verbose: int</span>
<span class="sd">    :param warm_start: If `True`, reuse the solution of the previous call to fit</span>
<span class="sd">        as initialization, otherwise, just erase the previous solution. Defauls to `False`.</span>
<span class="sd">    :type warm_start: bool</span>
<span class="sd">    :param n_jobs: Number of CPU cores used when parallelizing over classes of multi_class=`ovr`.</span>
<span class="sd">        Defaults to None.</span>
<span class="sd">    :type n_jobs: int</span>
<span class="sd">    :param l1_ratio: The Elastic-Net mixing parameter, with 0 &lt;= l1_ratio &lt;= 1.</span>
<span class="sd">        Only used if penalty = `elasticnetV. Defaults to None.</span>
<span class="sd">    :type l1_ratio: float</span>

<span class="sd">    :return: A Logistic Regression model initialize with arguments</span>
<span class="sd">    :rtype: LogisticRegression</span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="k">return</span> <span class="n">LogisticRegression</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span><span class="p">,</span> <span class="n">penalty</span> <span class="o">=</span> <span class="n">penalty</span><span class="p">,</span> <span class="n">dual</span> <span class="o">=</span> <span class="n">dual</span><span class="p">,</span> <span class="n">tol</span> <span class="o">=</span> <span class="n">tol</span><span class="p">,</span>
                 <span class="n">C</span> <span class="o">=</span> <span class="n">C</span><span class="p">,</span> <span class="n">fit_intercept</span> <span class="o">=</span> <span class="n">fit_intercept</span><span class="p">,</span> <span class="n">intercept_scaling</span> <span class="o">=</span> <span class="n">intercept_scaling</span><span class="p">,</span> <span class="n">class_weight</span> <span class="o">=</span> <span class="n">class_weight</span><span class="p">,</span> <span class="n">random_state</span> <span class="o">=</span> <span class="n">random_state</span><span class="p">,</span>
                 <span class="n">solver</span> <span class="o">=</span> <span class="n">solver</span><span class="p">,</span> <span class="n">max_iter</span> <span class="o">=</span> <span class="n">max_iter</span><span class="p">,</span> <span class="n">multi_class</span> <span class="o">=</span> <span class="n">multi_class</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="n">verbose</span><span class="p">,</span> <span class="n">warm_start</span> <span class="o">=</span> <span class="n">warm_start</span><span class="p">,</span> 
                 <span class="n">n_jobs</span> <span class="o">=</span> <span class="n">n_jobs</span><span class="p">,</span> <span class="n">l1_ratio</span> <span class="o">=</span> <span class="n">l1_ratio</span><span class="p">)</span></div>


<div class="viewcode-block" id="model_random_forest"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.model_random_forest">[docs]</a><span class="k">def</span> <span class="nf">model_random_forest</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">n_estimators</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="n">criterion</span> <span class="o">=</span> <span class="s1">&#39;gini&#39;</span><span class="p">,</span> <span class="n">max_depth</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">min_samples_split</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
               <span class="n">min_samples_leaf</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">min_weight_fraction_leaf</span> <span class="o">=</span> <span class="mf">0.0</span><span class="p">,</span> <span class="n">max_features</span> <span class="o">=</span> <span class="s1">&#39;sqrt&#39;</span><span class="p">,</span> <span class="n">max_leaf_nodes</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
               <span class="n">min_impurity_decrease</span> <span class="o">=</span> <span class="mf">0.0</span><span class="p">,</span> <span class="n">bootstrap</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">oob_score</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">n_jobs</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">random_state</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
               <span class="n">warm_start</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">class_weight</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">ccp_alpha</span> <span class="o">=</span> <span class="mf">0.0</span><span class="p">,</span> <span class="n">max_samples</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Generate a Random Forest classifier algorithm from scikit-learn library.</span>

<span class="sd">    :param name: Name of the model.</span>
<span class="sd">    :type name: str</span>
<span class="sd">    :param n_estimators: Number of trees in the forest. Defaults to `100`.</span>
<span class="sd">    :type n_estimators: int</span>
<span class="sd">    :param criterion: Function to measure the quality of a split. Defaults to `gini`.</span>
<span class="sd">    :type criterion: {`gini`, `entropy`, `log_loss`}.</span>
<span class="sd">    :param max_depth: The maximum depth of the tree. If None, nodes are expanded</span>
<span class="sd">        until all leaves are pure or contains less than `min_sample_split` samples.</span>
<span class="sd">        Defaults to None.</span>
<span class="sd">    :type max_depth: int, None</span>
<span class="sd">    :param min_samples_split: The minimum number of samples required to split an</span>
<span class="sd">        internal node. Defaults to `2`.</span>
<span class="sd">    :type min_samples_split: int or float</span>
<span class="sd">    :param min_samples_leaf: The minimum number of samples required to be at a leaf node.</span>
<span class="sd">        Defaults to `1`.</span>
<span class="sd">    :type min_samples_leaf: int or float</span>
<span class="sd">    :param min_weight_fraction_leaf: The minimum weighted fraction of the sum total</span>
<span class="sd">        of weights (of all the input samples) required to be at a leaf node.</span>
<span class="sd">        Defaults to `0.0`.</span>
<span class="sd">    :type min_weight_fraction_leaf: float</span>
<span class="sd">    :param max_features: The number of features to consider when looking for</span>
<span class="sd">        the best split. Defaults to `sqrt`.</span>
<span class="sd">    :type max_features:{`sqrt`, `log2`, None}, int or float</span>
<span class="sd">    :param max_leaf_nodes: Grow trees with `max_leaf_nodes` in best-first fashion.</span>
<span class="sd">        Defaults to None.</span>
<span class="sd">    :type max_leaf_nodes: int, None</span>
<span class="sd">    :param min_impurity_decrease: A node will be split if this split induces a decrease</span>
<span class="sd">        of the impurity greater than or equal to this value. Defaults to `0.0`.</span>
<span class="sd">    :type min_impurity_decrease: float</span>
<span class="sd">    :param bootstrap: Whether bootstrap samples are used when building trees. </span>
<span class="sd">        If `False`, the whole dataset is used to build each tree. Defauls to `True`.</span>
<span class="sd">    :type bootstrap: bool</span>
<span class="sd">    :param oob_score: Wheter to use out-of-bag samples to estimate the generalization score.</span>
<span class="sd">        Only available if boostrap is `True`. Defaults to `False`.</span>
<span class="sd">    :type oob_score: bool</span>
<span class="sd">    :param n_jobs: Number of CPU cores used when parallelizing over classes of multi_class=`ovr`.</span>
<span class="sd">        Defaults to None.</span>
<span class="sd">    :type n_jobs: int</span>
<span class="sd">    :param random_state: Used when solver == `sag`, `saga` or `liblinear` to shuffle the data. </span>
<span class="sd">        Defaults to None.</span>
<span class="sd">    :type random_state: int, RandomState instance, or None</span>
<span class="sd">    :param verbose: For the liblinear and lbfgs solvers set verbose to any positive</span>
<span class="sd">        number for verbosity. Defaults to `0`.</span>
<span class="sd">    :type verbose: int</span>
<span class="sd">    :param warm_start: If `True`, reuse the solution of the previous call to fit</span>
<span class="sd">        as initialization, otherwise, just erase the previous solution. Defauls to `False`.</span>
<span class="sd">    :type warm_start: bool</span>
<span class="sd">    :param class_weight: Weights associated with classes in the form {class_label: weight}.</span>
<span class="sd">        Defaults to None.</span>
<span class="sd">    :type class_weight: dict, `balanced`, `balanced_subsample`, or None</span>
<span class="sd">    :param ccp_alpha: Complexity parameter used for Minimal Cost-Complexity Pruning.</span>
<span class="sd">        Defaults to `0.0`.</span>
<span class="sd">    :type ccp_alpha: non-negative float</span>
<span class="sd">    :param max_samples: If boostrap is `True`, the number of samples to draw from X to </span>
<span class="sd">        train each base estimator. Defaults to None.</span>
<span class="sd">    :type max_samples: int, float, None</span>

<span class="sd">    :return: Random Forest classifier algorithm</span>
<span class="sd">    :rtype: RandomForest</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">RandomForest</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span><span class="p">,</span> <span class="n">n_estimators</span> <span class="o">=</span> <span class="n">n_estimators</span><span class="p">,</span> <span class="n">criterion</span> <span class="o">=</span> <span class="n">criterion</span><span class="p">,</span> <span class="n">max_depth</span> <span class="o">=</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">min_samples_split</span> <span class="o">=</span> <span class="n">min_samples_split</span><span class="p">,</span>
               <span class="n">min_samples_leaf</span> <span class="o">=</span> <span class="n">min_samples_leaf</span><span class="p">,</span> <span class="n">min_weight_fraction_leaf</span> <span class="o">=</span> <span class="n">min_weight_fraction_leaf</span><span class="p">,</span> <span class="n">max_features</span> <span class="o">=</span> <span class="n">max_features</span><span class="p">,</span> <span class="n">max_leaf_nodes</span> <span class="o">=</span> <span class="n">max_leaf_nodes</span><span class="p">,</span> 
               <span class="n">min_impurity_decrease</span> <span class="o">=</span> <span class="n">min_impurity_decrease</span><span class="p">,</span> <span class="n">bootstrap</span> <span class="o">=</span> <span class="n">bootstrap</span><span class="p">,</span> <span class="n">oob_score</span> <span class="o">=</span> <span class="n">oob_score</span><span class="p">,</span> <span class="n">n_jobs</span> <span class="o">=</span> <span class="n">n_jobs</span><span class="p">,</span> <span class="n">random_state</span> <span class="o">=</span> <span class="n">random_state</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="n">verbose</span><span class="p">,</span>
               <span class="n">warm_start</span> <span class="o">=</span> <span class="n">warm_start</span><span class="p">,</span> <span class="n">class_weight</span> <span class="o">=</span> <span class="n">class_weight</span><span class="p">,</span> <span class="n">ccp_alpha</span> <span class="o">=</span> <span class="n">ccp_alpha</span><span class="p">,</span> <span class="n">max_samples</span> <span class="o">=</span> <span class="n">max_samples</span><span class="p">)</span></div>


<div class="viewcode-block" id="model_scikit_learn"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.model_scikit_learn">[docs]</a><span class="k">def</span> <span class="nf">model_scikit_learn</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">model</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Creates a generic model that is provided by</span>
<span class="sd">    the Scikit-learn library.</span>

<span class="sd">    :param name: Name of the model.</span>
<span class="sd">    :type name: str</span>
<span class="sd">    :param model: Model to work with.</span>
<span class="sd">    :type model: object</span>

<span class="sd">    :return: a generic ScikitLearn model</span>
<span class="sd">    :rtype: ScikitLearnModel</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ScikitLearnModel</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">model</span><span class="p">)</span></div>


<div class="viewcode-block" id="model_support_vector_classif"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.model_support_vector_classif">[docs]</a><span class="k">def</span> <span class="nf">model_support_vector_classif</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">C</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span> <span class="n">kernel</span> <span class="o">=</span> <span class="s1">&#39;rbf&#39;</span><span class="p">,</span> <span class="n">degree</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">gamma</span> <span class="o">=</span> <span class="s1">&#39;scale&#39;</span><span class="p">,</span> <span class="n">coef0</span> <span class="o">=</span> <span class="mf">0.0</span><span class="p">,</span> 
               <span class="n">shrinking</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">probability</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">tol</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">pow</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="o">-</span><span class="mi">3</span><span class="p">),</span> <span class="n">cache_size</span> <span class="o">=</span> <span class="mi">200</span><span class="p">,</span> <span class="n">class_weight</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
               <span class="n">verbose</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">max_iter</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">decision_function_shape</span> <span class="o">=</span> <span class="s1">&#39;ovr&#39;</span><span class="p">,</span> <span class="n">break_ties</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">random_state</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Generate a Support Vector Machine algorithm from scikit-learn library.</span>

<span class="sd">    :param name: Name of the model.</span>
<span class="sd">    :type name: str</span>
<span class="sd">    :param C: Regularization parameter. The strenght of the regularization is </span>
<span class="sd">        inversely proportional to C. Defaults to `1.0`.</span>
<span class="sd">    :type C: float</span>
<span class="sd">    :param kernel: Specifies the kernel type to be used in the algorithm. </span>
<span class="sd">        Defaults to `rbf`</span>
<span class="sd">    :type kernel: {`linear`, `poly`, `rbf`, `sigmoid`, `precomputed`} or callable</span>
<span class="sd">    :param degree: Degree of the polynomial kernel function when kernel is `poly`.</span>
<span class="sd">        Defaults to `3`.</span>
<span class="sd">    :type degree: int</span>
<span class="sd">    :param gamma: Kernek coefficient fot `rbf`, `poly` and `sigmoid`.</span>
<span class="sd">        Defaults to `scale`.</span>
<span class="sd">    :type gamma: {`scale`, `auto`} or float</span>
<span class="sd">    :param coef0: Independent term in kernel function, when kernel is `poly` or</span>
<span class="sd">        `sigmoid`. Defaults to `0.0`.</span>
<span class="sd">    :type coef0: float</span>
<span class="sd">    :param shrinking: Whether to use the shrinking heuristic. Defaults to `True`.</span>
<span class="sd">    :type shrinking: bool</span>
<span class="sd">    :param probability: Wheter to enable probability estimates. Deafuls to `False`.</span>
<span class="sd">    :type probability: bool</span>
<span class="sd">    :param tol: Tolerance for stopping criteria. Defaults to `1e-3`.</span>
<span class="sd">    :type tol: float</span>
<span class="sd">    :param cache_size: Specify the size of the kernelk cache (in MB). Defaults to `200`.</span>
<span class="sd">    :type cache_size: float</span>
<span class="sd">    :param class_weight: Weights associated with classes in the form {class_label: weight}.</span>
<span class="sd">        Defaults to None.</span>
<span class="sd">    :type class_weight: dict, `balanced`, or None</span>
<span class="sd">    :param verbose: For the liblinear and lbfgs solvers set verbose to any positive</span>
<span class="sd">            number for verbosity.  Defaults to None.</span>
<span class="sd">    :type verbose: bool</span>
<span class="sd">    :param max_iter: Hard limit on iterations within solver, or -1 for no limit.</span>
<span class="sd">        Defaults to -1.</span>
<span class="sd">    :type max_iter: int</span>
<span class="sd">    :param decision_function_shape: Wheter to return a one-vs-rest (`ovr`) or </span>
<span class="sd">        one-vs-one (`ovo`) decision function. Defaults to `ovr`.</span>
<span class="sd">    :type decision_function_shape: {`ovo`, `ovr`}</span>
<span class="sd">    :param break_ties: If `True`, `decision_functoin_shape` = `ovr` and number of</span>
<span class="sd">        classes &gt; 2, predict will break ties according to the confidence values</span>
<span class="sd">        of `decision_function`. Defaults to `False`.</span>
<span class="sd">    :type break_ties: bool</span>
<span class="sd">    :param random_state: Used when solver == `sag`, `saga` or `liblinear` to shuffle the data. </span>
<span class="sd">        Defaults to None.</span>
<span class="sd">    :type random_state: int, RandomState instance, or None</span>

<span class="sd">    :return: a support vector machine algorithm</span>
<span class="sd">    :rtype: SupportVectorClassif</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">SupportVectorClassif</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span><span class="p">,</span> <span class="n">C</span> <span class="o">=</span> <span class="n">C</span><span class="p">,</span> <span class="n">kernel</span> <span class="o">=</span> <span class="n">kernel</span><span class="p">,</span> <span class="n">degree</span> <span class="o">=</span> <span class="n">degree</span><span class="p">,</span> <span class="n">gamma</span> <span class="o">=</span> <span class="n">gamma</span><span class="p">,</span> <span class="n">coef0</span> <span class="o">=</span> <span class="n">coef0</span><span class="p">,</span> 
               <span class="n">shrinking</span> <span class="o">=</span> <span class="n">shrinking</span><span class="p">,</span> <span class="n">probability</span> <span class="o">=</span> <span class="n">probability</span><span class="p">,</span> <span class="n">tol</span> <span class="o">=</span> <span class="n">tol</span><span class="p">,</span> <span class="n">cache_size</span> <span class="o">=</span> <span class="n">cache_size</span><span class="p">,</span> <span class="n">class_weight</span> <span class="o">=</span> <span class="n">class_weight</span><span class="p">,</span> 
               <span class="n">verbose</span> <span class="o">=</span> <span class="n">verbose</span><span class="p">,</span> <span class="n">max_iter</span> <span class="o">=</span> <span class="n">max_iter</span><span class="p">,</span> <span class="n">decision_function_shape</span> <span class="o">=</span> <span class="n">decision_function_shape</span><span class="p">,</span> <span class="n">break_ties</span> <span class="o">=</span> <span class="n">break_ties</span><span class="p">,</span> <span class="n">random_state</span> <span class="o">=</span> <span class="n">random_state</span><span class="p">)</span></div>


<div class="viewcode-block" id="model_neural_network"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.model_neural_network">[docs]</a><span class="k">def</span> <span class="nf">model_neural_network</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">hidden_layer_sizes</span> <span class="o">=</span> <span class="p">(</span><span class="mi">100</span><span class="p">,),</span> <span class="n">activation</span> <span class="o">=</span> <span class="s1">&#39;relu&#39;</span><span class="p">,</span> <span class="n">solver</span> <span class="o">=</span> <span class="s1">&#39;adam&#39;</span><span class="p">,</span>
               <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.0001</span><span class="p">,</span> <span class="n">batch_size</span> <span class="o">=</span> <span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="n">learning_rate</span> <span class="o">=</span> <span class="s1">&#39;constant&#39;</span><span class="p">,</span> <span class="n">learning_rate_init</span> <span class="o">=</span> <span class="mf">0.001</span><span class="p">,</span> <span class="n">power_t</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span>
               <span class="n">max_iter</span> <span class="o">=</span> <span class="mi">200</span><span class="p">,</span> <span class="n">shuffle</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">random_state</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">tol</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">pow</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="o">-</span><span class="mi">4</span><span class="p">),</span> <span class="n">verbose</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">warm_start</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
               <span class="n">momentum</span> <span class="o">=</span> <span class="mf">0.9</span><span class="p">,</span> <span class="n">nesterovs_momentum</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">early_stopping</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">validation_fraction</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> <span class="n">beta_1</span> <span class="o">=</span> <span class="mf">0.9</span><span class="p">,</span> 
               <span class="n">beta_2</span> <span class="o">=</span> <span class="mf">0.999</span><span class="p">,</span> <span class="n">epsilon</span> <span class="o">=</span>  <span class="n">math</span><span class="o">.</span><span class="n">pow</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="o">-</span><span class="mi">8</span><span class="p">),</span> <span class="n">n_iter_no_change</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">pow</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="o">-</span><span class="mi">8</span><span class="p">),</span> <span class="n">max_fun</span> <span class="o">=</span> <span class="mi">15000</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Generates a Neural Network algorithm from scikit-learn library.</span>

<span class="sd">    :param name: Name of the model.</span>
<span class="sd">    :type name: str</span>
<span class="sd">    :param hidden_layer_sizes: The ith element represents the number of neurons in the</span>
<span class="sd">        ith hidden layer. Defaults to `(100,)`.</span>
<span class="sd">    :type hidden_layer_sizes: tuple, length = n_layers - 2</span>
<span class="sd">    :param activation: Activation function for the hidden layer. Defaults to `relu`.</span>
<span class="sd">    :type activation: {`identity`, `logistic`, `tanh`, `relu`}</span>
<span class="sd">    :param solver: The solver for weight optimization. Defaults to `adam`.</span>
<span class="sd">    :type solver: {`lbfgs`, `sgd`, `adam`}</span>
<span class="sd">    :param alpha: Strength of the L2 regularization term (which is divided by the sample</span>
<span class="sd">        size when added to the loss). Defaults to `0.0001`.</span>
<span class="sd">    :type alpha: float</span>
<span class="sd">    :param batch_size: Size of minibatches for stochastic optimizers. Defaults to `auto`.</span>
<span class="sd">    :type batch_size: int, str</span>
<span class="sd">    :param learning_rate: Learning rate schedule for weight purposes. Defaults to `constant`.</span>
<span class="sd">    :type learning_rate: {`constant`, `invscaling`, `adaptive`}</span>
<span class="sd">    :param learning_rate_init: The initial learning rate used. It controls the step-size</span>
<span class="sd">        in updating the weights. Only when solver is `sgd` or `adam`. Defaults to `0.001`.</span>
<span class="sd">    :type learning_rate_init: float</span>
<span class="sd">    :param power_t: The exponent for inverse scaling learning rate. It is used in updating</span>
<span class="sd">        effective learning rate when the `learning_rate` is `invscaling`. Only when solver </span>
<span class="sd">        is `sgd`. Defaults to `0.5`</span>
<span class="sd">    :type power_t: float</span>
<span class="sd">    :param max_iter: Maximum number of iterations. Defaults to `200`.</span>
<span class="sd">    :type max_iter: int</span>
<span class="sd">    :param shuffle: Whether to shuffle samples in each iteration. Only when solver is `sgd` </span>
<span class="sd">        or `adam`. Defaults to `True`.</span>
<span class="sd">    :type shuffle: bool</span>
<span class="sd">    :param random_state: Determines random number generation for weights and bias initialization,</span>
<span class="sd">        train-test split and batch sampling. Defaults to None.</span>
<span class="sd">    :type random_state: int, RandomInstance, None</span>
<span class="sd">    :param tol:Tolerance for stopping criteria. Defaults to `1e-4`.</span>
<span class="sd">    :type tol: float</span>
<span class="sd">    :param verbose: For the liblinear and lbfgs solvers set verbose to any positive</span>
<span class="sd">            number for verbosity.  Defaults to `False`.</span>
<span class="sd">    :type verbose: bool</span>
<span class="sd">    :param warm_start: If `True`, reuse the solution of the previous call to fit</span>
<span class="sd">        as initialization, otherwise, just erase the previous solution. Defauls to `False`.</span>
<span class="sd">    :type warm_start: bool</span>
<span class="sd">    :param momentum: Momentum for gradient descent update. Should be between 0 and 1. </span>
<span class="sd">        Defaults to `0.9`.</span>
<span class="sd">    :type momentum: float</span>
<span class="sd">    :param nesterovs_momentum: Wheter to use Nesterov&#39;s momentum. Only when solves is `sgd`</span>
<span class="sd">        and momentum &gt; 0. Defaults to `True`.</span>
<span class="sd">    :type nesterovs_momentum: bool</span>
<span class="sd">    :param early_stopping: Whether to use early stopping to terminate training when </span>
<span class="sd">        validation score is not improving. Defaults to `False`.</span>
<span class="sd">    :type early_stopping: bool</span>
<span class="sd">    :param validation_fraction: The proportion of training data to set aside as validation </span>
<span class="sd">        set fot early stopping. Must be between 0 and 1. Only if early stopping is `True`.</span>
<span class="sd">        Defaults to `0.1`.</span>
<span class="sd">    :type validation_fraction: float</span>
<span class="sd">    :param beta_1: Exponential decay rate for estimates of first moment vector in adam, </span>
<span class="sd">        should be between 0 and 1. Only used when solver is `adam`. Defaults to `0.9`.</span>
<span class="sd">    :type beta_1: float</span>
<span class="sd">    :param beta_2: Exponential decay rate for estimates of second moment vector in adam, </span>
<span class="sd">        should be between 0 and 1. Only used when solver is `adam`. Defaults to `0.999`.</span>
<span class="sd">    :type beta_2: float</span>
<span class="sd">    :param epsilon: Value for numerical stability in adam. Only used when solver is `adam`. </span>
<span class="sd">        Defaults to `1e-8`.</span>
<span class="sd">    :type epsilon: float</span>
<span class="sd">    :param n_iter_no_change: Maximum number of epoch to not meet tol improvement. </span>
<span class="sd">        Only when solver is `sgd` or `adam`. Defaults to `10`.</span>
<span class="sd">    :type n_iter_no_change: int</span>
<span class="sd">    :param max_fun: Maximum number of loss functions calls. Only when solver is `lbfgs`. </span>
<span class="sd">        Defaults to `15000`.</span>
<span class="sd">    :type max_fun: int</span>

<span class="sd">    :return: Neural Network algorithm</span>
<span class="sd">    :rtype: NeuralNetwork</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">NeuralNetwork</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span><span class="p">,</span> <span class="n">hidden_layer_sizes</span> <span class="o">=</span> <span class="n">hidden_layer_sizes</span><span class="p">,</span> <span class="n">activation</span> <span class="o">=</span> <span class="n">activation</span><span class="p">,</span> <span class="n">solver</span> <span class="o">=</span> <span class="n">solver</span><span class="p">,</span>
               <span class="n">alpha</span> <span class="o">=</span> <span class="n">alpha</span><span class="p">,</span> <span class="n">batch_size</span> <span class="o">=</span> <span class="n">batch_size</span><span class="p">,</span> <span class="n">learning_rate</span> <span class="o">=</span> <span class="n">learning_rate</span><span class="p">,</span> <span class="n">learning_rate_init</span> <span class="o">=</span> <span class="n">learning_rate_init</span><span class="p">,</span> <span class="n">power_t</span> <span class="o">=</span> <span class="n">power_t</span><span class="p">,</span>
               <span class="n">max_iter</span> <span class="o">=</span> <span class="n">max_iter</span><span class="p">,</span> <span class="n">shuffle</span> <span class="o">=</span> <span class="n">shuffle</span><span class="p">,</span> <span class="n">random_state</span> <span class="o">=</span> <span class="n">random_state</span><span class="p">,</span> <span class="n">tol</span> <span class="o">=</span> <span class="n">tol</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="n">verbose</span><span class="p">,</span> <span class="n">warm_start</span> <span class="o">=</span> <span class="n">warm_start</span><span class="p">,</span>
               <span class="n">momentum</span> <span class="o">=</span> <span class="n">momentum</span><span class="p">,</span> <span class="n">nesterovs_momentum</span> <span class="o">=</span> <span class="n">nesterovs_momentum</span><span class="p">,</span> <span class="n">early_stopping</span> <span class="o">=</span> <span class="n">early_stopping</span><span class="p">,</span> <span class="n">validation_fraction</span> <span class="o">=</span> <span class="n">validation_fraction</span><span class="p">,</span> <span class="n">beta_1</span> <span class="o">=</span> <span class="n">beta_1</span><span class="p">,</span> 
               <span class="n">beta_2</span> <span class="o">=</span> <span class="n">beta_2</span><span class="p">,</span> <span class="n">epsilon</span> <span class="o">=</span> <span class="n">epsilon</span><span class="p">,</span> <span class="n">n_iter_no_change</span> <span class="o">=</span> <span class="n">n_iter_no_change</span><span class="p">,</span> <span class="n">max_fun</span> <span class="o">=</span> <span class="n">max_fun</span><span class="p">)</span></div>



<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">######################################################</span>

<span class="sd">                    PROCESSGDC</span>

<span class="sd">######################################################</span>
<span class="sd">&quot;&quot;&quot;</span>


<span class="c1"># =============================================================================</span>
<span class="c1">#                                   URL </span>
<span class="c1"># =============================================================================</span>


<div class="viewcode-block" id="gdc_endpt_base"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_endpt_base">[docs]</a><span class="k">def</span> <span class="nf">gdc_endpt_base</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Returns a string with the URL to search in the TCGA base</span>
<span class="sd">        </span>
<span class="sd">    :return: The URL to search in the GDC base.</span>
<span class="sd">    :rtype: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">get_endpt_base</span><span class="p">()</span></div>


<div class="viewcode-block" id="gdc_endpt_files"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_endpt_files">[docs]</a><span class="k">def</span> <span class="nf">gdc_endpt_files</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Returns a string with the URL to files endpoint</span>
<span class="sd">        </span>
<span class="sd">    :return: The URL to search in the GDC files.</span>
<span class="sd">    :rtype: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">get_endpt_files</span><span class="p">()</span></div>


<div class="viewcode-block" id="gdc_endpt_status"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_endpt_status">[docs]</a><span class="k">def</span> <span class="nf">gdc_endpt_status</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Returns a string with the URL to check status</span>
<span class="sd">        </span>
<span class="sd">    :return: The URL to check status</span>
<span class="sd">    :rtype: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">get_endpt_status</span><span class="p">()</span></div>


<div class="viewcode-block" id="gdc_endpt_data"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_endpt_data">[docs]</a><span class="k">def</span> <span class="nf">gdc_endpt_data</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Returns a string with the URL to files endpoint</span>
<span class="sd">        </span>
<span class="sd">    :return: The URL to search in the GDC files.</span>
<span class="sd">    :rtype: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">get_endpt_data</span><span class="p">()</span></div>


<div class="viewcode-block" id="gdc_endpt_projects"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_endpt_projects">[docs]</a><span class="k">def</span> <span class="nf">gdc_endpt_projects</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Returns a string with the URL to projects endpoint</span>
<span class="sd">        </span>
<span class="sd">    :return: The URL to search in the GDC projects.</span>
<span class="sd">    :rtype: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">get_endpt_projects</span><span class="p">()</span></div>


<div class="viewcode-block" id="gdc_endpt_legacy"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_endpt_legacy">[docs]</a><span class="k">def</span> <span class="nf">gdc_endpt_legacy</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Returns a string with the URL to search in the TCGA Legacy Archive</span>
<span class="sd">        </span>
<span class="sd">    :return: The URL to search in the GDC Legacy Archive.</span>
<span class="sd">    :rtype: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">get_endpt_legacy</span><span class="p">()</span></div>


<div class="viewcode-block" id="gdc_clinical_filename"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_clinical_filename">[docs]</a><span class="k">def</span> <span class="nf">gdc_clinical_filename</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Returns a string with the main name of a clinical filename of GDC</span>
<span class="sd">        </span>
<span class="sd">    :return: a string with the main name.</span>
<span class="sd">    :rtype: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">get_clinical_filename</span><span class="p">()</span></div>


<div class="viewcode-block" id="gdc_clinical_options"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_clinical_options">[docs]</a><span class="k">def</span> <span class="nf">gdc_clinical_options</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Returns a list of options for clinical information</span>
<span class="sd">        </span>
<span class="sd">    :return: list of options</span>
<span class="sd">    :rtype: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">get_clinical_options</span><span class="p">()</span></div>


<div class="viewcode-block" id="gdc_data_legacy"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_data_legacy">[docs]</a><span class="k">def</span> <span class="nf">gdc_data_legacy</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot; Returns a boolean that indicates if the data is from the GDC</span>
<span class="sd">        Legacy Archive or from the GDC Data Portal.</span>
<span class="sd">    </span>
<span class="sd">        :return: `True` to search for data in the GDC Legacy Archive,</span>
<span class="sd">            and `False` otherwise.</span>
<span class="sd">        :rtype: bool</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">get_data_legacy</span><span class="p">()</span></div>


<div class="viewcode-block" id="gdc_set_data_legacy"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_set_data_legacy">[docs]</a><span class="k">def</span> <span class="nf">gdc_set_data_legacy</span><span class="p">(</span><span class="n">legacy</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Set if data comes from the GDC Legacy Archive or not.</span>
<span class="sd">        </span>
<span class="sd">        :param legacy: `True` to search for data in the GDC Legacy Archive,</span>
<span class="sd">            and `False` otherwise.</span>
<span class="sd">        :type legacy: bool </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">get_enset_data_legacydpt_legacy</span><span class="p">(</span><span class="n">legacy</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                               PROJECTS </span>
<span class="c1"># =============================================================================</span>



<div class="viewcode-block" id="gdc_get_projects"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_get_projects">[docs]</a><span class="k">def</span> <span class="nf">gdc_get_projects</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Gets all available GDC projects</span>
<span class="sd">            </span>
<span class="sd">        :return: List of available projects in the GDC Data Portal.</span>
<span class="sd">        :rtype: list</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">get_projects</span><span class="p">()</span></div>


<div class="viewcode-block" id="gdc_search_project"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_search_project">[docs]</a><span class="k">def</span> <span class="nf">gdc_search_project</span><span class="p">(</span><span class="n">ident</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Search the project by the input &quot;ident&quot; variable. It could be the identifier</span>
<span class="sd">        or the complete name of the cancer. If exists, the ID will be returned.</span>
<span class="sd">        </span>
<span class="sd">        :param ident: Identificator (ID/Name) to search the project.</span>
<span class="sd">        :type ident: str</span>
<span class="sd">        :return: The identifier of the project. E.g: TCGA-BRCA.</span>
<span class="sd">        :rtype: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">search_project</span><span class="p">(</span><span class="n">ident</span><span class="p">)</span></div>


<div class="viewcode-block" id="gdc_check_parameter"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_check_parameter">[docs]</a><span class="k">def</span> <span class="nf">gdc_check_parameter</span><span class="p">(</span><span class="n">endpt</span><span class="p">,</span> <span class="n">param</span><span class="p">,</span> <span class="n">value</span><span class="p">,</span> <span class="n">filters</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Check if the parameter selected is available with the actual query.</span>
<span class="sd">            </span>
<span class="sd">        :param endpt: URL where the request is going to be done.</span>
<span class="sd">        :type endpt: str</span>
<span class="sd">        :param param: Name of the parameter to check.</span>
<span class="sd">        :type param: str</span>
<span class="sd">        :param value: Value of the parameter to check.</span>
<span class="sd">        :type value: str</span>
<span class="sd">        :param filters: current list of filters.</span>
<span class="sd">        :type filters: list</span>
<span class="sd">        :return: The value of parameter with the correct format to add </span>
<span class="sd">            it to the query.</span>
<span class="sd">        :rtype: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">check_parameter</span><span class="p">(</span><span class="n">endpt</span><span class="p">,</span> <span class="n">param</span><span class="p">,</span> <span class="n">value</span><span class="p">,</span> <span class="n">filters</span><span class="p">)</span></div>


<div class="viewcode-block" id="gdc_check_data_type"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_check_data_type">[docs]</a><span class="k">def</span> <span class="nf">gdc_check_data_type</span><span class="p">(</span><span class="n">endpt</span><span class="p">,</span> <span class="n">filters</span><span class="p">,</span> <span class="n">data_type</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Check the data type selected. We need to have a single data type to </span>
<span class="sd">        proceed with the download. So, data type cannot be a list.</span>
<span class="sd">        If None, we have to check if all the filters selected return </span>
<span class="sd">        only one data type to download.</span>
<span class="sd">        </span>
<span class="sd">        :param endpt: URL where the request is going to be done.</span>
<span class="sd">        :type endpt: str</span>
<span class="sd">        :param filters: current list of filters.</span>
<span class="sd">        :type filters: list</span>
<span class="sd">        :param data_type: Data Type to check.</span>
<span class="sd">        :type data_type: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">check_data_type</span><span class="p">(</span><span class="n">endpt</span><span class="p">,</span> <span class="n">filters</span><span class="p">,</span> <span class="n">data_type</span><span class="p">)</span></div>


<div class="viewcode-block" id="gdc_form_filter"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_form_filter">[docs]</a><span class="k">def</span> <span class="nf">gdc_form_filter</span><span class="p">(</span><span class="n">filters</span><span class="p">,</span> <span class="n">nameFilter</span><span class="p">,</span> <span class="n">valueFilter</span><span class="p">,</span> <span class="n">fullNameFilter</span><span class="p">,</span> <span class="n">endpt_</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Creates the filter to do the query that gets the data.</span>
<span class="sd">        If the new value for the query is available, it will be added to</span>
<span class="sd">        the filter&#39;s list.</span>
<span class="sd">        </span>
<span class="sd">        :param filters: Current list with the filters.</span>
<span class="sd">        :type filters: list</span>
<span class="sd">        :param nameFilter: Name of the filter to be checked and added.</span>
<span class="sd">        :type nameFilter: str</span>
<span class="sd">        :param valueFilter: Value of the filter to be checked and added.</span>
<span class="sd">        :type valueFilter: str</span>
<span class="sd">        :param fullNameFilter: Name of the filter to be added plus the path </span>
<span class="sd">            where it is searched in the API.</span>
<span class="sd">        :type fullNameFilter: str </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">form_filter</span><span class="p">(</span><span class="n">filters</span><span class="p">,</span> <span class="n">nameFilter</span><span class="p">,</span> <span class="n">valueFilter</span><span class="p">,</span> <span class="n">fullNameFilter</span><span class="p">,</span> <span class="n">endpt_</span><span class="p">)</span></div>


<div class="viewcode-block" id="gdc_regex_file_name"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_regex_file_name">[docs]</a><span class="k">def</span> <span class="nf">gdc_regex_file_name</span><span class="p">(</span><span class="n">proj_id</span><span class="p">,</span> <span class="n">clinical_option</span><span class="p">,</span> <span class="n">file_name</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Finds the clinical option even if it is a new version of the file.</span>
<span class="sd">        The version of the file is controlled by a regex expression. </span>
<span class="sd">        E.g:  _omf_v1.0 or _omf_v12.9 can be part of the file name</span>
<span class="sd">        </span>
<span class="sd">        :param proj_id: ID of the project. E.g: if the project is &quot;TCGA-SKCM&quot;, the id</span>
<span class="sd">            will be &quot;skcm&quot;.</span>
<span class="sd">        :type proj_id: str</span>
<span class="sd">        :param clinical_option: Clinical option to search in the regex expression.</span>
<span class="sd">        :type clinical_option: str</span>
<span class="sd">        :param file_name: File to check if it has the clinical option in its name, even</span>
<span class="sd">            if it has a version pattern.</span>
<span class="sd">        :type file_name: str</span>
<span class="sd">        :return: `True` if the clinical option has been found and `False` otherwise.</span>
<span class="sd">        :rtype: bool</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">regex_file_name</span><span class="p">(</span><span class="n">proj_id</span><span class="p">,</span> <span class="n">clinical_option</span><span class="p">,</span> <span class="n">file_name</span><span class="p">)</span></div>


<div class="viewcode-block" id="gdc_find_clinical"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_find_clinical">[docs]</a><span class="k">def</span> <span class="nf">gdc_find_clinical</span><span class="p">(</span><span class="n">file_list</span><span class="p">,</span> <span class="n">clinical_option</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Finds if the selected clinical option is one of the available list </span>
<span class="sd">        of files.</span>
<span class="sd">        First, we only kept with those files from the list of files that match</span>
<span class="sd">        a pattern where it is searched tyhe word &quot;clinical&quot; in the name of the file.</span>
<span class="sd">        Then, in a new list with only the clinical files, search for a file that have</span>
<span class="sd">        the specified clinical option.</span>
<span class="sd">        </span>
<span class="sd">        :param file_list: List of files to search the clinical option&#39;s file.</span>
<span class="sd">        :type file_list: list</span>
<span class="sd">        :return: `True` if the clinical option has been found and `False` otherwise.</span>
<span class="sd">        :rtype: bool</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">find_clinical</span><span class="p">(</span><span class="n">file_list</span><span class="p">,</span> <span class="n">clinical_option</span><span class="p">)</span></div>


<div class="viewcode-block" id="gdc_write_to_dataframe"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_write_to_dataframe">[docs]</a><span class="k">def</span> <span class="nf">gdc_write_to_dataframe</span><span class="p">(</span><span class="n">list_</span><span class="p">,</span> <span class="n">file</span><span class="p">,</span> <span class="n">col</span><span class="p">,</span> <span class="n">entity_id</span><span class="p">,</span> <span class="n">legacy</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Creates a list of the DataFrame&#39;s lines, so it allows multithreading</span>
<span class="sd">        while reading the file.</span>
<span class="sd">        </span>
<span class="sd">        :param list_: List to append DataFrame&#39;s content. It is modified by reference.</span>
<span class="sd">        :type list_: list</span>
<span class="sd">        :param file: file to be read.</span>
<span class="sd">        :type file: str</span>
<span class="sd">        :param col: Name of the column to be read. If legacy is `True`, </span>
<span class="sd">            it is set to the second column of the DataFrame; if legacy is</span>
<span class="sd">            `False`, column is set to `unstranded`.</span>
<span class="sd">        :type col: str</span>
<span class="sd">        :param legacy:`True` to search for data in the GDC Legacy Archive,</span>
<span class="sd">            and `False` otherwise; defaults to `False`.</span>
<span class="sd">        :type legacy: bool </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">write_to_dataframe</span><span class="p">(</span><span class="n">list_</span><span class="p">,</span> <span class="n">file</span><span class="p">,</span> <span class="n">col</span><span class="p">,</span> <span class="n">entity_id</span><span class="p">,</span> <span class="n">legacy</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                               PROJECTS </span>
<span class="c1"># =============================================================================</span>



<div class="viewcode-block" id="gdc_query"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_query">[docs]</a><span class="k">def</span> <span class="nf">gdc_query</span><span class="p">(</span><span class="n">query_object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; It creates the correct query using the data specified in the input DataObject</span>
<span class="sd">        and gets the data that it is going to be downloaded.</span>
<span class="sd">        To do it, it is checked each parameter availability taking into account the filter</span>
<span class="sd">        selected.</span>
<span class="sd">        If the parameter is available, it is added to the filter&#39;s list before checking</span>
<span class="sd">        the next parameter. </span>
<span class="sd">        Once all filters are set, we get the request.</span>
<span class="sd">        </span>
<span class="sd">        :param query_object: GDCQuery that specified all the fields to get the data desired.</span>
<span class="sd">        :type query_object: GDCQuery</span>
<span class="sd">        :return: The response of the request, in case all parameters are available.</span>
<span class="sd">        :rtype: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">get_query</span><span class="p">(</span><span class="n">query_object</span><span class="p">)</span></div>


<div class="viewcode-block" id="gdc_clinical_query"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_clinical_query">[docs]</a><span class="k">def</span> <span class="nf">gdc_clinical_query</span><span class="p">(</span><span class="n">query_object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Query in the TCGA API to get the clinical info linked to a </span>
<span class="sd">        specific project.</span>
<span class="sd">        </span>
<span class="sd">        :param project_name: Name of the project.</span>
<span class="sd">        :type project_name: str</span>
<span class="sd">        :param legacy:`True` to search for data in the GDC Legacy Archive,</span>
<span class="sd">            and `False` otherwise; defaults to `False`.</span>
<span class="sd">        :type legacy: bool </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">get_clinical_query</span><span class="p">(</span><span class="n">query_object</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                               PROJECTS </span>
<span class="c1"># =============================================================================</span>



<div class="viewcode-block" id="gdc_download_data"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_download_data">[docs]</a><span class="k">def</span> <span class="nf">gdc_download_data</span><span class="p">(</span><span class="n">query_object</span><span class="p">,</span> <span class="n">os_path</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Download the data with the characteristics specidied in the DataObject object.</span>
<span class="sd">        Before downloading the data, it is checked if all fields have a correct value and</span>
<span class="sd">        existing information.</span>
<span class="sd">        A folder with the ID of the cancer will be created to save the data. It will also be</span>
<span class="sd">        saved a manifest.txt file with the list of the ID of the files and their Entity ID.</span>
<span class="sd">        The Entity ID is useful to match samples with clinical information downloaded from the </span>
<span class="sd">        portal.</span>
<span class="sd">        All the data is downloaded using the TCGA API via requests.</span>
<span class="sd">        </span>
<span class="sd">        :param query_object: GDCQuery object with the features of the data that will be downloaded.</span>
<span class="sd">        :type query_object: GDCQuery</span>
<span class="sd">        :param os_path: Directory path were the file will be saved.Defaults to None, so</span>
<span class="sd">            it will be saved in the current directory.</span>
<span class="sd">        :type os_path: str, optional</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">download_data</span><span class="p">(</span><span class="n">query_object</span><span class="p">,</span> <span class="n">os_path</span><span class="p">)</span></div>


<div class="viewcode-block" id="gdc_download_clinical_data"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_download_clinical_data">[docs]</a><span class="k">def</span> <span class="nf">gdc_download_clinical_data</span><span class="p">(</span><span class="n">project_name</span><span class="p">,</span> <span class="n">clinic_info</span><span class="p">,</span> <span class="n">legacy</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">os_path</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Download clinical data from a specified project. A folder with the name of the</span>
<span class="sd">        project will be created if it does not already exist. If it exists, it will be</span>
<span class="sd">        checked if the clinical information has been previously downloaded.</span>
<span class="sd">        </span>
<span class="sd">        :param project_name: Name of the project.</span>
<span class="sd">        :type project_name: str</span>
<span class="sd">        :param clinic_info: Type of clinical information that will be downloaded. </span>
<span class="sd">            It will be downloaded the latest version found. E.g: &quot;patient&quot; or &quot;drug&quot;.</span>
<span class="sd">        :type clinic_info: str</span>
<span class="sd">        :param legacy:`True` to search for data in the GDC Legacy Archive,</span>
<span class="sd">            and `False` otherwise; defaults to `False`.</span>
<span class="sd">        :type legacy: bool</span>
<span class="sd">        :param os_path: Directory path were the file will be saved. Defaults to None, so</span>
<span class="sd">            it will be saved in the current directory.</span>
<span class="sd">        :type os_path: str, optional</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">download_clinical_data</span><span class="p">(</span><span class="n">project_name</span><span class="p">,</span> <span class="n">clinic_info</span><span class="p">,</span> <span class="n">legacy</span> <span class="o">=</span> <span class="n">legacy</span><span class="p">,</span> <span class="n">os_path</span> <span class="o">=</span> <span class="n">os_path</span><span class="p">)</span></div>


<div class="viewcode-block" id="gdc_download_genecode"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_download_genecode">[docs]</a><span class="k">def</span> <span class="nf">gdc_download_genecode</span><span class="p">(</span><span class="n">url_gtf</span> <span class="o">=</span> <span class="s2">&quot;https://api.gdc.cancer.gov/data/be002a2c-3b27-43f3-9e0f-fd47db92a6b5&quot;</span><span class="p">,</span> <span class="n">name_gtf</span> <span class="o">=</span> <span class="s2">&quot;genecode_v36&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Download genecode to create a data structure with information about</span>
<span class="sd">        specific genes. It will allow building normalization functions.</span>

<span class="sd">        :param url_gtf: URL where the genecode is hosted. By default, the program download from GDC and its 36th version.</span>
<span class="sd">        :type url_gtf: str</span>
<span class="sd">        :param name_gft: Name of the file (WITHOUT extension). Its extension must be .gtf</span>
<span class="sd">        :param name_gft: str</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">download_genecode</span><span class="p">(</span><span class="n">url_gtf</span><span class="p">,</span> <span class="n">name_gtf</span><span class="p">)</span></div>



<div class="viewcode-block" id="gdc_create_gene_model"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_create_gene_model">[docs]</a><span class="k">def</span> <span class="nf">gdc_create_gene_model</span><span class="p">(</span><span class="n">path_gtf</span> <span class="o">=</span> <span class="s2">&quot;genecode_v36.gtf&quot;</span><span class="p">,</span> <span class="n">path_save</span> <span class="o">=</span> <span class="s2">&quot;genecode_v36.csv&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Create a DataFrame that allow an easy treatment with information about Genes.</span>
<span class="sd">        The function will create a CSV readable with features of Genes. If CSV exist, there is</span>
<span class="sd">        not neccesary to call this function, only read it.</span>

<span class="sd">        :param path_gtf: Path where genecode with GTF format is located</span>
<span class="sd">        :type path_gtf: str</span>
<span class="sd">        :param path_save: Path where the CSV result will be save</span>
<span class="sd">        :param path_save: str</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">create_gene_model</span><span class="p">(</span><span class="n">path_gtf</span><span class="p">,</span> <span class="n">path_save</span><span class="p">)</span></div>



<div class="viewcode-block" id="read_genecode_csv"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.read_genecode_csv">[docs]</a><span class="k">def</span> <span class="nf">read_genecode_csv</span><span class="p">(</span><span class="n">file_csv</span><span class="o">=</span><span class="s2">&quot;../genecode_v36.csv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Read a CSV file that contains information about genes</span>
<span class="sd">        and returns a DataFrame to work with it</span>

<span class="sd">        :param file_csv: Path where the genecode on CSV format is saved</span>
<span class="sd">        :type file_csv: str</span>
<span class="sd">        :param sep: Character that delimit its structure</span>
<span class="sd">        :type sep: str</span>

<span class="sd">        :return: A DataFrame with Gene Information</span>
<span class="sd">        :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">read_genecode_csv</span><span class="p">(</span><span class="n">file_csv</span><span class="p">,</span> <span class="n">sep</span><span class="p">)</span></div>


<span class="c1"># =============================================================================</span>
<span class="c1">#                           READ DATA FUNCTIONS </span>
<span class="c1"># =============================================================================</span>



<div class="viewcode-block" id="gdc_read_file_rna_legacy"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_read_file_rna_legacy">[docs]</a><span class="k">def</span> <span class="nf">gdc_read_file_rna_legacy</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">col</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Read a single Rna-Seq file from The TCGA Legacy Archive.</span>
<span class="sd">        The first line is the column&#39;s names.</span>
<span class="sd">        An undefined number of rows are skipped while their first character</span>
<span class="sd">        is a `?`.</span>
<span class="sd">        The index is set to the first column, which is the gene_id.</span>
<span class="sd">        In this type of files the gene_id are made up of the ID of the gene and a number,</span>
<span class="sd">        separated by a &quot;|&quot;. Only the ID of the gene is saved.</span>
<span class="sd">        </span>
<span class="sd">        :param file_name: File name to be read.</span>
<span class="sd">        :type dir_path: str</span>
<span class="sd">        :param col: Name of the column to be read. Set to `unstranded`.</span>
<span class="sd">        :type col: str</span>
<span class="sd">        :param sep: Separator character between fields. Defaults to &quot;\t&quot; (Tab-separated values).</span>
<span class="sd">        :type sep: str</span>
<span class="sd">        :return: DataFrame with file&#39;s data read.</span>
<span class="sd">        :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">gdc_download_clinical_data</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">col</span><span class="p">,</span> <span class="n">sep</span><span class="p">)</span></div>


<div class="viewcode-block" id="gdc_read_file_rna"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_read_file_rna">[docs]</a><span class="k">def</span> <span class="nf">gdc_read_file_rna</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">col</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Read a single Rna-Seq file from The GDC portal. </span>
<span class="sd">        The first line of this type of files is a comment that it is skipped.</span>
<span class="sd">        The second line will be the header of the data.</span>
<span class="sd">        Next lines are skipped / saved in a summary_file because they store </span>
<span class="sd">        summary information of the file.</span>
<span class="sd">        The index is set to the first column, which is the gene_id.</span>
<span class="sd">        </span>
<span class="sd">        :param file_name: File name to be read.</span>
<span class="sd">        :type dir_path: str</span>
<span class="sd">        :param col: Name of the column to be read. Set to `unstranded`.</span>
<span class="sd">        :type col: str</span>
<span class="sd">        :param sep: Separator character between fields. Defaults to &quot;\t&quot; (Tab-separated values).</span>
<span class="sd">        :type sep: str</span>
<span class="sd">        :return: DataFrame with file&#39;s data read.</span>
<span class="sd">        :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">read_file_rna</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">col</span><span class="p">,</span> <span class="n">sep</span><span class="p">)</span></div>



<div class="viewcode-block" id="gdc_read_rna"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_read_rna">[docs]</a><span class="k">def</span> <span class="nf">gdc_read_rna</span><span class="p">(</span><span class="n">dir_path</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Read Rna-Seq files from The Cancer Genome Atlas (TCGA) portal.</span>
<span class="sd">        The file will be read in a different way depending on whether it is from </span>
<span class="sd">        the Legacy Data portal or from the GDC Data portal.</span>
<span class="sd">        For the GDC portal, unstranded data is downloaded by default.</span>
<span class="sd">        For the Legacy Archive, it will be read the second column of the file</span>
<span class="sd">        that can be raw counts or normalized counts. While downloading the data, </span>
<span class="sd">        the user can indicate the type of data to dowload by the parameter &quot;normalized&quot;.</span>
<span class="sd">        </span>
<span class="sd">        :param dir_path: The path of the folder where the project has been downloaded. </span>
<span class="sd">            By feault, the program search the folder in the current directory, so if </span>
<span class="sd">            the project has been saved in a different folder, the user must indicate</span>
<span class="sd">            the full path.</span>
<span class="sd">        :type dir_path: str</span>
<span class="sd">        :param sep: Separator character between fields. Defaults to &quot;\t&quot; (Tab-separated values).</span>
<span class="sd">        :type sep: str</span>
<span class="sd">        :param save: `True` to keep the downloaded data in the folder and `False` to remove</span>
<span class="sd">        the downloaded data after read it; defaults to `True`.</span>
<span class="sd">        :type save: bool</span>
<span class="sd">        :return: DataFrame with the RNA-Seq information read.</span>
<span class="sd">        :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">read_rna</span><span class="p">(</span><span class="n">dir_path</span><span class="p">,</span> <span class="n">sep</span><span class="p">,</span> <span class="n">save</span><span class="p">)</span></div>


<div class="viewcode-block" id="gdc_read_clinical"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.gdc_read_clinical">[docs]</a><span class="k">def</span> <span class="nf">gdc_read_clinical</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Read a clinical file preivously downaloaded from The Cancer Genome Atlas (TCGA) portal.</span>
<span class="sd">        In TCGA clinical files, the first line is a comment that it is skipped. The two following lines</span>
<span class="sd">        have the header names; only the second header&#39;s line is kept, due to they have the same values</span>
<span class="sd">        but with different names.</span>
<span class="sd">        Returns a dataframe with the clinical information where the index is set to `bcr_patient_barcode`,</span>
<span class="sd">        that match with the Entity ID of the samples.</span>
<span class="sd">        </span>
<span class="sd">        :param file_name: The name of the clinical project to read. By default, the programn search the file</span>
<span class="sd">            in the actual path, so the location of the file is required to find the file. </span>
<span class="sd">            E.g: &quot;TCGA-BRCA/nationwidechildrens.org_clinical_patient_brca.txt&quot;.</span>
<span class="sd">        :type file_name: str</span>
<span class="sd">        :param sep: Separator character between fields. Defaults to &quot;\t&quot; (Tab-separated values).</span>
<span class="sd">        :type sep: str</span>
<span class="sd">        :return: DataFrame with the clinical information read. The index has been set to</span>
<span class="sd">            `bcr_patient_barcode`, matching with the Entity ID of the samples.</span>
<span class="sd">        :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">ProcessGDC</span><span class="o">.</span><span class="n">read_clinical</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">sep</span><span class="p">)</span></div>



<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">######################################################</span>

<span class="sd">                        OBJECTS</span>

<span class="sd">######################################################</span>
<span class="sd">&quot;&quot;&quot;</span>


<span class="c1"># =============================================================================</span>
<span class="c1">#                               TCGAQUERY </span>
<span class="c1"># =============================================================================</span>


<div class="viewcode-block" id="create_gdc_query"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.create_gdc_query">[docs]</a><span class="k">def</span> <span class="nf">create_gdc_query</span><span class="p">(</span><span class="n">project</span> <span class="p">,</span><span class="n">experimental_strategy</span><span class="p">,</span> <span class="n">data_category</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">data_type</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
                 <span class="n">workflow_type</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">legacy</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>  <span class="n">platform</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
                 <span class="n">data_format</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">sample_type</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">normalized</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Generate a object to do arequest on Genomic Data Commoms to download data</span>
<span class="sd">    </span>
<span class="sd">    :param project: Cancer name or ID to download. E.g: &quot;TCGA-BRCA&quot;, </span>
<span class="sd">        &quot;Acute Myeloid Leukemia&quot;, &quot;adrenocortical carcinoma&quot; ...</span>
<span class="sd">    :type project: str </span>
<span class="sd">    :param experimental_strategy: Experimental strategies used for molecular</span>
<span class="sd">        characterization of the cancer.</span>
<span class="sd">    :type experimental_strategy: str</span>
<span class="sd">    :param data_category: A high-level data file category. E.g &quot;Raw Sequencing</span>
<span class="sd">        Data&quot; or &quot;Transcriptome Profiling&quot;, dafults to None</span>
<span class="sd">    :type data_category: str, optional</span>
<span class="sd">    :param data_type: Data file type. Data Type is more granular than Data Category.</span>
<span class="sd">        E.g: &quot;Align Reads&quot; or &quot;Gene Expression Quantification&quot;, defaults to None.</span>
<span class="sd">    :type data_type: str, optional</span>
<span class="sd">    :param workflow_type: Bioinformatics workflow used to generate or harmonize</span>
<span class="sd">        the data file. E.g: &quot;STAR - Counts&quot;, &quot;BCGSC miRNA Profiling&quot;, defaults to None.</span>
<span class="sd">    :type workflow_type: str, optional</span>
<span class="sd">    :param legacy: The GCD Legacy Archive hosts unharmonized legacy dara from repositories</span>
<span class="sd">        that predate the GDC (e.g: CGHub). Legacy data is not actively maintained, processed, or</span>
<span class="sd">        harmonized by the GDC. Legacy is `True` to search for data in the GDC Legacy Archive,</span>
<span class="sd">        and `False` otherwise; defaults to `False`.</span>
<span class="sd">    :type legacy: bool</span>
<span class="sd">    :param platform: Technolofical platform on which experimental data was produced.</span>
<span class="sd">        E.g: &quot;Illumina Human Methylation 450&quot;, defaults to None. </span>
<span class="sd">    :type platform: str, optional</span>
<span class="sd">    :param data_format: Format of the Data File, defaults to None.</span>
<span class="sd">    :type data_format: str, optional</span>
<span class="sd">    :param sample_type: Describes the source of a biospecimen used for a laboratory test. </span>
<span class="sd">        E.g: &quot;Primary Tumor&quot;, [&quot;Primary Tumor&quot;, &quot;Solid Tissue Normal&quot;], defaults to None.</span>
<span class="sd">    :type sample_type: list, str, optional</span>
<span class="sd">    :param normalized: This parameter is only valid when legacy is set as `True`.</span>
<span class="sd">        If `True`, normalized data from the GDC Legacy Archive will be downloaded, </span>
<span class="sd">        and `False` otherwise; defaults to `False`.</span>
<span class="sd">    :type normalized: bool</span>

<span class="sd">    :return: a object with all neccesary information to download the project</span>
<span class="sd">    :rtype: GDCQuery</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">GDCQuery</span><span class="p">(</span><span class="n">project</span> <span class="p">,</span><span class="n">experimental_strategy</span><span class="p">,</span> <span class="n">data_category</span><span class="p">,</span> <span class="n">data_type</span><span class="p">,</span> 
                 <span class="n">workflow_type</span><span class="p">,</span> <span class="n">legacy</span><span class="p">,</span>  <span class="n">platform</span><span class="p">,</span> 
                 <span class="n">data_format</span><span class="p">,</span> <span class="n">sample_type</span><span class="p">,</span> <span class="n">normalized</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                               DATAPROJECT </span>
<span class="c1"># =============================================================================</span>



<div class="viewcode-block" id="create_data_project"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.create_data_project">[docs]</a><span class="k">def</span> <span class="nf">create_data_project</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Generate a object that allow to prepare a download for a project.</span>
<span class="sd">    Essential information such as the project&#39;s name, the path where it has been downloaded,</span>
<span class="sd">    the list of related files and the total size of the project is saved.</span>
<span class="sd">    </span>
<span class="sd">    :return: a object with all information save on its attributes</span>
<span class="sd">    :rtype: DataProject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">DataProject</span><span class="p">()</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                               FILEPROJECT </span>
<span class="c1"># =============================================================================</span>



<div class="viewcode-block" id="create_file_project"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.create_file_project">[docs]</a><span class="k">def</span> <span class="nf">create_file_project</span><span class="p">(</span><span class="n">id_</span><span class="p">,</span> <span class="n">data_format</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">access</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">file_name</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">submitter_id</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
               <span class="n">data_category</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">type_</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">file_size</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">created_datetime</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">md5sum</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
               <span class="n">updated_datetime</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">data_type</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">state</span><span class="o">=</span> <span class="kc">None</span><span class="p">,</span>  <span class="n">experimental_strategy</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">version</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
               <span class="n">data_release</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">entity_id</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Creates a object with the information and features of a downloaded file. </span>
<span class="sd">        Only the id_ will be a mandatory field due to the differences</span>
<span class="sd">        in the information that different data sources can provide.</span>
<span class="sd">    </span>
<span class="sd">    :param id_: UUID of the file. E.g: 48d05650-971e-4407-9cf1-7e9d71b1ff20</span>
<span class="sd">    :type id_: str  </span>
<span class="sd">    :param data_format: Format of the Data File, defaults to None.</span>
<span class="sd">    :type data_format: str, optional</span>
<span class="sd">    :param access: Indicator of whether acces to the data file is open or controlled,</span>
<span class="sd">        defaults to None.</span>
<span class="sd">    :type access: str, optional</span>
<span class="sd">    :param file_name: Name of the file. </span>
<span class="sd">        E.g: BLAIN_p_TCGA_282_304_b2_N_GenomeWideSNP_6_G08_1348614.grch38.seg.v2.txt, defaults to None.</span>
<span class="sd">    :type file_name: str, optional</span>
<span class="sd">    :param submitter_id: A unique key to identify the case. Submitter ID is different from the universally</span>
<span class="sd">        unique identifier (UUID). E.g: 7f4e8b00-d73f-479e-a8f7-30a8d7aee093_dr12_allcnv, defaults to None.</span>
<span class="sd">    :type submitter_id: str, optional  </span>
<span class="sd">    :param data_category: A high-level data file category. E.g &quot;Raw Sequencing</span>
<span class="sd">        Data&quot; or &quot;Transcriptome Profiling&quot;, dafults to None</span>
<span class="sd">    :type data_category: str, optional   </span>
<span class="sd">    :param type_: Type of the File, defaults to None.</span>
<span class="sd">    :type type_: str, optional   </span>
<span class="sd">    :param file_size: Size of the file in MB, defaults to None.</span>
<span class="sd">    :type file_size: int, optional</span>
<span class="sd">    :param created_datetime: Datetime of when the file has been created.</span>
<span class="sd">    :type created_datetime: datetime, optional</span>
<span class="sd">    :param md5sum: hash code. MD5 Checksum is used to verify the integrity of files,</span>
<span class="sd">        as virtually any change to a file will cause its MD5 hash to change.</span>
<span class="sd">    :type md5sum: str, optional</span>
<span class="sd">    :param updated_datetime: Datetatime of when a file had its latest update.</span>
<span class="sd">    :type updated_datetime: datetime, optional</span>
<span class="sd">    :param data_type: Data file type. Data Type is more granular than Data Category.</span>
<span class="sd">        E.g: &quot;Align Reads&quot; or &quot;Gene Expression Quantification&quot;, defaults to None.</span>
<span class="sd">    :type data_type: str, optional</span>
<span class="sd">    :param state: Current state of the file. E.g: &quot;releassed&quot;, defaults to None.</span>
<span class="sd">    :type state: str, optional</span>
<span class="sd">    :param experimental_strategy: Experimental strategies used for molecular</span>
<span class="sd">        characterization of the cancer.</span>
<span class="sd">    :type experimental_strategy: str, optional</span>
<span class="sd">    :param version: Number of the current version of the file.</span>
<span class="sd">    :type version: int, float, optional</span>
<span class="sd">    :param data_release: Number of the data release.</span>
<span class="sd">    :type data_release: float, optional</span>
<span class="sd">    :param entity_id: Also defined as Barcode. Entity ID is the primary identifier of</span>
<span class="sd">        biospecimen data within the TCGA project. E.g: TCGA-OR-A5JL-01A-11D-A29H-01, defaults to None.</span>
<span class="sd">    :type entity_id: str, optional</span>

<span class="sd">    :return: a object with information of a specific File</span>
<span class="sd">    :rtype: FileProject</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">FileProject</span><span class="p">(</span><span class="n">id_</span><span class="p">,</span> <span class="n">data_format</span><span class="p">,</span> <span class="n">access</span><span class="p">,</span> <span class="n">file_name</span><span class="p">,</span> <span class="n">submitter_id</span><span class="p">,</span>
               <span class="n">data_category</span><span class="p">,</span> <span class="n">type_</span><span class="p">,</span> <span class="n">file_size</span><span class="p">,</span> <span class="n">created_datetime</span><span class="p">,</span> <span class="n">md5sum</span><span class="p">,</span>
               <span class="n">updated_datetime</span><span class="p">,</span> <span class="n">data_type</span><span class="p">,</span> <span class="n">state</span><span class="p">,</span>  <span class="n">experimental_strategy</span><span class="p">,</span> <span class="n">version</span><span class="p">,</span>
               <span class="n">data_release</span><span class="p">,</span> <span class="n">entity_id</span><span class="p">)</span></div>



<span class="c1"># =============================================================================</span>
<span class="c1">#                                WORKFLOW </span>
<span class="c1"># =============================================================================</span>



<div class="viewcode-block" id="create_workflow"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.create_workflow">[docs]</a><span class="k">def</span> <span class="nf">create_workflow</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Creates a object workflow that allows to save several Log of a DataObject</span>
<span class="sd">        The methods executed over the data can be saved and applied over bulk data </span>
<span class="sd">        to replicate/expand the same analysis in the future.</span>
<span class="sd">    </span>
<span class="sd">    :return: an empty object prepare to save a Workflow of DataObject</span>
<span class="sd">    :rtype: Workflow</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">Workflow</span><span class="p">()</span></div>



<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">######################################################</span>

<span class="sd">                    SERVER STATUS</span>

<span class="sd">######################################################</span>
<span class="sd">&quot;&quot;&quot;</span>

<div class="viewcode-block" id="server_status_code"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.server_status_code">[docs]</a><span class="k">def</span> <span class="nf">server_status_code</span><span class="p">(</span><span class="n">stat_code</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the status response from the server</span>
<span class="sd">                </span>
<span class="sd">        :param stat_code: Current status of the checked URL endpoint.</span>
<span class="sd">        :type stat_code: int</span>
<span class="sd">        </span>
<span class="sd">        :return: Message explaining the obtained server&#39;s code.</span>
<span class="sd">        :rtype: str</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">ut</span><span class="o">.</span><span class="n">status_code</span><span class="p">(</span><span class="n">stat_code</span><span class="p">)</span></div>


<div class="viewcode-block" id="server_check_status"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.server_check_status">[docs]</a><span class="k">def</span> <span class="nf">server_check_status</span><span class="p">(</span><span class="n">endpt_status</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Check the status of the server.</span>
<span class="sd">            </span>
<span class="sd">        :param endpt_status: URL (endpoint) to be checked.</span>
<span class="sd">        :type endpt_status: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="k">return</span> <span class="n">ut</span><span class="o">.</span><span class="n">check_server_status</span><span class="p">(</span><span class="n">endpt_status</span><span class="p">)</span></div>


<div class="viewcode-block" id="get_json_query"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.get_json_query">[docs]</a><span class="k">def</span> <span class="nf">get_json_query</span><span class="p">(</span><span class="n">url_query</span><span class="p">,</span> <span class="n">params_query</span> <span class="o">=</span> <span class="p">{}):</span>
    <span class="sd">&quot;&quot;&quot;Get a structured JSON just by passing a URL, so it can be used by all the models.</span>
<span class="sd">        Each model will processes it as it is structured in its API.</span>
<span class="sd">        Params are empty by default, to get the data with or without parameters</span>
<span class="sd">        </span>
<span class="sd">        :param url_query: URL to do the request.</span>
<span class="sd">        :type url_query: str</span>
<span class="sd">        :param params_query: params to filter the query.</span>
<span class="sd">        :type params_query: dict</span>

<span class="sd">        :return: data obtained from the request.</span>
<span class="sd">        :rtype: json</span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="k">return</span> <span class="n">ut</span><span class="o">.</span><span class="n">get_json_query</span><span class="p">(</span><span class="n">url_query</span><span class="p">,</span> <span class="n">params_query</span><span class="p">)</span></div>


<div class="viewcode-block" id="create_opin"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.create_opin">[docs]</a><span class="k">def</span> <span class="nf">create_opin</span><span class="p">(</span><span class="n">filt</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Create a single in-operator in json format for the query, adding a field and its value.</span>

<span class="sd">        :param filt: values of a filter.</span>
<span class="sd">        :type filt: list[str]</span>
<span class="sd">        </span>
<span class="sd">        :return: the structured in-operator</span>
<span class="sd">        :rtype: dict</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">ut</span><span class="o">.</span><span class="n">create_opin</span><span class="p">(</span><span class="n">filt</span><span class="p">)</span></div>


<div class="viewcode-block" id="find_parameter"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.find_parameter">[docs]</a><span class="k">def</span> <span class="nf">find_parameter</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="n">param_results</span><span class="p">,</span> <span class="n">param</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Search in the result&#39;s list the desired value.</span>
<span class="sd">        If it is not found, the method shows all available parameters.</span>
<span class="sd">        If missing is returned, that option will be removed from the result&#39;s list</span>
<span class="sd">        because it is an undesired value.</span>

<span class="sd">        :param value: desired value to find on a parameter.</span>
<span class="sd">        :type value: str</span>
<span class="sd">        :param param_results: available parameters.</span>
<span class="sd">        :type param_results: list[str]</span>
<span class="sd">        :param param: param where to find.</span>
<span class="sd">        :type param: str</span>
<span class="sd">        </span>
<span class="sd">        :return: param found.</span>
<span class="sd">        :rtype: str</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">ut</span><span class="o">.</span><span class="n">find_parameter</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="n">param_results</span><span class="p">,</span> <span class="n">param</span><span class="p">)</span></div>


<div class="viewcode-block" id="print_request"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.print_request">[docs]</a><span class="k">def</span> <span class="nf">print_request</span><span class="p">(</span><span class="n">json_obj</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Decodes a JSON object and display it in an elegant format per screen,</span>

<span class="sd">        :param json_obj: JSON object to print.</span>
<span class="sd">        :type json_obj: JSON Object</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">ut</span><span class="o">.</span><span class="n">print_request</span><span class="p">(</span><span class="n">json_obj</span><span class="p">)</span></div>



<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">######################################################</span>

<span class="sd">                    UTILITIES</span>

<span class="sd">######################################################</span>
<span class="sd">&quot;&quot;&quot;</span>


<div class="viewcode-block" id="check_have_header"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.check_have_header">[docs]</a><span class="k">def</span> <span class="nf">check_have_header</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Check if the file has or not header.</span>
<span class="sd">            </span>
<span class="sd">        :param filename: Name of the file to check if has header or not.</span>
<span class="sd">        :type filename: str</span>

<span class="sd">        :return: `True` if file has header, `False` if otherwise.</span>
<span class="sd">        :rtype: bool</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">ut</span><span class="o">.</span><span class="n">check_have_header</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span></div>


<div class="viewcode-block" id="read_file"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.read_file">[docs]</a><span class="k">def</span> <span class="nf">read_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns a DataFrame with the read file.</span>

<span class="sd">        :param file: File to read</span>
<span class="sd">        :type file: str</span>
<span class="sd">        :param index: Value of the index of the DataFrame.</span>
<span class="sd">        :type index: str</span>
<span class="sd">        :param sep: Separator character between fields. Defaults to &quot;\t&quot; (Tab-separated values).</span>
<span class="sd">        :type sep: str</span>
<span class="sd">        </span>
<span class="sd">        :return: DataFrame with the file read.</span>
<span class="sd">        :rtype: DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">ut</span><span class="o">.</span><span class="n">read_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">index</span><span class="p">,</span> <span class="n">sep</span><span class="p">)</span></div>


<div class="viewcode-block" id="copy_object"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.copy_object">[docs]</a><span class="k">def</span> <span class="nf">copy_object</span><span class="p">(</span><span class="n">obj</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Comparison between objects doesn¬¥t make a copy, </span>
<span class="sd">        but just add another reference. </span>
<span class="sd">        This method do a copy of an object.</span>

<span class="sd">        :param obj: Object to do the copy</span>
<span class="sd">        :type obj: Object</span>

<span class="sd">        :return: A copy of the object</span>
<span class="sd">        :rtype: Object</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">ut</span><span class="o">.</span><span class="n">copy_object</span><span class="p">(</span><span class="n">obj</span><span class="p">)</span></div>


<div class="viewcode-block" id="recursive_delete_data"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.recursive_delete_data">[docs]</a><span class="k">def</span> <span class="nf">recursive_delete_data</span><span class="p">(</span><span class="n">dir_to_delete</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Search in the directory tree the input directory, so it remove its</span>
<span class="sd">        content recursively before remove the root directory.</span>

<span class="sd">        :param dir_to_delete: Directory path to delete its content.</span>
<span class="sd">        :type dir_to_delete: str</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">ut</span><span class="o">.</span><span class="n">recursive_delete_data</span><span class="p">(</span><span class="n">dir_to_delete</span><span class="p">)</span></div>


<div class="viewcode-block" id="calc_plot_dim"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.calc_plot_dim">[docs]</a><span class="k">def</span> <span class="nf">calc_plot_dim</span><span class="p">(</span><span class="n">n_elems</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Calculate the number of rows and columns of a given number</span>
<span class="sd">        in order to find the most squared and symetric image that it is possible</span>

<span class="sd">        :param n_elems: The number of elements.</span>
<span class="sd">        :type n_elems: int</span>
<span class="sd">    </span>
<span class="sd">        :return: The number of rows and columns.</span>
<span class="sd">        :rtype: int, int</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">ut</span><span class="o">.</span><span class="n">calc_plot_dim</span><span class="p">(</span><span class="n">n_elems</span><span class="p">)</span></div>


<div class="viewcode-block" id="print_progress_bar"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.print_progress_bar">[docs]</a><span class="k">def</span> <span class="nf">print_progress_bar</span><span class="p">(</span><span class="n">iteration</span><span class="p">,</span> <span class="n">total</span><span class="p">,</span> <span class="n">mb_iter</span><span class="p">,</span> <span class="n">mb_total</span><span class="p">,</span> <span class="n">prefix</span> <span class="o">=</span> <span class="s1">&#39;Progress:&#39;</span><span class="p">,</span> <span class="n">suffix</span> <span class="o">=</span> <span class="s1">&#39;Complete&#39;</span><span class="p">,</span> <span class="n">length</span> <span class="o">=</span> <span class="mi">50</span><span class="p">,</span> <span class="n">decimals</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>  <span class="n">fill</span> <span class="o">=</span> <span class="s1">&#39;‚ñà&#39;</span><span class="p">,</span> <span class="n">printEnd</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\r</span><span class="s2">&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Print a bar to show the progress of the download.</span>
<span class="sd">        </span>
<span class="sd">        :param iteration: Iterator that indicates the current file that it is being downloaded.</span>
<span class="sd">        :type iteration: int </span>
<span class="sd">        :param total: Total number of files to be downloaded.</span>
<span class="sd">        :type total: int </span>
<span class="sd">        :param mb_iter: Size of the current file in MB.</span>
<span class="sd">        :type mb_iter: float </span>
<span class="sd">        :param mb_total: Total size of the project in MB.</span>
<span class="sd">        :type mb_total: float </span>
<span class="sd">        :param prefix: Message to appear next to the amount of current size downloaded.</span>
<span class="sd">        :type prefix: str </span>
<span class="sd">        :param suffix: Message to appear next to the completed part.</span>
<span class="sd">        :type suffix: str </span>
<span class="sd">        :param length: Total length of the progress bar. Defaults to 50.</span>
<span class="sd">        :type length: int </span>
<span class="sd">        :param decimals: Integer value that indicated the percentage of downloaded data.</span>
<span class="sd">            Defaults to 1.</span>
<span class="sd">        :type decimals: int </span>
<span class="sd">        :param fill: Charcater to print the progress of the bar. Defaults to &#39;‚ñà&#39;.</span>
<span class="sd">        :type fill: str </span>
<span class="sd">        :param printEnd: Character to print at the end of the iteration.    </span>
<span class="sd">            Defaults to &quot;\r&quot; (carriage return). A carriage return means moving </span>
<span class="sd">            the cursor to the beginning of the line.</span>
<span class="sd">        :type printEnd: str </span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">ut</span><span class="o">.</span><span class="n">print_progress_bar</span><span class="p">(</span><span class="n">iteration</span><span class="p">,</span> <span class="n">total</span><span class="p">,</span> <span class="n">mb_iter</span><span class="p">,</span> <span class="n">mb_total</span><span class="p">,</span> <span class="n">prefix</span><span class="p">,</span> <span class="n">suffix</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">decimals</span><span class="p">,</span>  <span class="n">fill</span><span class="p">,</span> <span class="n">printEnd</span><span class="p">)</span></div>



<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">######################################################</span>

<span class="sd">                    READ JSON-DICT</span>

<span class="sd">######################################################</span>
<span class="sd">&quot;&quot;&quot;</span>


<div class="viewcode-block" id="dict_from_dataobject"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.dict_from_dataobject">[docs]</a><span class="k">def</span> <span class="nf">dict_from_dataobject</span><span class="p">(</span><span class="n">dobject</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Generates a Dictionary from DataObject information</span>

<span class="sd">    :param dobject: a object of class DataObject to extract information</span>
<span class="sd">    :type dobject: DataObject</span>

<span class="sd">    :return: a dictionary with the same information as input</span>
<span class="sd">    :rtype: dict</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ut</span><span class="o">.</span><span class="n">dict_from_dataobject</span><span class="p">(</span><span class="n">dobject</span><span class="p">)</span></div>


<div class="viewcode-block" id="json_from_dict"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.json_from_dict">[docs]</a><span class="k">def</span> <span class="nf">json_from_dict</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Generate a Json from dictionary information</span>

<span class="sd">    :param data: a dictionary to extract information</span>
<span class="sd">    :type data: dict</span>

<span class="sd">    :return: a str with json syntax with the same information as input</span>
<span class="sd">    :rtype: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ut</span><span class="o">.</span><span class="n">json_from_dict</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></div>


<div class="viewcode-block" id="dict_from_json"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.dict_from_json">[docs]</a><span class="k">def</span> <span class="nf">dict_from_json</span><span class="p">(</span><span class="n">json</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Serialize a dictionary from json information</span>

<span class="sd">    :param json_data: a json structure to extract information</span>
<span class="sd">    :type json_data: str</span>

<span class="sd">    :return: a dictionary with the same information as input</span>
<span class="sd">    :rtype: dict</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ut</span><span class="o">.</span><span class="n">dict_from_json</span><span class="p">(</span><span class="n">json</span><span class="p">)</span></div>


<div class="viewcode-block" id="dict_from_json_file"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.dict_from_json_file">[docs]</a><span class="k">def</span> <span class="nf">dict_from_json_file</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Serialize a dictionary from json file</span>

<span class="sd">    :param path: a path with the json file to extract information</span>
<span class="sd">    :type path: str</span>

<span class="sd">    :return: a dictionary with the same information as input</span>
<span class="sd">    :rtype: dict</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ut</span><span class="o">.</span><span class="n">dict_from_json_file</span><span class="p">(</span><span class="n">path</span><span class="p">)</span></div>


<div class="viewcode-block" id="json_file_from_dict"><a class="viewcode-back" href="../../geneflow.html#geneflow.geneflow.json_file_from_dict">[docs]</a><span class="k">def</span> <span class="nf">json_file_from_dict</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">path</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Generate a Json file from dictionary information</span>

<span class="sd">    :param data: a dictionary to extract information</span>
<span class="sd">    :type data: dict</span>
<span class="sd">    :param path: path to save new file created. If None, it creates a file on the same</span>
<span class="sd">        path as program with the generic name &#39;object.json&#39;</span>
<span class="sd">    :type path: dict</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ut</span><span class="o">.</span><span class="n">json_file_from_dict</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">path</span><span class="p">)</span></div>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Alba Casillas Rodr√≠guez.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>  

<style>
    .wy-nav-content-wrap {
        height: 100%;
        width: 100%;
        background: hsl(0, 0%, 100%);
        display: block;
    }

    .wy-nav-content {
        height: 100%;
        width: 100%;
        background: hsl(0, 0%, 100%);
    }

    .rst-content {
        width: 100%;
        height: 100%;
    }

    /* Sidebar header (and topbar for mobile) */
    .wy-side-nav-search,
    .wy-nav-top {
        background: hsl(0, 2%, 10%);
    }

    /* Sidebar */
    .wy-nav-side {
        background: #85e4a5;
    }

    .caption {
        background: hsl(0, 0%, 100%);
    }

    .caption-text {
        color: hsl(0, 0%, 0%)
    }

    .wy-menu-vertical a {
        color: hsl(0, 0%, 0%)
    }

    .wy-menu-vertical a:hover {
        color: hsl(0, 0%, 100%)
    }
</style>


</body>
</html>