<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>GeneFlow &mdash; GeneFlow 0.8 documentation</title><link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="&lt;no title&gt;" href="modules.html" />
    <link rel="prev" title="GeneFlow: Tool to analyze gene expression data" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html"><img src="_static/geneflow_logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption"><span class="caption-text">GeneFlow:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">GeneFlow</a></li>
</ul>
<p class="caption"><span class="caption-text">Modules:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="src.html">DataObject</a></li>
<li class="toctree-l1"><a class="reference internal" href="src.html#module-src.etl">ETL</a></li>
<li class="toctree-l1"><a class="reference internal" href="src.html#module-src.model">Model</a></li>
<li class="toctree-l1"><a class="reference internal" href="src.html#module-src.objects">Objects</a></li>
<li class="toctree-l1"><a class="reference internal" href="src.html#module-src.processing">Processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="src.html#module-src.utils">Utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="src.html#module-src.visualize">Visualize</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GeneFlow</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>GeneFlow</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/CasillasAlba/GeneFlow/blob/main/README.mdgeneflow.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="module-geneflow.geneflow">
<span id="geneflow"></span><h1>GeneFlow<a class="headerlink" href="#module-geneflow.geneflow" title="Permalink to this headline">¶</a></h1>
<p>&#64;author: Alba Casillas Rodríguez (<a class="reference external" href="mailto:albacaro&#37;&#52;&#48;correo&#46;ugr&#46;es">albacaro<span>&#64;</span>correo<span>&#46;</span>ugr<span>&#46;</span>es</a>)</p>
<dl class="py function">
<dt id="geneflow.geneflow.add_column">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">add_column</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">name_column</span><span class="o">=</span><span class="default_value">''</span></em>, <em class="sig-param"><span class="n">values</span><span class="o">=</span><span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#add_column"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.add_column" title="Permalink to this definition">¶</a></dt>
<dd><p>Add new column to a DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>name_column</strong> (<em>str</em>) – Name of the new column</p></li>
<li><p><strong>values</strong> (<em>list</em><em>, </em><em>ndarray</em>) – values of the new column. Must have the same lenght as number of rows of DataFrame</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the input DataFrame with a new column with its values</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.add_label">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">add_label</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em>, <em class="sig-param"><span class="n">list_0s</span></em>, <em class="sig-param"><span class="n">column_observed</span></em>, <em class="sig-param"><span class="n">name_label</span><span class="o">=</span><span class="default_value">'label'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#add_label"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.add_label" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Add a special column to a DataFrame.</dt><dd><p>It will be used on Machine Learning as the Target Column of Classification</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>list_0s</strong> (<em>list</em><em>, </em><em>ndarray</em>) – List of values that consider as 0 on Label value</p></li>
<li><p><strong>column_observed</strong> (<em>str</em>) – Column of the DataFrame used to decide when value of target is 0 or 1</p></li>
<li><p><strong>name_label</strong> (<em>list</em><em>, </em><em>ndarray</em>) – values of the new column. Must have the same lenght as number of rows of DataFrame</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a DataFrame with a new column that represents the labels of data to classify</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.bar_plot">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">bar_plot</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em>, <em class="sig-param"><span class="n">x</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">y</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#bar_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.bar_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a bar diagram with the data received as the input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – Data values to create the diagram.</p></li>
<li><p><strong>x</strong> (<em>str</em>) – Name of X axis. Defaults to None.</p></li>
<li><p><strong>y</strong> (<em>str</em>) – Name of Y axis. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Figure that contains the bar plot</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.box_plot">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">box_plot</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em>, <em class="sig-param"><span class="n">x</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">y</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">median</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#box_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.box_plot" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Creates a box plot with the data received as the input.</dt><dd><p>If “median” parameter is true, the function will display a line with the median value.
If x and y are not None, they must be name of row and column of the data.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The data used to create the plot.</p></li>
<li><p><strong>x</strong> (<em>str</em>) – Name of X axis. Defaults to None.</p></li>
<li><p><strong>y</strong> (<em>str</em>) – Name of Y axis. Defaults to None.</p></li>
<li><p><strong>median</strong> (<em>bool</em>) – Indicates if the diagram will show a median line.
Defaults to <cite>False</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Figure that contains the box plot.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.calc_plot_dim">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">calc_plot_dim</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n_elems</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#calc_plot_dim"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.calc_plot_dim" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the number of rows and columns of a given number
in order to find the most squared and symetric image that it is possible</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n_elems</strong> (<em>int</em>) – The number of elements.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The number of rows and columns.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int, int</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.check_all_element">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">check_all_element</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">list_</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">sublist</span><span class="o">=</span><span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#check_all_element"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.check_all_element" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if a given list of elements are on a list or array</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>list</strong> (<em>list</em>) – The current list to be process</p></li>
<li><p><strong>sublist</strong> (<em>list</em><em>, </em><em>ndarray</em>) – list of element to find</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if all elements is on a list, False otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.check_element">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">check_element</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">list_</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">element</span><span class="o">=</span><span class="default_value">''</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#check_element"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.check_element" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if a given element is on a list or array</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>list</strong> (<em>list</em>) – The current list to be process</p></li>
<li><p><strong>element</strong> (<em>int</em><em>, </em><em>float</em><em>, </em><em>str</em><em>, </em><em>bool</em>) – element to find on a list</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if element is on a list, False otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.check_have_header">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">check_have_header</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filename</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#check_have_header"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.check_have_header" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the file has or not header.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>filename</strong> (<em>str</em>) – Name of the file to check if has header or not.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><cite>True</cite> if file has header, <cite>False</cite> if otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.check_sub_element">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">check_sub_element</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">list_</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">partial_sublist</span><span class="o">=</span><span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#check_sub_element"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.check_sub_element" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>It receives a list with elements where it is only needed to find</dt><dd><p>a SUBSTRING of the name to be valid.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>list</strong> (<em>list</em>) – The current list to be process</p></li>
<li><p><strong>partial_sublist</strong> (<em>list</em><em>, </em><em>ndarray</em>) – list of element to find without need to match all name</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of coincidence with the large name</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list, ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.check_sub_element_short">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">check_sub_element_short</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">list_</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">partial_sublist</span><span class="o">=</span><span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#check_sub_element_short"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.check_sub_element_short" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>It receives a list with elements where it is only needed to find</dt><dd><p>a SUBSTRING of the name to be valid, but returns the short name founded</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>list</strong> (<em>list</em>) – The current list to be process</p></li>
<li><p><strong>partial_sublist</strong> (<em>list</em><em>, </em><em>ndarray</em>) – list of element to find without need to match all name</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of coincidence with the short name</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list, ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.clustermap">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">clustermap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em>, <em class="sig-param"><span class="n">color</span><span class="o">=</span><span class="default_value">'YlOrRd'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#clustermap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.clustermap" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Create a clustermap with the data received as the input.</dt><dd><p>A clustermap order data by similarity.
This reorganizes the data for the rows and columns and displays
similar content next to one another for even more
depth of understanding the data.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The data to create the clustermap.</p></li>
<li><p><strong>color</strong> (<em>str</em>) – color spectrum used on heatmap of clustermap</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Figure that contains the clustermap.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.columnames">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">columnames</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#columnames"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.columnames" title="Permalink to this definition">¶</a></dt>
<dd><p>Return on a list the name of the columns of the DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>DataFrame</em>) – a dataframe to obtain the name of its columns</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list with the column names of the data</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.copy_object">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">copy_object</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">obj</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#copy_object"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.copy_object" title="Permalink to this definition">¶</a></dt>
<dd><p>Comparison between objects doesn´t make a copy,
but just add another reference.
This method do a copy of an object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>obj</strong> (<em>Object</em>) – Object to do the copy</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A copy of the object</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Object</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.count_types">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">count_types</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">col_name</span><span class="o">=</span><span class="default_value">''</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#count_types"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.count_types" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Shows a Series containing counts of unique values of a specific column.</dt><dd><p>It will be in descending order so that the first element is the most frequently-occurring element</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>col_name</strong> (<em>str</em>) – name of the column to count types</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>information about types of a column</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.counts_cpm">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">counts_cpm</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">src.dataobject.DataObject</a></span></em>, <em class="sig-param"><span class="n">log_method</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">prior_count</span><span class="o">=</span><span class="default_value">2</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#counts_cpm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.counts_cpm" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Normalize DataObject’s Count Matrix using CPM method</dt><dd><p>It only modify values inside Count Matrix.
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The DataObject to apply normalization.</p></li>
<li><p><strong>log_method</strong> (<em>bool</em>) – True if CPM will do with logaritmic alghoritm, False otherwise</p></li>
<li><p><strong>prior_count</strong> (<em>int</em>) – If log is True, ends up getting scaled by the ratio of a library size to the average library size and then multiplied by its value before getting added to each library size</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject with CPM’s values.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.counts_drop_duplicates">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">counts_drop_duplicates</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">src.dataobject.DataObject</a></span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">keep</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">by_name</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#counts_drop_duplicates"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.counts_drop_duplicates" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Drop DataObject’s Count Matrix duplicates.</dt><dd><p>Drops duplicates in DataObject following a ‘keep’ criteria.
Control that the rest of the data match with new Count Matrix</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>axis</strong> (<em>int</em>) – axis 0 means by rows, 1 by columns</p></li>
<li><p><strong>keep</strong> (<em>str</em><em>, </em><em>bool</em>) – Indicates which duplicates will keep on Count Matrix.
“first” keeps only the first duplicated column/row; “last” keeps only the last duplicated column/row; False keeps both duplicated columns/rows</p></li>
<li><p><strong>by_name</strong> (<em>bool</em>) – True if only evaluate the name of the column/row, False if evaluate the content of the column/row to decide if is duplicated or not</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.counts_drop_nan_by_thresh">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">counts_drop_nan_by_thresh</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">src.dataobject.DataObject</a></span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">thresh</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#counts_drop_nan_by_thresh"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.counts_drop_nan_by_thresh" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Drop DataObject’s Count Matrix values limited by Threshold.</dt><dd><p>Control that the rest of the data match with new Count Matrix
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>axis</strong> (<em>int</em>) – Indicates 0 to row and 1 to columns</p></li>
<li><p><strong>thresh</strong> (<em>float</em><em>, </em><em>int</em>) – threshold to evaluate element. If value less than threshold, that will be selected to drop</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.counts_drop_values">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">counts_drop_values</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">src.dataobject.DataObject</a></span></em>, <em class="sig-param"><span class="n">to_delete</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">method</span><span class="o">=</span><span class="default_value">'all'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#counts_drop_values"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.counts_drop_values" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Drop DataObject’s Count Matrix values.</dt><dd><p>Control that the rest of the data match with new Count Matrix
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>to_delete</strong> (<em>str</em><em>, </em><em>float</em><em>, </em><em>int</em><em>, </em><em>bool</em><em>, </em><em>list</em><em>, </em><em>ndarray</em><em>,</em>) – value to be delete.</p></li>
<li><p><strong>axis</strong> (<em>int</em>) – Indicates 0 to row and 1 to columns</p></li>
<li><p><strong>method</strong> (<em>str</em>) – all to check all values are the same, any if a partial value match</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.counts_projection">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">counts_projection</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">src.dataobject.DataObject</a></span></em>, <em class="sig-param"><span class="n">col_list</span><span class="o">=</span><span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#counts_projection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.counts_projection" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Set DataObject’s Count Matrix selecting a specific column list.</dt><dd><p>Control that the rest of the data match with new Count Matrix
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>col_list</strong> (<em>list</em>) – A list with the names of the columns to keep</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.counts_projection_substring">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">counts_projection_substring</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">src.dataobject.DataObject</a></span></em>, <em class="sig-param"><span class="n">col_list</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">rename</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#counts_projection_substring"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.counts_projection_substring" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Set DataObject’s Count Matrix selecting a specific column list.</dt><dd><p>In that case, is not neccesary that name of the column were exactly
Control that the rest of the data match with new Count Matrix
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>col_list</strong> (<em>list</em>) – A list with the names of the columns to keep</p></li>
<li><p><strong>rename</strong> (<em>bool</em>) – True if the column names of Count Matrix need to be renamed, False otherwise</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.counts_replace">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">counts_replace</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">src.dataobject.DataObject</a></span></em>, <em class="sig-param"><span class="n">to_replace</span><span class="o">=</span><span class="default_value">'Unknown'</span></em>, <em class="sig-param"><span class="n">replaced_by</span><span class="o">=</span><span class="default_value">nan</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#counts_replace"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.counts_replace" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Replace DataObject’s Count Matrix values with another.</dt><dd><p>Control that the rest of the data match with new Count Matrix
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>to_replace</strong> (<em>str</em><em>, </em><em>float</em><em>, </em><em>int</em><em>, </em><em>bool</em><em>, </em><em>list</em><em>, </em><em>ndarray</em>) – values to be replaced</p></li>
<li><p><strong>replaced_by</strong> (<em>str</em><em>, </em><em>float</em><em>, </em><em>int</em><em>, </em><em>bool</em>) – new value to replaced the value before</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.counts_replace_nan">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">counts_replace_nan</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">src.dataobject.DataObject</a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#counts_replace_nan"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.counts_replace_nan" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Replace nan values to mean of the row of Count Matrix</dt><dd><p>Only modifies values of Count Matrix.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.counts_selection">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">counts_selection</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">src.dataobject.DataObject</a></span></em>, <em class="sig-param"><span class="n">row_list</span><span class="o">=</span><span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#counts_selection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.counts_selection" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Set DataObject’s Count Matrix selecting a specific row list.</dt><dd><p>Control that the rest of the data match with new Count Matrix
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>row_list</strong> (<em>list</em>) – A list with the names of the rows to keep</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.counts_transpose">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">counts_transpose</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">src.dataobject.DataObject</a></span></em>, <em class="sig-param"><span class="n">index_name</span><span class="o">=</span><span class="default_value">''</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#counts_transpose"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.counts_transpose" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Do transpose to DataObject’s Count Matrix</dt><dd><p>Control that the rest of the data match with new Count Matrix
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>index_name</strong> (<em>str</em><em>, </em><em>int</em>) – the name of the index before transpose to keep it</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.cpm">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">cpm</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">log_method</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">prior_count</span><span class="o">=</span><span class="default_value">2</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#cpm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.cpm" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalize values of DataFrame using CPM method</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>log_method</strong> (<em>bool</em>) – True if CPM will do with logaritmic alghoritm, False otherwise</p></li>
<li><p><strong>prior_count</strong> (<em>int</em>) – If log is True, ends up getting scaled by the ratio of a library size to the average library size and then multiplied by its value before getting added to each library size</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a DataFrame to computed one of the CPM method</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.create_data_object">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">create_data_object</code><span class="sig-paren">(</span><em class="sig-param">counts_dframe: pandas.core.frame.DataFrame</em>, <em class="sig-param">obs_=Empty DataFrame Columns: [] Index: []</em>, <em class="sig-param">var_=Empty DataFrame Columns: [] Index: []</em>, <em class="sig-param">log=&lt;src.objects.Workflow object&gt;</em>, <em class="sig-param">unsdict={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#create_data_object"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.create_data_object" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a Data Object to save information about a project downloaded or read</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>counts_dframe</strong> (<em>DataFrame</em>) – That will be the count matrix of the DataObject</p></li>
<li><p><strong>obs</strong> (<em>DataFrame</em>) – That will be the clinical information associated to count matrix
on DataObject</p></li>
<li><p><strong>var</strong> (<em>DataFrame</em>) – That will be the gene information associated to count matrix
on DataObject</p></li>
<li><p><strong>log</strong> (<a class="reference internal" href="src.html#src.objects.Workflow" title="src.objects.Workflow"><em>Workflow</em></a>) – That will contains a flow that save step by step the changes of
the DataObject. Useful to replicate an experiment with another information</p></li>
<li><p><strong>unsdict</strong> (<em>dict</em>) – It is a variable used to save extra data with an unclassified distribution
on DataObject.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A DataObject with all information organize</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.create_data_project">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">create_data_project</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#create_data_project"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.create_data_project" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a object that allow to prepare a download for a project.
Essential information such as the project’s name, the path where it has been downloaded,
the list of related files and the total size of the project is saved.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a object with all information save on its attributes</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="src.html#src.objects.DataProject" title="src.objects.DataProject">DataProject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.create_file_project">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">create_file_project</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">id_</span></em>, <em class="sig-param"><span class="n">data_format</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">access</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">file_name</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">submitter_id</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">data_category</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">type_</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">file_size</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">created_datetime</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">md5sum</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">updated_datetime</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">data_type</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">state</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">experimental_strategy</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">version</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">data_release</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">entity_id</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#create_file_project"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.create_file_project" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Creates a object with the information and features of a downloaded file.</dt><dd><p>Only the <a href="#id5"><span class="problematic" id="id6">id_</span></a> will be a mandatory field due to the differences
in the information that different data sources can provide.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>id</strong> (<em>str</em>) – UUID of the file. E.g: 48d05650-971e-4407-9cf1-7e9d71b1ff20</p></li>
<li><p><strong>data_format</strong> (<em>str</em><em>, </em><em>optional</em>) – Format of the Data File, defaults to None.</p></li>
<li><p><strong>access</strong> (<em>str</em><em>, </em><em>optional</em>) – Indicator of whether acces to the data file is open or controlled,
defaults to None.</p></li>
<li><p><strong>file_name</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the file.
E.g: BLAIN_p_TCGA_282_304_b2_N_GenomeWideSNP_6_G08_1348614.grch38.seg.v2.txt, defaults to None.</p></li>
<li><p><strong>submitter_id</strong> (<em>str</em><em>, </em><em>optional</em>) – A unique key to identify the case. Submitter ID is different from the universally
unique identifier (UUID). E.g: 7f4e8b00-d73f-479e-a8f7-30a8d7aee093_dr12_allcnv, defaults to None.</p></li>
<li><p><strong>data_category</strong> (<em>str</em><em>, </em><em>optional</em>) – A high-level data file category. E.g “Raw Sequencing
Data” or “Transcriptome Profiling”, dafults to None</p></li>
<li><p><strong>type</strong> (<em>str</em><em>, </em><em>optional</em>) – Type of the File, defaults to None.</p></li>
<li><p><strong>file_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of the file in MB, defaults to None.</p></li>
<li><p><strong>created_datetime</strong> (<em>datetime</em><em>, </em><em>optional</em>) – Datetime of when the file has been created.</p></li>
<li><p><strong>md5sum</strong> (<em>str</em><em>, </em><em>optional</em>) – hash code. MD5 Checksum is used to verify the integrity of files,
as virtually any change to a file will cause its MD5 hash to change.</p></li>
<li><p><strong>updated_datetime</strong> (<em>datetime</em><em>, </em><em>optional</em>) – Datetatime of when a file had its latest update.</p></li>
<li><p><strong>data_type</strong> (<em>str</em><em>, </em><em>optional</em>) – Data file type. Data Type is more granular than Data Category.
E.g: “Align Reads” or “Gene Expression Quantification”, defaults to None.</p></li>
<li><p><strong>state</strong> (<em>str</em><em>, </em><em>optional</em>) – Current state of the file. E.g: “releassed”, defaults to None.</p></li>
<li><p><strong>experimental_strategy</strong> (<em>str</em><em>, </em><em>optional</em>) – Experimental strategies used for molecular
characterization of the cancer.</p></li>
<li><p><strong>version</strong> (<em>int</em><em>, </em><em>float</em><em>, </em><em>optional</em>) – Number of the current version of the file.</p></li>
<li><p><strong>data_release</strong> (<em>float</em><em>, </em><em>optional</em>) – Number of the data release.</p></li>
<li><p><strong>entity_id</strong> (<em>str</em><em>, </em><em>optional</em>) – Also defined as Barcode. Entity ID is the primary identifier of
biospecimen data within the TCGA project. E.g: TCGA-OR-A5JL-01A-11D-A29H-01, defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a object with information of a specific File</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.objects.FileProject" title="src.objects.FileProject">FileProject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.create_gdc_query">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">create_gdc_query</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">project</span></em>, <em class="sig-param"><span class="n">experimental_strategy</span></em>, <em class="sig-param"><span class="n">data_category</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">data_type</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">workflow_type</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">legacy</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">platform</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">data_format</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">sample_type</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">normalized</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#create_gdc_query"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.create_gdc_query" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a object to do arequest on Genomic Data Commoms to download data</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>project</strong> (<em>str</em>) – Cancer name or ID to download. E.g: “TCGA-BRCA”,
“Acute Myeloid Leukemia”, “adrenocortical carcinoma” …</p></li>
<li><p><strong>experimental_strategy</strong> (<em>str</em>) – Experimental strategies used for molecular
characterization of the cancer.</p></li>
<li><p><strong>data_category</strong> (<em>str</em><em>, </em><em>optional</em>) – A high-level data file category. E.g “Raw Sequencing
Data” or “Transcriptome Profiling”, dafults to None</p></li>
<li><p><strong>data_type</strong> (<em>str</em><em>, </em><em>optional</em>) – Data file type. Data Type is more granular than Data Category.
E.g: “Align Reads” or “Gene Expression Quantification”, defaults to None.</p></li>
<li><p><strong>workflow_type</strong> (<em>str</em><em>, </em><em>optional</em>) – Bioinformatics workflow used to generate or harmonize
the data file. E.g: “STAR - Counts”, “BCGSC miRNA Profiling”, defaults to None.</p></li>
<li><p><strong>legacy</strong> (<em>bool</em>) – The GCD Legacy Archive hosts unharmonized legacy dara from repositories
that predate the GDC (e.g: CGHub). Legacy data is not actively maintained, processed, or
harmonized by the GDC. Legacy is <cite>True</cite> to search for data in the GDC Legacy Archive,
and <cite>False</cite> otherwise; defaults to <cite>False</cite>.</p></li>
<li><p><strong>platform</strong> (<em>str</em><em>, </em><em>optional</em>) – Technolofical platform on which experimental data was produced.
E.g: “Illumina Human Methylation 450”, defaults to None.</p></li>
<li><p><strong>data_format</strong> (<em>str</em><em>, </em><em>optional</em>) – Format of the Data File, defaults to None.</p></li>
<li><p><strong>sample_type</strong> (<em>list</em><em>, </em><em>str</em><em>, </em><em>optional</em>) – Describes the source of a biospecimen used for a laboratory test.
E.g: “Primary Tumor”, [“Primary Tumor”, “Solid Tissue Normal”], defaults to None.</p></li>
<li><p><strong>normalized</strong> (<em>bool</em>) – This parameter is only valid when legacy is set as <cite>True</cite>.
If <cite>True</cite>, normalized data from the GDC Legacy Archive will be downloaded,
and <cite>False</cite> otherwise; defaults to <cite>False</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a object with all neccesary information to download the project</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.objects.GDCQuery" title="src.objects.GDCQuery">GDCQuery</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.create_model_selection">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">create_model_selection</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">source</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#create_model_selection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.create_model_selection" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>This class performs different techniques to the data such as: splitting data,</dt><dd><p>over-sampling and under-sampling, normalization, computes confusion matrix and AUC, etc.,
to the data in order to improve Machine Learning algorithm’s performance.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The data to apply Machine Learning methods. Defaults to None.
If it is not specified, user must provide the <cite>source</cite> and <cite>target</cite> values.</p></li>
<li><p><strong>source</strong> (<em>DataFrame</em>) – The features that describe the samples. Defaults to None.
If it is noy specified, user must provide <cite>data</cite> value.
Otherwise, if it is different to None, <cite>target</cite> values must be provided.</p></li>
<li><p><strong>target</strong> (<em>Series</em>) – The labels to be predicted. Defaults to None.
If it is noy specified, user must provide <cite>data</cite> value.
Otherwise, if it is different to None, <cite>source</cite> values must be provided.</p></li>
<li><p><strong>X_train</strong> (<em>DataFrame</em>) – All the observations that will be used to train the model.
Defaults to None. Only can be updated after use <cite>get_train_test_sample</cite> method.</p></li>
<li><p><strong>Y_train</strong> (<em>Series</em>) – The dependent variable wich need to be predicted by the model.
Defaults to None. Only can be updated after use <cite>get_train_test_sample</cite> method.</p></li>
<li><p><strong>X_test</strong> (<em>DataFrame</em>) – The remaining portion of the independent  variables which
will not be used in the training phase and will be used
to make predictions to test the accuracy of the model.
Defaults to None. Only can be updated after use <cite>get_train_test_sample</cite> method.</p></li>
<li><p><strong>Y_test</strong> (<em>Series</em>) – The labels of the test data. These labels will be used to test
the accuracy between actual and predicted categories. Defaults to None.
Only can be updated after use <cite>get_train_test_sample</cite> method.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a Model Selection with a data used on Machine Learning</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.model.ModelSelection" title="src.model.ModelSelection">ModelSelection</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.create_opin">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">create_opin</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filt</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#create_opin"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.create_opin" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a single in-operator in json format for the query, adding a field and its value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>filt</strong> (<em>list</em><em>[</em><em>str</em><em>]</em>) – values of a filter.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the structured in-operator</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.create_workflow">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">create_workflow</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#create_workflow"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.create_workflow" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Creates a object workflow that allows to save several Log of a DataObject</dt><dd><p>The methods executed over the data can be saved and applied over bulk data
to replicate/expand the same analysis in the future.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>an empty object prepare to save a Workflow of DataObject</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="src.html#src.objects.Workflow" title="src.objects.Workflow">Workflow</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.data_projection_filter">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">data_projection_filter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">filter_</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#data_projection_filter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.data_projection_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract a Sub-DataFrame selecting columns by filter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrane to be process</p></li>
<li><p><strong>filter</strong> (<em>pd.Series</em><em>, </em><em>pd.DataFrame</em>) – a filter to do the projection by columns</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a Sub-DataFrame after apply a filter by columns</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.data_projection_index">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">data_projection_index</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">indx</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#data_projection_index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.data_projection_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract a Sub-DataFrame selecting a specific column by index.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrane to be process</p></li>
<li><p><strong>indx</strong> (<em>int</em>) – a number of a column. Must be on range of lenght columns</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a Sub-DataFrame with selected column</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.data_projection_list">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">data_projection_list</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">column_list</span><span class="o">=</span><span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#data_projection_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.data_projection_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract a Sub-DataFrame selecting a specific column list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrane to be process</p></li>
<li><p><strong>column_list</strong> (<em>list</em>) – A list with the names of the columns to keep</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a Sub-DataFrame with selected columns</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.data_projection_name">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">data_projection_name</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">name</span><span class="o">=</span><span class="default_value">''</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#data_projection_name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.data_projection_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract a Sub-DataFrame selecting a specific column by name.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrane to be process</p></li>
<li><p><strong>name</strong> (<em>str</em>) – a name of a column. Must exists</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a Sub-DataFrame with selected column</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.data_projection_range">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">data_projection_range</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ini_column</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">fini_column</span><span class="o">=</span><span class="default_value">- 1</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#data_projection_range"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.data_projection_range" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract a Sub-DataFrame selecting a range of columns</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrane to be process</p></li>
<li><p><strong>ini_column</strong> (<em>int</em>) – a initial position to project. Must be on range of lenght columns</p></li>
<li><p><strong>fini_column</strong> (<em>int</em>) – a final position to project. Must be on range of lenght columns</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a Sub-DataFrame with the selected columns in an interval</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.data_projection_substring">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">data_projection_substring</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">column_list</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">rename</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#data_projection_substring"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.data_projection_substring" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract a Sub-DataFrame selecting a specific column list no matter if name doesn’t match complete.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrane to be process</p></li>
<li><p><strong>column_list</strong> (<em>list</em>) – A list with the names of the columns to keep</p></li>
<li><p><strong>rename</strong> (<em>bool</em>) – check if should rename the column to the short name. True, then rename, False, not</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a Sub-DataFrame with selected columns</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.data_selection_filter">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">data_selection_filter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">filter_</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#data_selection_filter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.data_selection_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract a Sub-DataFrame selecting rows by filter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrane to be process</p></li>
<li><p><strong>filter</strong> (<em>pd.Series</em><em>, </em><em>pd.DataFrame</em>) – a filter to do the selection by rows</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a Sub-DataFrame after apply a filter by rows</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.data_selection_index">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">data_selection_index</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">indx</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#data_selection_index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.data_selection_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract a Sub-DataFrame selecting a specific row by index.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrane to be process</p></li>
<li><p><strong>indx</strong> (<em>int</em>) – a number of a row. Must be on range of lenght row</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a Sub-DataFrame with selected row</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.data_selection_list">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">data_selection_list</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">row_list</span><span class="o">=</span><span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#data_selection_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.data_selection_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract a Sub-DataFrame selecting a specific row list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrane to be process</p></li>
<li><p><strong>row_list</strong> (<em>list</em>) – A list with the names of the rows to keep</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a Sub-DataFrame with selected rows</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.data_selection_name">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">data_selection_name</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">name</span><span class="o">=</span><span class="default_value">''</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#data_selection_name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.data_selection_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract a Sub-DataFrame selecting a specific row by name.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrane to be process</p></li>
<li><p><strong>name</strong> (<em>str</em>) – a name of a row. Must exists</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a Sub-DataFrame with selected row</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.data_selection_range">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">data_selection_range</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ini_row</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">fini_row</span><span class="o">=</span><span class="default_value">- 1</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#data_selection_range"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.data_selection_range" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract a Sub-DataFrame selecting a range of rows</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrane to be process</p></li>
<li><p><strong>ini_row</strong> (<em>int</em>) – a initial position to select. Must be on range of lenght rows</p></li>
<li><p><strong>fini_row</strong> (<em>int</em>) – a final position to select. Must be on range of lenght rows</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a Sub-DataFrame with the selected rows in an interval</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.data_selection_series">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">data_selection_series</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">series</span><span class="p">:</span> <span class="n">pandas.core.series.Series</span></em>, <em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#data_selection_series"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.data_selection_series" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract a Sub-DataFrame selecting a specific row by series.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrane to be process</p></li>
<li><p><strong>series</strong> (<em>list</em>) – A series with a structure to keep</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a Sub-DataFrame with selected rows</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.data_zip">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">data_zip</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">key_column</span><span class="o">=</span><span class="default_value">''</span></em>, <em class="sig-param"><span class="n">values_column</span><span class="o">=</span><span class="default_value">''</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#data_zip"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.data_zip" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Creates a dictionary using information about two columns</dt><dd><p>of a DataFrame</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>key_column</strong> (<em>str</em>) – column name of a DataFrame used for the dictionary keys</p></li>
<li><p><strong>values_column</strong> (<em>str</em>) – column name of a DataFrame used for the dictionary values</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>with both columns indicates, create and returns a dictionary</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.describe">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">describe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">perc</span><span class="o">=</span><span class="default_value">''</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#describe"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.describe" title="Permalink to this definition">¶</a></dt>
<dd><p>Show a statistic description about a DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>perc</strong> (<em>list</em><em>, </em><em>float</em>) – The percentiles to include in the output</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>descriptive statistics that summarize the central tendency, dispersion and shape
of the dataset’s distribution, excluding NaN values</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.dict_from_dataobject">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">dict_from_dataobject</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#dict_from_dataobject"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.dict_from_dataobject" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a Dictionary from DataObject information</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – a object of class DataObject to extract information</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a dictionary with the same information as input</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.dict_from_json">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">dict_from_json</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">json</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#dict_from_json"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.dict_from_json" title="Permalink to this definition">¶</a></dt>
<dd><p>Serialize a dictionary from json information</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>json_data</strong> (<em>str</em>) – a json structure to extract information</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a dictionary with the same information as input</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.dict_from_json_file">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">dict_from_json_file</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#dict_from_json_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.dict_from_json_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Serialize a dictionary from json file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>path</strong> (<em>str</em>) – a path with the json file to extract information</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a dictionary with the same information as input</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.drop_duplicated">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">drop_duplicated</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">keep</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">by_name</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#drop_duplicated"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.drop_duplicated" title="Permalink to this definition">¶</a></dt>
<dd><p>Drop data duplicates of a DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>axis</strong> (<em>int</em>) – axis 0 means by rows, 1 by columns</p></li>
<li><p><strong>keep</strong> (<em>str</em><em>, </em><em>bool</em>) – Indicates which duplicates will keep
“first” keeps only the first duplicated column/row; “last” keeps only the last duplicated column/row; False keeps both duplicated columns/rows</p></li>
<li><p><strong>by_name</strong> (<em>bool</em>) – True if only evaluate the name of the column/row, False if evaluate the content of the column/row to decide if is duplicated or not</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A new dataframe without duplicates rows or columns</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.drop_nan_by_thresh">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">drop_nan_by_thresh</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">thresh</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#drop_nan_by_thresh"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.drop_nan_by_thresh" title="Permalink to this definition">¶</a></dt>
<dd><p>Drop DataFrame’s values limited by Threshold.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>axis</strong> (<em>int</em>) – Indicates 0 to row and 1 to columns</p></li>
<li><p><strong>thresh</strong> (<em>float</em><em>, </em><em>int</em>) – threshold to evaluate element. If value less than threshold, that will be selected to drop</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dataframe with drop element checking a threshold</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.drop_rows_by_colname">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">drop_rows_by_colname</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">col_name</span><span class="o">=</span><span class="default_value">''</span></em>, <em class="sig-param"><span class="n">to_delete</span><span class="o">=</span><span class="default_value">nan</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#drop_rows_by_colname"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.drop_rows_by_colname" title="Permalink to this definition">¶</a></dt>
<dd><p>Drop DataFrame’s rows searching a value on a specific column</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>col_name</strong> (<em>str</em>) – name of the column to search</p></li>
<li><p><strong>to_delete</strong> (<em>str</em><em>, </em><em>float</em><em>, </em><em>int</em><em>, </em><em>bool</em><em>, </em><em>list</em><em>, </em><em>ndarray</em>) – value to find on a column to decide with row delete is exists</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dataframe with dropped rows search a value on a specific column</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.drop_values">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">drop_values</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">to_delete</span><span class="o">=</span><span class="default_value">nan</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">method</span><span class="o">=</span><span class="default_value">'all'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#drop_values"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.drop_values" title="Permalink to this definition">¶</a></dt>
<dd><p>Drop values of a DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>to_delete</strong> (<em>str</em><em>, </em><em>float</em><em>, </em><em>int</em><em>, </em><em>bool</em><em>, </em><em>list</em><em>, </em><em>ndarray</em><em>,</em>) – value to be delete.</p></li>
<li><p><strong>axis</strong> (<em>int</em>) – Indicates 0 to row and 1 to columns</p></li>
<li><p><strong>method</strong> (<em>str</em>) – all to check all values are the same, any if a partial value match</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A new dataframe within this values</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.duplicated_corr_plot">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">duplicated_corr_plot</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data_duplicated</span></em>, <em class="sig-param"><span class="n">ini_row</span><span class="o">=</span><span class="default_value">1001</span></em>, <em class="sig-param"><span class="n">fin_row</span><span class="o">=</span><span class="default_value">2000</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#duplicated_corr_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.duplicated_corr_plot" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Create a diagram to display the correlation between duplicate variables.</dt><dd><p>In case the original dataFrame is set as the input, the method will be check
for duplicates. If the DataFrame contains only the duplicates, this checking
will not affect the result.
The layout wil be created the most squared and symetric as possible.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_duplicated</strong> (<em>DataFrame</em>) – The duplicated data used to create the diagram</p></li>
<li><p><strong>ini_row</strong> (<em>int</em>) – First row to select the data will be used. Defaults to 1001.</p></li>
<li><p><strong>fin_row</strong> (<em>int</em>) – Final row to select the data will be used. Dafults to 2000.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>figure that contains a diagram with duplicated data of a correlation</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.duplicates">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">duplicates</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">keep</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#duplicates"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.duplicates" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a DataFrame with the duplicates rows or columns of a input DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>axis</strong> (<em>int</em>) – axis 0 means by rows, 1 by columns</p></li>
<li><p><strong>keep</strong> (<em>str</em><em>, </em><em>bool</em>) – Indicates which duplicates will keep
“first” keeps only the first duplicated column/row; “last” keeps only the last duplicated column/row; False keeps both duplicated columns/rows</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dataframe that only contains duplicated columns or rows</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.explain_all_variable_colname">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">explain_all_variable_colname</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">samples</span></em>, <em class="sig-param"><span class="n">colnm</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#explain_all_variable_colname"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.explain_all_variable_colname" title="Permalink to this definition">¶</a></dt>
<dd><p>Return specific field about all variables of a DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samples</strong> – a dataframe with information of variables</p></li>
<li><p><strong>colnm</strong> (<em>str</em><em>, </em><em>int</em>) – specific column name of the DataFrame (field)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a dataframe with specific field of information about all variables</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.explain_variable">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">explain_variable</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">samples</span></em>, <em class="sig-param"><span class="n">var_id</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#explain_variable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.explain_variable" title="Permalink to this definition">¶</a></dt>
<dd><p>Return information about a specific variable of a DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samples</strong> – a dataframe with information of variables</p></li>
<li><p><strong>var_id</strong> (<em>str</em><em>, </em><em>int</em>) – specific identificator of a variable</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>specific information of a variable</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>series</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.explain_variable_colname">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">explain_variable_colname</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">samples</span></em>, <em class="sig-param"><span class="n">var_id</span></em>, <em class="sig-param"><span class="n">colnm</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#explain_variable_colname"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.explain_variable_colname" title="Permalink to this definition">¶</a></dt>
<dd><p>Return specific field about a specific variable of a DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samples</strong> – a dataframe with information of variables</p></li>
<li><p><strong>var_id</strong> (<em>str</em><em>, </em><em>int</em>) – specific identificator of a variable</p></li>
<li><p><strong>colnm</strong> (<em>str</em><em>, </em><em>int</em>) – specific column name of the DataFrame (field)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>specific field of information of a variable</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>series</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.fill_nan">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">fill_nan</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">to_replace</span><span class="o">=</span><span class="default_value">'Unknown'</span></em>, <em class="sig-param"><span class="n">replaced_by</span><span class="o">=</span><span class="default_value">nan</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#fill_nan"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.fill_nan" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Replace nan values to mean of the row of DataFrame</dt><dd><p>Only modifies values of DataFrame.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A DataFrame with all nan values replace to mean</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.filter_by_index">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">filter_by_index</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">index_cond</span><span class="o">=</span><span class="default_value">''</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#filter_by_index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.filter_by_index" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Extract a Sub-DataFrame selected by a filter of index condition.</dt><dd><p>For multi-index cases:
index_cond must have index type (it is returned after  do dataframe.index)
It keeps only the rows that are in index_cond</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrane to be process</p></li>
<li><p><strong>index_cond</strong> (<em>str</em>) – a condition as filter</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a Sub-DataFrame that keeps only the rows that are in index_cond</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.filter_dictionary">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">filter_dictionary</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">list_var</span><span class="o">=</span><span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#filter_dictionary"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.filter_dictionary" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a Dictionary, select only elements on a passed list</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dict</strong> (<em>dict</em>) – The current Dictionary to be process</p></li>
<li><p><strong>list_var</strong> (<em>list</em>) – new Count Matrix</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A new dictionary with filtered elements</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.find_parameter">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">find_parameter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">value</span></em>, <em class="sig-param"><span class="n">param_results</span></em>, <em class="sig-param"><span class="n">param</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#find_parameter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.find_parameter" title="Permalink to this definition">¶</a></dt>
<dd><p>Search in the result’s list the desired value.
If it is not found, the method shows all available parameters.
If missing is returned, that option will be removed from the result’s list
because it is an undesired value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>value</strong> (<em>str</em>) – desired value to find on a parameter.</p></li>
<li><p><strong>param_results</strong> (<em>list</em><em>[</em><em>str</em><em>]</em>) – available parameters.</p></li>
<li><p><strong>param</strong> (<em>str</em>) – param where to find.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>param found.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.from_dict">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">from_dict</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">json_task</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#from_dict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.from_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a object Task replicate from json or dictionary</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a object Task replicated</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="src.html#src.processing.Task" title="src.processing.Task">Task</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_check_data_type">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_check_data_type</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">endpt</span></em>, <em class="sig-param"><span class="n">filters</span></em>, <em class="sig-param"><span class="n">data_type</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_check_data_type"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_check_data_type" title="Permalink to this definition">¶</a></dt>
<dd><p>Check the data type selected. We need to have a single data type to
proceed with the download. So, data type cannot be a list.
If None, we have to check if all the filters selected return
only one data type to download.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>endpt</strong> (<em>str</em>) – URL where the request is going to be done.</p></li>
<li><p><strong>filters</strong> (<em>list</em>) – current list of filters.</p></li>
<li><p><strong>data_type</strong> (<em>str</em>) – Data Type to check.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_check_parameter">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_check_parameter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">endpt</span></em>, <em class="sig-param"><span class="n">param</span></em>, <em class="sig-param"><span class="n">value</span></em>, <em class="sig-param"><span class="n">filters</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_check_parameter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_check_parameter" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the parameter selected is available with the actual query.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>endpt</strong> (<em>str</em>) – URL where the request is going to be done.</p></li>
<li><p><strong>param</strong> (<em>str</em>) – Name of the parameter to check.</p></li>
<li><p><strong>value</strong> (<em>str</em>) – Value of the parameter to check.</p></li>
<li><p><strong>filters</strong> (<em>list</em>) – current list of filters.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The value of parameter with the correct format to add
it to the query.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_clinical_filename">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_clinical_filename</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_clinical_filename"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_clinical_filename" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a string with the main name of a clinical filename of GDC</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a string with the main name.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_clinical_options">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_clinical_options</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_clinical_options"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_clinical_options" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of options for clinical information</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>list of options</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_clinical_query">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_clinical_query</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">query_object</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_clinical_query"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_clinical_query" title="Permalink to this definition">¶</a></dt>
<dd><p>Query in the TCGA API to get the clinical info linked to a
specific project.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>project_name</strong> (<em>str</em>) – Name of the project.</p>
</dd>
</dl>
<dl class="simple">
<dt>:param legacy:<cite>True</cite> to search for data in the GDC Legacy Archive,</dt><dd><p>and <cite>False</cite> otherwise; defaults to <cite>False</cite>.</p>
</dd>
</dl>
<dl class="field-list simple">
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_create_gene_model">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_create_gene_model</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path_gtf</span><span class="o">=</span><span class="default_value">'genecode_v36.gtf'</span></em>, <em class="sig-param"><span class="n">path_save</span><span class="o">=</span><span class="default_value">'genecode_v36.csv'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_create_gene_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_create_gene_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a DataFrame that allow an easy treatment with information about Genes.
The function will create a CSV readable with features of Genes. If CSV exist, there is
not neccesary to call this function, only read it.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path_gtf</strong> (<em>str</em>) – Path where genecode with GTF format is located</p></li>
<li><p><strong>path_save</strong> – Path where the CSV result will be save</p></li>
<li><p><strong>path_save</strong> – str</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_data_legacy">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_data_legacy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_data_legacy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_data_legacy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a boolean that indicates if the data is from the GDC
Legacy Archive or from the GDC Data Portal.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><cite>True</cite> to search for data in the GDC Legacy Archive,
and <cite>False</cite> otherwise.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_download_clinical_data">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_download_clinical_data</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">project_name</span></em>, <em class="sig-param"><span class="n">clinic_info</span></em>, <em class="sig-param"><span class="n">legacy</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">os_path</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_download_clinical_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_download_clinical_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Download clinical data from a specified project. A folder with the name of the
project will be created if it does not already exist. If it exists, it will be
checked if the clinical information has been previously downloaded.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>project_name</strong> (<em>str</em>) – Name of the project.</p></li>
<li><p><strong>clinic_info</strong> (<em>str</em>) – Type of clinical information that will be downloaded.
It will be downloaded the latest version found. E.g: “patient” or “drug”.</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>:param legacy:<cite>True</cite> to search for data in the GDC Legacy Archive,</dt><dd><p>and <cite>False</cite> otherwise; defaults to <cite>False</cite>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>os_path</strong> (<em>str</em><em>, </em><em>optional</em>) – Directory path were the file will be saved. Defaults to None, so
it will be saved in the current directory.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_download_data">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_download_data</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">query_object</span></em>, <em class="sig-param"><span class="n">os_path</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_download_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_download_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Download the data with the characteristics specidied in the DataObject object.
Before downloading the data, it is checked if all fields have a correct value and
existing information.
A folder with the ID of the cancer will be created to save the data. It will also be
saved a manifest.txt file with the list of the ID of the files and their Entity ID.
The Entity ID is useful to match samples with clinical information downloaded from the
portal.
All the data is downloaded using the TCGA API via requests.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>query_object</strong> (<a class="reference internal" href="src.html#src.objects.GDCQuery" title="src.objects.GDCQuery"><em>GDCQuery</em></a>) – GDCQuery object with the features of the data that will be downloaded.</p></li>
<li><p><strong>os_path</strong> (<em>str</em><em>, </em><em>optional</em>) – Directory path were the file will be saved.Defaults to None, so
it will be saved in the current directory.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_download_genecode">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_download_genecode</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">url_gtf</span><span class="o">=</span><span class="default_value">'https://api.gdc.cancer.gov/data/be002a2c-3b27-43f3-9e0f-fd47db92a6b5'</span></em>, <em class="sig-param"><span class="n">name_gtf</span><span class="o">=</span><span class="default_value">'genecode_v36'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_download_genecode"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_download_genecode" title="Permalink to this definition">¶</a></dt>
<dd><p>Download genecode to create a data structure with information about
specific genes. It will allow building normalization functions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>url_gtf</strong> (<em>str</em>) – URL where the genecode is hosted. By default, the program download from GDC and its 36th version.</p></li>
<li><p><strong>name_gft</strong> – Name of the file (WITHOUT extension). Its extension must be .gtf</p></li>
<li><p><strong>name_gft</strong> – str</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_endpt_base">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_endpt_base</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_endpt_base"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_endpt_base" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a string with the URL to search in the TCGA base</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The URL to search in the GDC base.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_endpt_data">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_endpt_data</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_endpt_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_endpt_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a string with the URL to files endpoint</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The URL to search in the GDC files.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_endpt_files">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_endpt_files</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_endpt_files"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_endpt_files" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a string with the URL to files endpoint</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The URL to search in the GDC files.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_endpt_legacy">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_endpt_legacy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_endpt_legacy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_endpt_legacy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a string with the URL to search in the TCGA Legacy Archive</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The URL to search in the GDC Legacy Archive.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_endpt_projects">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_endpt_projects</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_endpt_projects"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_endpt_projects" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a string with the URL to projects endpoint</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The URL to search in the GDC projects.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_endpt_status">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_endpt_status</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_endpt_status"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_endpt_status" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a string with the URL to check status</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The URL to check status</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_find_clinical">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_find_clinical</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_list</span></em>, <em class="sig-param"><span class="n">clinical_option</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_find_clinical"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_find_clinical" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds if the selected clinical option is one of the available list
of files.
First, we only kept with those files from the list of files that match
a pattern where it is searched tyhe word “clinical” in the name of the file.
Then, in a new list with only the clinical files, search for a file that have
the specified clinical option.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file_list</strong> (<em>list</em>) – List of files to search the clinical option’s file.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><cite>True</cite> if the clinical option has been found and <cite>False</cite> otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_form_filter">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_form_filter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filters</span></em>, <em class="sig-param"><span class="n">nameFilter</span></em>, <em class="sig-param"><span class="n">valueFilter</span></em>, <em class="sig-param"><span class="n">fullNameFilter</span></em>, <em class="sig-param"><span class="n">endpt_</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_form_filter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_form_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates the filter to do the query that gets the data.
If the new value for the query is available, it will be added to
the filter’s list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filters</strong> (<em>list</em>) – Current list with the filters.</p></li>
<li><p><strong>nameFilter</strong> (<em>str</em>) – Name of the filter to be checked and added.</p></li>
<li><p><strong>valueFilter</strong> (<em>str</em>) – Value of the filter to be checked and added.</p></li>
<li><p><strong>fullNameFilter</strong> (<em>str</em>) – Name of the filter to be added plus the path
where it is searched in the API.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_get_projects">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_get_projects</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_get_projects"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_get_projects" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets all available GDC projects</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>List of available projects in the GDC Data Portal.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_query">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_query</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">query_object</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_query"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_query" title="Permalink to this definition">¶</a></dt>
<dd><p>It creates the correct query using the data specified in the input DataObject
and gets the data that it is going to be downloaded.
To do it, it is checked each parameter availability taking into account the filter
selected.
If the parameter is available, it is added to the filter’s list before checking
the next parameter.
Once all filters are set, we get the request.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>query_object</strong> (<a class="reference internal" href="src.html#src.objects.GDCQuery" title="src.objects.GDCQuery"><em>GDCQuery</em></a>) – GDCQuery that specified all the fields to get the data desired.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The response of the request, in case all parameters are available.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_read_clinical">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_read_clinical</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_name</span></em>, <em class="sig-param"><span class="n">sep</span><span class="o">=</span><span class="default_value">'\t'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_read_clinical"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_read_clinical" title="Permalink to this definition">¶</a></dt>
<dd><p>Read a clinical file preivously downaloaded from The Cancer Genome Atlas (TCGA) portal.
In TCGA clinical files, the first line is a comment that it is skipped. The two following lines
have the header names; only the second header’s line is kept, due to they have the same values
but with different names.
Returns a dataframe with the clinical information where the index is set to <cite>bcr_patient_barcode</cite>,
that match with the Entity ID of the samples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> (<em>str</em>) – The name of the clinical project to read. By default, the programn search the file
in the actual path, so the location of the file is required to find the file.
E.g: “TCGA-BRCA/nationwidechildrens.org_clinical_patient_brca.txt”.</p></li>
<li><p><strong>sep</strong> (<em>str</em>) – Separator character between fields. Defaults to ”   ” (Tab-separated values).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame with the clinical information read. The index has been set to
<cite>bcr_patient_barcode</cite>, matching with the Entity ID of the samples.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_read_file_rna">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_read_file_rna</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_name</span></em>, <em class="sig-param"><span class="n">col</span></em>, <em class="sig-param"><span class="n">sep</span><span class="o">=</span><span class="default_value">'\t'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_read_file_rna"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_read_file_rna" title="Permalink to this definition">¶</a></dt>
<dd><p>Read a single Rna-Seq file from The GDC portal.
The first line of this type of files is a comment that it is skipped.
The second line will be the header of the data.
Next lines are skipped / saved in a summary_file because they store
summary information of the file.
The index is set to the first column, which is the gene_id.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – File name to be read.</p></li>
<li><p><strong>col</strong> (<em>str</em>) – Name of the column to be read. Set to <cite>unstranded</cite>.</p></li>
<li><p><strong>sep</strong> (<em>str</em>) – Separator character between fields. Defaults to ”   ” (Tab-separated values).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame with file’s data read.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_read_file_rna_legacy">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_read_file_rna_legacy</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_name</span></em>, <em class="sig-param"><span class="n">col</span></em>, <em class="sig-param"><span class="n">sep</span><span class="o">=</span><span class="default_value">'\t'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_read_file_rna_legacy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_read_file_rna_legacy" title="Permalink to this definition">¶</a></dt>
<dd><p>Read a single Rna-Seq file from The TCGA Legacy Archive.
The first line is the column’s names.
An undefined number of rows are skipped while their first character
is a <cite>?</cite>.
The index is set to the first column, which is the gene_id.
In this type of files the gene_id are made up of the ID of the gene and a number,
separated by a “|”. Only the ID of the gene is saved.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – File name to be read.</p></li>
<li><p><strong>col</strong> (<em>str</em>) – Name of the column to be read. Set to <cite>unstranded</cite>.</p></li>
<li><p><strong>sep</strong> (<em>str</em>) – Separator character between fields. Defaults to ”   ” (Tab-separated values).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame with file’s data read.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_read_rna">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_read_rna</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dir_path</span></em>, <em class="sig-param"><span class="n">sep</span><span class="o">=</span><span class="default_value">'\t'</span></em>, <em class="sig-param"><span class="n">save</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_read_rna"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_read_rna" title="Permalink to this definition">¶</a></dt>
<dd><p>Read Rna-Seq files from The Cancer Genome Atlas (TCGA) portal.
The file will be read in a different way depending on whether it is from
the Legacy Data portal or from the GDC Data portal.
For the GDC portal, unstranded data is downloaded by default.
For the Legacy Archive, it will be read the second column of the file
that can be raw counts or normalized counts. While downloading the data,
the user can indicate the type of data to dowload by the parameter “normalized”.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dir_path</strong> (<em>str</em>) – The path of the folder where the project has been downloaded.
By feault, the program search the folder in the current directory, so if
the project has been saved in a different folder, the user must indicate
the full path.</p></li>
<li><p><strong>sep</strong> (<em>str</em>) – Separator character between fields. Defaults to ”   ” (Tab-separated values).</p></li>
<li><p><strong>save</strong> – <cite>True</cite> to keep the downloaded data in the folder and <cite>False</cite> to remove</p></li>
</ul>
</dd>
</dl>
<p>the downloaded data after read it; defaults to <cite>True</cite>.
:type save: bool
:return: DataFrame with the RNA-Seq information read.
:rtype: DataFrame</p>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_regex_file_name">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_regex_file_name</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">proj_id</span></em>, <em class="sig-param"><span class="n">clinical_option</span></em>, <em class="sig-param"><span class="n">file_name</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_regex_file_name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_regex_file_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds the clinical option even if it is a new version of the file.
The version of the file is controlled by a regex expression.
E.g:  _omf_v1.0 or _omf_v12.9 can be part of the file name</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>proj_id</strong> (<em>str</em>) – ID of the project. E.g: if the project is “TCGA-SKCM”, the id
will be “skcm”.</p></li>
<li><p><strong>clinical_option</strong> (<em>str</em>) – Clinical option to search in the regex expression.</p></li>
<li><p><strong>file_name</strong> (<em>str</em>) – File to check if it has the clinical option in its name, even
if it has a version pattern.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><cite>True</cite> if the clinical option has been found and <cite>False</cite> otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_search_project">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_search_project</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">ident</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_search_project"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_search_project" title="Permalink to this definition">¶</a></dt>
<dd><p>Search the project by the input “ident” variable. It could be the identifier
or the complete name of the cancer. If exists, the ID will be returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ident</strong> (<em>str</em>) – Identificator (ID/Name) to search the project.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The identifier of the project. E.g: TCGA-BRCA.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_set_data_legacy">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_set_data_legacy</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">legacy</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_set_data_legacy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_set_data_legacy" title="Permalink to this definition">¶</a></dt>
<dd><p>Set if data comes from the GDC Legacy Archive or not.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>legacy</strong> (<em>bool</em>) – <cite>True</cite> to search for data in the GDC Legacy Archive,
and <cite>False</cite> otherwise.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.gdc_write_to_dataframe">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">gdc_write_to_dataframe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">list_</span></em>, <em class="sig-param"><span class="n">file</span></em>, <em class="sig-param"><span class="n">col</span></em>, <em class="sig-param"><span class="n">entity_id</span></em>, <em class="sig-param"><span class="n">legacy</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#gdc_write_to_dataframe"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.gdc_write_to_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a list of the DataFrame’s lines, so it allows multithreading
while reading the file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>list</strong> (<em>list</em>) – List to append DataFrame’s content. It is modified by reference.</p></li>
<li><p><strong>file</strong> (<em>str</em>) – file to be read.</p></li>
<li><p><strong>col</strong> (<em>str</em>) – Name of the column to be read. If legacy is <cite>True</cite>,
it is set to the second column of the DataFrame; if legacy is
<cite>False</cite>, column is set to <cite>unstranded</cite>.</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>:param legacy:<cite>True</cite> to search for data in the GDC Legacy Archive,</dt><dd><p>and <cite>False</cite> otherwise; defaults to <cite>False</cite>.</p>
</dd>
</dl>
<dl class="field-list simple">
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.get_json_query">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">get_json_query</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">url_query</span></em>, <em class="sig-param"><span class="n">params_query</span><span class="o">=</span><span class="default_value">{}</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#get_json_query"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.get_json_query" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a structured JSON just by passing a URL, so it can be used by all the models.
Each model will processes it as it is structured in its API.
Params are empty by default, to get the data with or without parameters</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>url_query</strong> (<em>str</em>) – URL to do the request.</p></li>
<li><p><strong>params_query</strong> (<em>dict</em>) – params to filter the query.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>data obtained from the request.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>json</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.heatmap">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">heatmap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">x</span></em>, <em class="sig-param"><span class="n">y</span></em>, <em class="sig-param"><span class="n">z</span></em>, <em class="sig-param"><span class="n">color</span><span class="o">=</span><span class="default_value">'YlOrRd'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#heatmap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.heatmap" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a heatmap with the data received as the input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<em>list</em>) – The data to create the heatmap.</p></li>
<li><p><strong>y</strong> (<em>list</em>) – The data to create the heatmap.</p></li>
<li><p><strong>z</strong> (<em>list</em>) – The data to create the heatmap.</p></li>
<li><p><strong>color</strong> (<em>str</em>) – color spectrum used on heatmap</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Figure that contains the heatmap.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.json_file_from_dict">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">json_file_from_dict</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em>, <em class="sig-param"><span class="n">path</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#json_file_from_dict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.json_file_from_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a Json file from dictionary information</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>dict</em>) – a dictionary to extract information</p></li>
<li><p><strong>path</strong> (<em>dict</em>) – path to save new file created. If None, it creates a file on the same
path as program with the generic name ‘object.json’</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.json_from_dict">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">json_from_dict</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#json_from_dict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.json_from_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a Json from dictionary information</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>dict</em>) – a dictionary to extract information</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a str with json syntax with the same information as input</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.kde_plot">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">kde_plot</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#kde_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.kde_plot" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Creates a Kernel Density Estimation (kde) diagram with the data received as the input.</dt><dd><p>A kernel density estimate (KDE) plot is a method for visualizing the distribution
of observations in a dataset, analagous to a histogram.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>DataFrame</em>) – Data values to create the diagram.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Figure that contains the KDE plot.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.list_intersection">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">list_intersection</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">list_</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">sublist</span><span class="o">=</span><span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#list_intersection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.list_intersection" title="Permalink to this definition">¶</a></dt>
<dd><p>Obtains a new list with commom elements of two list or array</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>list</strong> (<em>list</em>) – The primary list to be process</p></li>
<li><p><strong>sublist</strong> (<em>list</em>) – secondary list to find commom elements</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list which is the intersection between both input lists</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.list_intersection_substring">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">list_intersection_substring</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">list_</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">substrings</span><span class="o">=</span><span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#list_intersection_substring"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.list_intersection_substring" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Obtains a new list with commom elements of two lists or array</dt><dd><p>It is enough if only match partial name of element</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>list</strong> (<em>list</em>) – The primary list to be process</p></li>
<li><p><strong>substrings</strong> (<em>list</em>) – secondary list to find commom elements</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list with founded common substring between two list</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.list_zip">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">list_zip</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">list_</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">sublist</span><span class="o">=</span><span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#list_zip"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.list_zip" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Creates a new dictionary using one list</dt><dd><p>for the keys, and other to its values</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>list</strong> (<em>list</em>) – The current list to be process</p></li>
<li><p><strong>sublist</strong> (<em>list</em>) – a list with values of the dictionary</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>with the current list used as keys, creates and returns a dictionary with second list as values</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.mds_plot">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">mds_plot</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em>, <em class="sig-param"><span class="n">color_list</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">clinical_info</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">symbols</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">text_plot</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#mds_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.mds_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a multidimensional diagram plot with the data received as the input.
If “color_list” parameter is empty, colors will be formed automatically by function,
otherwise, lenght of them must be the same as lenght of data columns
If “text_plot” is true, the name of the columns will appear over the point its point on the plot,
if False will only appear the point</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The data used to create the plot.</p></li>
<li><p><strong>color_list</strong> (<em>list</em>) – A list of colors to be used on plot</p></li>
<li><p><strong>clinical_info</strong> (<em>list</em>) – A classification of the samples to represent them by types</p></li>
<li><p><strong>symbols</strong> (<em>list</em>) – A classification of the samples to represent them by symbols</p></li>
<li><p><strong>text_plot</strong> (<em>bool</em>) – Indicates if name of column should appear over its representation on plot</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Figure that contains the mds plot.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.model_logistic_regression">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">model_logistic_regression</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span></em>, <em class="sig-param"><span class="n">penalty</span><span class="o">=</span><span class="default_value">'l2'</span></em>, <em class="sig-param"><span class="n">dual</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">tol</span><span class="o">=</span><span class="default_value">0.0001</span></em>, <em class="sig-param"><span class="n">C</span><span class="o">=</span><span class="default_value">1.0</span></em>, <em class="sig-param"><span class="n">fit_intercept</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">intercept_scaling</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">class_weight</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">random_state</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">solver</span><span class="o">=</span><span class="default_value">'lbfgs'</span></em>, <em class="sig-param"><span class="n">max_iter</span><span class="o">=</span><span class="default_value">100</span></em>, <em class="sig-param"><span class="n">multi_class</span><span class="o">=</span><span class="default_value">'auto'</span></em>, <em class="sig-param"><span class="n">verbose</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">warm_start</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">n_jobs</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">l1_ratio</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#model_logistic_regression"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.model_logistic_regression" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a Logistic Regression algorithm from scikit-learn library.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – Name of the model.</p></li>
<li><p><strong>penalty</strong> (<cite>l1</cite>,`l2`,`elasticnet`, <cite>none</cite>) – Specify the norm of the penalty, where <cite>none</cite> means no penalty is added,
<cite>l2</cite> uses Ridge regularization, <cite>l1</cite> uses Lasso regularization, and <cite>elasticnet</cite>
means that both <cite>l1</cite> and <cite>l2</cite> penalty terms are added. Defaults to <cite>l2</cite>.</p></li>
<li><p><strong>dual</strong> (<em>bool</em>) – Dual or primal formulation. Dual formulation is only implemented
for l2 penalty with liblinear solver. Defauls to <cite>False</cite>.</p></li>
<li><p><strong>tol</strong> (<em>float</em>) – Tolerance for stopping criteria. Defaults to <cite>1e-4</cite>.</p></li>
<li><p><strong>C</strong> (<em>float</em>) – Inverse of regularization strenght, must be a positive float.
Smaller values specify stronger regularization. Defaults to <cite>1.0</cite>.</p></li>
<li><p><strong>fit_intercept</strong> (<em>bool</em>) – Specifies if a constant (bias or intercept) should be added
to the decision function. Defaults to <cite>True</cite>.</p></li>
<li><p><strong>intercept_scaling</strong> (<em>float</em>) – Useful only when the solver <cite>liblinear</cite> is used and
<cite>fit_intercept</cite> True. In this case, x becomes [x, self.intercept_scaling].
Defaults to <cite>1</cite>.</p></li>
<li><p><strong>class_weight</strong> (dict, <cite>balanced</cite>, or None) – Weights associated with classes in the form {class_label: weight}.
Defaults to None.</p></li>
<li><p><strong>random_state</strong> (<em>int</em><em>, </em><em>RandomState instance</em><em>, or </em><em>None</em>) – Used when solver == <cite>sag</cite>, <cite>saga</cite> or <cite>liblinear</cite> to shuffle the data.
Defaults to None.</p></li>
<li><p><strong>solver</strong> ({<cite>newton-cg</cite>, <cite>lbfgs</cite>, <cite>liblinear</cite>, <cite>sag</cite>, <cite>saga</cite>}) – Algorith to use in the optimization problem. Defaults to <cite>lbfgs</cite>.</p></li>
<li><p><strong>max_iter</strong> (<em>int</em>) – Maximum number of iterations taken for the solvers to converge.
Defaults to <cite>100</cite>.</p></li>
<li><p><strong>multi_class</strong> ({<cite>auto</cite> , <cite>ovr</cite> , <cite>multinomial</cite>}) – If the option chosen is <cite>ovr</cite>, then a binary problem is fit
for each label. For <cite>multinomial</cite> the loss minimised is the multinomial loss fit
across the entire probability distribution, even when the data is binary.
Defaults to <cite>auto</cite>.</p></li>
<li><p><strong>verbose</strong> (<em>int</em>) – For the liblinear and lbfgs solvers set verbose to any positive
number for verbosity. Defaults to <a href="#id1"><span class="problematic" id="id2">``</span></a>0`.</p></li>
<li><p><strong>warm_start</strong> (<em>bool</em>) – If <cite>True</cite>, reuse the solution of the previous call to fit
as initialization, otherwise, just erase the previous solution. Defauls to <cite>False</cite>.</p></li>
<li><p><strong>n_jobs</strong> (<em>int</em>) – Number of CPU cores used when parallelizing over classes of multi_class=`ovr`.
Defaults to None.</p></li>
<li><p><strong>l1_ratio</strong> (<em>float</em>) – The Elastic-Net mixing parameter, with 0 &lt;= l1_ratio &lt;= 1.
Only used if penalty = <a href="#id3"><span class="problematic" id="id4">`</span></a>elasticnetV. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Logistic Regression model initialize with arguments</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.model.LogisticRegression" title="src.model.LogisticRegression">LogisticRegression</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.model_neural_network">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">model_neural_network</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span></em>, <em class="sig-param"><span class="n">hidden_layer_sizes</span><span class="o">=</span><span class="default_value">100</span></em>, <em class="sig-param"><span class="n">activation</span><span class="o">=</span><span class="default_value">'relu'</span></em>, <em class="sig-param"><span class="n">solver</span><span class="o">=</span><span class="default_value">'adam'</span></em>, <em class="sig-param"><span class="n">alpha</span><span class="o">=</span><span class="default_value">0.0001</span></em>, <em class="sig-param"><span class="n">batch_size</span><span class="o">=</span><span class="default_value">'auto'</span></em>, <em class="sig-param"><span class="n">learning_rate</span><span class="o">=</span><span class="default_value">'constant'</span></em>, <em class="sig-param"><span class="n">learning_rate_init</span><span class="o">=</span><span class="default_value">0.001</span></em>, <em class="sig-param"><span class="n">power_t</span><span class="o">=</span><span class="default_value">0.5</span></em>, <em class="sig-param"><span class="n">max_iter</span><span class="o">=</span><span class="default_value">200</span></em>, <em class="sig-param"><span class="n">shuffle</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">random_state</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">tol</span><span class="o">=</span><span class="default_value">0.0001</span></em>, <em class="sig-param"><span class="n">verbose</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">warm_start</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">momentum</span><span class="o">=</span><span class="default_value">0.9</span></em>, <em class="sig-param"><span class="n">nesterovs_momentum</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">early_stopping</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">validation_fraction</span><span class="o">=</span><span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">beta_1</span><span class="o">=</span><span class="default_value">0.9</span></em>, <em class="sig-param"><span class="n">beta_2</span><span class="o">=</span><span class="default_value">0.999</span></em>, <em class="sig-param"><span class="n">epsilon</span><span class="o">=</span><span class="default_value">1e-08</span></em>, <em class="sig-param"><span class="n">n_iter_no_change</span><span class="o">=</span><span class="default_value">1e-08</span></em>, <em class="sig-param"><span class="n">max_fun</span><span class="o">=</span><span class="default_value">15000</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#model_neural_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.model_neural_network" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a Neural Network algorithm from scikit-learn library.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – Name of the model.</p></li>
<li><p><strong>hidden_layer_sizes</strong> (<em>tuple</em><em>, </em><em>length = n_layers - 2</em>) – The ith element represents the number of neurons in the
ith hidden layer. Defaults to <cite>(100,)</cite>.</p></li>
<li><p><strong>activation</strong> ({<cite>identity</cite>, <cite>logistic</cite>, <cite>tanh</cite>, <cite>relu</cite>}) – Activation function for the hidden layer. Defaults to <cite>relu</cite>.</p></li>
<li><p><strong>solver</strong> ({<cite>lbfgs</cite>, <cite>sgd</cite>, <cite>adam</cite>}) – The solver for weight optimization. Defaults to <cite>adam</cite>.</p></li>
<li><p><strong>alpha</strong> (<em>float</em>) – Strength of the L2 regularization term (which is divided by the sample
size when added to the loss). Defaults to <cite>0.0001</cite>.</p></li>
<li><p><strong>batch_size</strong> (<em>int</em><em>, </em><em>str</em>) – Size of minibatches for stochastic optimizers. Defaults to <cite>auto</cite>.</p></li>
<li><p><strong>learning_rate</strong> ({<cite>constant</cite>, <cite>invscaling</cite>, <cite>adaptive</cite>}) – Learning rate schedule for weight purposes. Defaults to <cite>constant</cite>.</p></li>
<li><p><strong>learning_rate_init</strong> (<em>float</em>) – The initial learning rate used. It controls the step-size
in updating the weights. Only when solver is <cite>sgd</cite> or <cite>adam</cite>. Defaults to <cite>0.001</cite>.</p></li>
<li><p><strong>power_t</strong> (<em>float</em>) – The exponent for inverse scaling learning rate. It is used in updating
effective learning rate when the <cite>learning_rate</cite> is <cite>invscaling</cite>. Only when solver
is <cite>sgd</cite>. Defaults to <cite>0.5</cite></p></li>
<li><p><strong>max_iter</strong> (<em>int</em>) – Maximum number of iterations. Defaults to <cite>200</cite>.</p></li>
<li><p><strong>shuffle</strong> (<em>bool</em>) – Whether to shuffle samples in each iteration. Only when solver is <cite>sgd</cite>
or <cite>adam</cite>. Defaults to <cite>True</cite>.</p></li>
<li><p><strong>random_state</strong> (<em>int</em><em>, </em><em>RandomInstance</em><em>, </em><em>None</em>) – Determines random number generation for weights and bias initialization,
train-test split and batch sampling. Defaults to None.</p></li>
</ul>
</dd>
</dl>
<p>:param tol:Tolerance for stopping criteria. Defaults to <cite>1e-4</cite>.
:type tol: float
:param verbose: For the liblinear and lbfgs solvers set verbose to any positive</p>
<blockquote>
<div><p>number for verbosity.  Defaults to <cite>False</cite>.</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>warm_start</strong> (<em>bool</em>) – If <cite>True</cite>, reuse the solution of the previous call to fit
as initialization, otherwise, just erase the previous solution. Defauls to <cite>False</cite>.</p></li>
<li><p><strong>momentum</strong> (<em>float</em>) – Momentum for gradient descent update. Should be between 0 and 1.
Defaults to <cite>0.9</cite>.</p></li>
<li><p><strong>nesterovs_momentum</strong> (<em>bool</em>) – Wheter to use Nesterov’s momentum. Only when solves is <cite>sgd</cite>
and momentum &gt; 0. Defaults to <cite>True</cite>.</p></li>
<li><p><strong>early_stopping</strong> (<em>bool</em>) – Whether to use early stopping to terminate training when
validation score is not improving. Defaults to <cite>False</cite>.</p></li>
<li><p><strong>validation_fraction</strong> (<em>float</em>) – The proportion of training data to set aside as validation
set fot early stopping. Must be between 0 and 1. Only if early stopping is <cite>True</cite>.
Defaults to <cite>0.1</cite>.</p></li>
<li><p><strong>beta_1</strong> (<em>float</em>) – Exponential decay rate for estimates of first moment vector in adam,
should be between 0 and 1. Only used when solver is <cite>adam</cite>. Defaults to <cite>0.9</cite>.</p></li>
<li><p><strong>beta_2</strong> (<em>float</em>) – Exponential decay rate for estimates of second moment vector in adam,
should be between 0 and 1. Only used when solver is <cite>adam</cite>. Defaults to <cite>0.999</cite>.</p></li>
<li><p><strong>epsilon</strong> (<em>float</em>) – Value for numerical stability in adam. Only used when solver is <cite>adam</cite>.
Defaults to <cite>1e-8</cite>.</p></li>
<li><p><strong>n_iter_no_change</strong> (<em>int</em>) – Maximum number of epoch to not meet tol improvement.
Only when solver is <cite>sgd</cite> or <cite>adam</cite>. Defaults to <cite>10</cite>.</p></li>
<li><p><strong>max_fun</strong> (<em>int</em>) – Maximum number of loss functions calls. Only when solver is <cite>lbfgs</cite>.
Defaults to <cite>15000</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Neural Network algorithm</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.model.NeuralNetwork" title="src.model.NeuralNetwork">NeuralNetwork</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.model_random_forest">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">model_random_forest</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span></em>, <em class="sig-param"><span class="n">n_estimators</span><span class="o">=</span><span class="default_value">100</span></em>, <em class="sig-param"><span class="n">criterion</span><span class="o">=</span><span class="default_value">'gini'</span></em>, <em class="sig-param"><span class="n">max_depth</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">min_samples_split</span><span class="o">=</span><span class="default_value">2</span></em>, <em class="sig-param"><span class="n">min_samples_leaf</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">min_weight_fraction_leaf</span><span class="o">=</span><span class="default_value">0.0</span></em>, <em class="sig-param"><span class="n">max_features</span><span class="o">=</span><span class="default_value">'sqrt'</span></em>, <em class="sig-param"><span class="n">max_leaf_nodes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">min_impurity_decrease</span><span class="o">=</span><span class="default_value">0.0</span></em>, <em class="sig-param"><span class="n">bootstrap</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">oob_score</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">n_jobs</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">random_state</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">verbose</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">warm_start</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">class_weight</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ccp_alpha</span><span class="o">=</span><span class="default_value">0.0</span></em>, <em class="sig-param"><span class="n">max_samples</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#model_random_forest"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.model_random_forest" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a Random Forest classifier algorithm from scikit-learn library.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – Name of the model.</p></li>
<li><p><strong>n_estimators</strong> (<em>int</em>) – Number of trees in the forest. Defaults to <cite>100</cite>.</p></li>
<li><p><strong>criterion</strong> ({<cite>gini</cite>, <cite>entropy</cite>, <cite>log_loss</cite>}.) – Function to measure the quality of a split. Defaults to <cite>gini</cite>.</p></li>
<li><p><strong>max_depth</strong> (<em>int</em><em>, </em><em>None</em>) – The maximum depth of the tree. If None, nodes are expanded
until all leaves are pure or contains less than <cite>min_sample_split</cite> samples.
Defaults to None.</p></li>
<li><p><strong>min_samples_split</strong> (<em>int</em><em> or </em><em>float</em>) – The minimum number of samples required to split an
internal node. Defaults to <cite>2</cite>.</p></li>
<li><p><strong>min_samples_leaf</strong> (<em>int</em><em> or </em><em>float</em>) – The minimum number of samples required to be at a leaf node.
Defaults to <cite>1</cite>.</p></li>
<li><p><strong>min_weight_fraction_leaf</strong> (<em>float</em>) – The minimum weighted fraction of the sum total
of weights (of all the input samples) required to be at a leaf node.
Defaults to <cite>0.0</cite>.</p></li>
<li><p><strong>max_features</strong> – The number of features to consider when looking for
the best split. Defaults to <cite>sqrt</cite>.</p></li>
</ul>
</dd>
</dl>
<p>:type max_features:{<cite>sqrt</cite>, <cite>log2</cite>, None}, int or float
:param max_leaf_nodes: Grow trees with <cite>max_leaf_nodes</cite> in best-first fashion.</p>
<blockquote>
<div><p>Defaults to None.</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>min_impurity_decrease</strong> (<em>float</em>) – A node will be split if this split induces a decrease
of the impurity greater than or equal to this value. Defaults to <cite>0.0</cite>.</p></li>
<li><p><strong>bootstrap</strong> (<em>bool</em>) – Whether bootstrap samples are used when building trees.
If <cite>False</cite>, the whole dataset is used to build each tree. Defauls to <cite>True</cite>.</p></li>
<li><p><strong>oob_score</strong> (<em>bool</em>) – Wheter to use out-of-bag samples to estimate the generalization score.
Only available if boostrap is <cite>True</cite>. Defaults to <cite>False</cite>.</p></li>
<li><p><strong>n_jobs</strong> (<em>int</em>) – Number of CPU cores used when parallelizing over classes of multi_class=`ovr`.
Defaults to None.</p></li>
<li><p><strong>random_state</strong> (<em>int</em><em>, </em><em>RandomState instance</em><em>, or </em><em>None</em>) – Used when solver == <cite>sag</cite>, <cite>saga</cite> or <cite>liblinear</cite> to shuffle the data.
Defaults to None.</p></li>
<li><p><strong>verbose</strong> (<em>int</em>) – For the liblinear and lbfgs solvers set verbose to any positive
number for verbosity. Defaults to <cite>0</cite>.</p></li>
<li><p><strong>warm_start</strong> (<em>bool</em>) – If <cite>True</cite>, reuse the solution of the previous call to fit
as initialization, otherwise, just erase the previous solution. Defauls to <cite>False</cite>.</p></li>
<li><p><strong>class_weight</strong> (dict, <cite>balanced</cite>, <cite>balanced_subsample</cite>, or None) – Weights associated with classes in the form {class_label: weight}.
Defaults to None.</p></li>
<li><p><strong>ccp_alpha</strong> (<em>non-negative float</em>) – Complexity parameter used for Minimal Cost-Complexity Pruning.
Defaults to <cite>0.0</cite>.</p></li>
<li><p><strong>max_samples</strong> (<em>int</em><em>, </em><em>float</em><em>, </em><em>None</em>) – If boostrap is <cite>True</cite>, the number of samples to draw from X to
train each base estimator. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Random Forest classifier algorithm</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.model.RandomForest" title="src.model.RandomForest">RandomForest</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.model_scikit_learn">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">model_scikit_learn</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span></em>, <em class="sig-param"><span class="n">model</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#model_scikit_learn"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.model_scikit_learn" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a generic model that is provided by
the Scikit-learn library.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – Name of the model.</p></li>
<li><p><strong>model</strong> (<em>object</em>) – Model to work with.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a generic ScikitLearn model</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.model.ScikitLearnModel" title="src.model.ScikitLearnModel">ScikitLearnModel</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.model_support_vector_classif">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">model_support_vector_classif</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span></em>, <em class="sig-param"><span class="n">C</span><span class="o">=</span><span class="default_value">1.0</span></em>, <em class="sig-param"><span class="n">kernel</span><span class="o">=</span><span class="default_value">'rbf'</span></em>, <em class="sig-param"><span class="n">degree</span><span class="o">=</span><span class="default_value">3</span></em>, <em class="sig-param"><span class="n">gamma</span><span class="o">=</span><span class="default_value">'scale'</span></em>, <em class="sig-param"><span class="n">coef0</span><span class="o">=</span><span class="default_value">0.0</span></em>, <em class="sig-param"><span class="n">shrinking</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">probability</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">tol</span><span class="o">=</span><span class="default_value">0.001</span></em>, <em class="sig-param"><span class="n">cache_size</span><span class="o">=</span><span class="default_value">200</span></em>, <em class="sig-param"><span class="n">class_weight</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">verbose</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">max_iter</span><span class="o">=</span><span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">decision_function_shape</span><span class="o">=</span><span class="default_value">'ovr'</span></em>, <em class="sig-param"><span class="n">break_ties</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">random_state</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#model_support_vector_classif"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.model_support_vector_classif" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a Support Vector Machine algorithm from scikit-learn library.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – Name of the model.</p></li>
<li><p><strong>C</strong> (<em>float</em>) – Regularization parameter. The strenght of the regularization is
inversely proportional to C. Defaults to <cite>1.0</cite>.</p></li>
<li><p><strong>kernel</strong> ({<cite>linear</cite>, <cite>poly</cite>, <cite>rbf</cite>, <cite>sigmoid</cite>, <cite>precomputed</cite>} or callable) – Specifies the kernel type to be used in the algorithm.
Defaults to <cite>rbf</cite></p></li>
<li><p><strong>degree</strong> (<em>int</em>) – Degree of the polynomial kernel function when kernel is <cite>poly</cite>.
Defaults to <cite>3</cite>.</p></li>
<li><p><strong>gamma</strong> ({<cite>scale</cite>, <cite>auto</cite>} or float) – Kernek coefficient fot <cite>rbf</cite>, <cite>poly</cite> and <cite>sigmoid</cite>.
Defaults to <cite>scale</cite>.</p></li>
<li><p><strong>coef0</strong> (<em>float</em>) – Independent term in kernel function, when kernel is <cite>poly</cite> or
<cite>sigmoid</cite>. Defaults to <cite>0.0</cite>.</p></li>
<li><p><strong>shrinking</strong> (<em>bool</em>) – Whether to use the shrinking heuristic. Defaults to <cite>True</cite>.</p></li>
<li><p><strong>probability</strong> (<em>bool</em>) – Wheter to enable probability estimates. Deafuls to <cite>False</cite>.</p></li>
<li><p><strong>tol</strong> (<em>float</em>) – Tolerance for stopping criteria. Defaults to <cite>1e-3</cite>.</p></li>
<li><p><strong>cache_size</strong> (<em>float</em>) – Specify the size of the kernelk cache (in MB). Defaults to <cite>200</cite>.</p></li>
<li><p><strong>class_weight</strong> (dict, <cite>balanced</cite>, or None) – Weights associated with classes in the form {class_label: weight}.
Defaults to None.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – For the liblinear and lbfgs solvers set verbose to any positive
number for verbosity.  Defaults to None.</p></li>
<li><p><strong>max_iter</strong> (<em>int</em>) – Hard limit on iterations within solver, or -1 for no limit.
Defaults to -1.</p></li>
<li><p><strong>decision_function_shape</strong> ({<cite>ovo</cite>, <cite>ovr</cite>}) – Wheter to return a one-vs-rest (<cite>ovr</cite>) or
one-vs-one (<cite>ovo</cite>) decision function. Defaults to <cite>ovr</cite>.</p></li>
<li><p><strong>break_ties</strong> (<em>bool</em>) – If <cite>True</cite>, <cite>decision_functoin_shape</cite> = <cite>ovr</cite> and number of
classes &gt; 2, predict will break ties according to the confidence values
of <cite>decision_function</cite>. Defaults to <cite>False</cite>.</p></li>
<li><p><strong>random_state</strong> (<em>int</em><em>, </em><em>RandomState instance</em><em>, or </em><em>None</em>) – Used when solver == <cite>sag</cite>, <cite>saga</cite> or <cite>liblinear</cite> to shuffle the data.
Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a support vector machine algorithm</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.model.SupportVectorClassif" title="src.model.SupportVectorClassif">SupportVectorClassif</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.obs_add_column">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">obs_add_column</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">src.dataobject.DataObject</a></span></em>, <em class="sig-param"><span class="n">name_column</span><span class="o">=</span><span class="default_value">''</span></em>, <em class="sig-param"><span class="n">values</span><span class="o">=</span><span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#obs_add_column"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.obs_add_column" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Add to DataObject’s Observations a new column</dt><dd><p>It is not neccesary control that the rest of the data match with new Observations.
Add new column doesn’t affect the rest of the DataObject
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>name_column</strong> (<em>str</em>) – Name of the new column</p></li>
<li><p><strong>values</strong> (<em>list</em><em>, </em><em>ndarray</em>) – values of the new column. Must have the same lenght as number of rows of Observations</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.obs_drop_duplicates">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">obs_drop_duplicates</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">src.dataobject.DataObject</a></span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">keep</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">by_name</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#obs_drop_duplicates"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.obs_drop_duplicates" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Drop DataObject’s Observations duplicates.</dt><dd><p>Control that the rest of the data match with new Observations
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>axis</strong> (<em>int</em>) – axis 0 means by rows, 1 by columns</p></li>
<li><p><strong>keep</strong> (<em>str</em><em>, </em><em>bool</em>) – Indicates which duplicates will keep on Observations.
“first” keeps only the first duplicated column/row; “last” keeps only the last duplicated column/row; False keeps both duplicated columns/rows</p></li>
<li><p><strong>by_name</strong> (<em>bool</em>) – True if only evaluate the name of the column/row, False if evaluate the content of the column/row to decide if is duplicated or not</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.obs_projection">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">obs_projection</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">src.dataobject.DataObject</a></span></em>, <em class="sig-param"><span class="n">col_list</span><span class="o">=</span><span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#obs_projection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.obs_projection" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Set DataObject’s Observations selecting a specific column list.</dt><dd><p>Control that the rest of the data match with new Observations
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>col_list</strong> (<em>list</em>) – A list with the names of the columns to keep</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.obs_replace">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">obs_replace</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">src.dataobject.DataObject</a></span></em>, <em class="sig-param"><span class="n">to_replace</span><span class="o">=</span><span class="default_value">'Unknown'</span></em>, <em class="sig-param"><span class="n">replaced_by</span><span class="o">=</span><span class="default_value">nan</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#obs_replace"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.obs_replace" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Replace DataObject’s Observations values with another.</dt><dd><p>Control that the rest of the data match with new Observations
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>to_replace</strong> (<em>str</em><em>, </em><em>float</em><em>, </em><em>int</em><em>, </em><em>bool</em><em>, </em><em>list</em><em>, </em><em>ndarray</em>) – values to be replaced</p></li>
<li><p><strong>replaced_by</strong> (<em>str</em><em>, </em><em>float</em><em>, </em><em>int</em><em>, </em><em>bool</em>) – new value to replaced the value before</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.obs_selection">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">obs_selection</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">src.dataobject.DataObject</a></span></em>, <em class="sig-param"><span class="n">row_list</span><span class="o">=</span><span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#obs_selection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.obs_selection" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Set DataObject’s Observations selecting a specific row list.</dt><dd><p>Control that the rest of the data match with new Observations
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>row_list</strong> (<em>list</em>) – A list with the names of the rows to keep</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.plot_confusion_matrix">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">plot_confusion_matrix</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">matrix</span></em>, <em class="sig-param"><span class="n">label_list</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">color</span><span class="o">=</span><span class="default_value">'Viridis'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#plot_confusion_matrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.plot_confusion_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a figure with the Confusion Matriz as Heatmap.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>matrix</strong> (<em>list</em><em>[</em><em>list</em><em>]</em>) – Contains the elements and data of a confusion matrix.</p></li>
<li><p><strong>label_list</strong> (<em>list</em>) – Used as value of axis. Defaults to None.</p></li>
<li><p><strong>color</strong> (<em>list</em>) – color spectrum used on heatmap. Defaults to <cite>Viridis</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Figure that contains the heatmap showing the colored confusion matrix.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.plot_prc">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">plot_prc</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">testy</span></em>, <em class="sig-param"><span class="n">predictions</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#plot_prc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.plot_prc" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a Precision-Recall (PR) Curve.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>testy</strong> (<em>list</em>) – Real value of a dataset if  a ML algorithm has been applied.</p></li>
<li><p><strong>predictions</strong> (<em>list</em>) – Result of apply ML on the dataset.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Figure that contains the PR Curve of the Real Value and Predictions.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.plot_prec_recall_vs_thresh">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">plot_prec_recall_vs_thresh</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">testy</span></em>, <em class="sig-param"><span class="n">predictions</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#plot_prec_recall_vs_thresh"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.plot_prec_recall_vs_thresh" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates the curve between Precision-Real and Thresholds.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>testy</strong> (<em>list</em>) – Real value of a dataset if  a ML algorithm has been applied.</p></li>
<li><p><strong>predictions</strong> (<em>list</em>) – Result of apply ML on the dataset.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Figure that contains the plot showing the precision recall curve over thresholds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.plot_roc">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">plot_roc</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">testy</span></em>, <em class="sig-param"><span class="n">predictions</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#plot_roc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.plot_roc" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a ROC Curve.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>testy</strong> (<em>list</em>) – Real value of a dataset if  a ML algorithm has been applied.</p></li>
<li><p><strong>predictions</strong> (<em>list</em>) – Result of apply ML on the dataset.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Figure that contains the ROC Curve of the Real Value and Predictions.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.print_progress_bar">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">print_progress_bar</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">iteration</span></em>, <em class="sig-param"><span class="n">total</span></em>, <em class="sig-param"><span class="n">mb_iter</span></em>, <em class="sig-param"><span class="n">mb_total</span></em>, <em class="sig-param"><span class="n">prefix</span><span class="o">=</span><span class="default_value">'Progress:'</span></em>, <em class="sig-param"><span class="n">suffix</span><span class="o">=</span><span class="default_value">'Complete'</span></em>, <em class="sig-param"><span class="n">length</span><span class="o">=</span><span class="default_value">50</span></em>, <em class="sig-param"><span class="n">decimals</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">fill</span><span class="o">=</span><span class="default_value">'█'</span></em>, <em class="sig-param"><span class="n">printEnd</span><span class="o">=</span><span class="default_value">'\r'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#print_progress_bar"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.print_progress_bar" title="Permalink to this definition">¶</a></dt>
<dd><p>Print a bar to show the progress of the download.</p>
<blockquote>
<div><dl class="field-list simple">
<dt class="field-odd">param iteration</dt>
<dd class="field-odd"><p>Iterator that indicates the current file that it is being downloaded.</p>
</dd>
<dt class="field-even">type iteration</dt>
<dd class="field-even"><p>int</p>
</dd>
<dt class="field-odd">param total</dt>
<dd class="field-odd"><p>Total number of files to be downloaded.</p>
</dd>
<dt class="field-even">type total</dt>
<dd class="field-even"><p>int</p>
</dd>
<dt class="field-odd">param mb_iter</dt>
<dd class="field-odd"><p>Size of the current file in MB.</p>
</dd>
<dt class="field-even">type mb_iter</dt>
<dd class="field-even"><p>float</p>
</dd>
<dt class="field-odd">param mb_total</dt>
<dd class="field-odd"><p>Total size of the project in MB.</p>
</dd>
<dt class="field-even">type mb_total</dt>
<dd class="field-even"><p>float</p>
</dd>
<dt class="field-odd">param prefix</dt>
<dd class="field-odd"><p>Message to appear next to the amount of current size downloaded.</p>
</dd>
<dt class="field-even">type prefix</dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">param suffix</dt>
<dd class="field-odd"><p>Message to appear next to the completed part.</p>
</dd>
<dt class="field-even">type suffix</dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">param length</dt>
<dd class="field-odd"><p>Total length of the progress bar. Defaults to 50.</p>
</dd>
<dt class="field-even">type length</dt>
<dd class="field-even"><p>int</p>
</dd>
<dt class="field-odd">param decimals</dt>
<dd class="field-odd"><p>Integer value that indicated the percentage of downloaded data.
Defaults to 1.</p>
</dd>
<dt class="field-even">type decimals</dt>
<dd class="field-even"><p>int</p>
</dd>
<dt class="field-odd">param fill</dt>
<dd class="field-odd"><p>Charcater to print the progress of the bar. Defaults to ‘█’.</p>
</dd>
<dt class="field-even">type fill</dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">param printEnd</dt>
<dd class="field-odd"><p>Character to print at the end of the iteration.
Defaults to “</p>
</dd>
</dl>
</div></blockquote>
<dl>
<dt>” (carriage return). A carriage return means moving</dt><dd><blockquote>
<div><p>the cursor to the beginning of the line.</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">type printEnd</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.print_request">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">print_request</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">json_obj</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#print_request"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.print_request" title="Permalink to this definition">¶</a></dt>
<dd><p>Decodes a JSON object and display it in an elegant format per screen,</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>json_obj</strong> (<em>JSON Object</em>) – JSON object to print.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.read_file">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">read_file</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file</span></em>, <em class="sig-param"><span class="n">index</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">sep</span><span class="o">=</span><span class="default_value">'\t'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#read_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.read_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a DataFrame with the read file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file</strong> (<em>str</em>) – File to read</p></li>
<li><p><strong>index</strong> (<em>str</em>) – Value of the index of the DataFrame.</p></li>
<li><p><strong>sep</strong> (<em>str</em>) – Separator character between fields. Defaults to ”   ” (Tab-separated values).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame with the file read.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.read_genecode_csv">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">read_genecode_csv</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_csv</span><span class="o">=</span><span class="default_value">'../genecode_v36.csv'</span></em>, <em class="sig-param"><span class="n">sep</span><span class="o">=</span><span class="default_value">','</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#read_genecode_csv"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.read_genecode_csv" title="Permalink to this definition">¶</a></dt>
<dd><p>Read a CSV file that contains information about genes
and returns a DataFrame to work with it</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_csv</strong> (<em>str</em>) – Path where the genecode on CSV format is saved</p></li>
<li><p><strong>sep</strong> (<em>str</em>) – Character that delimit its structure</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A DataFrame with Gene Information</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.recursive_delete_data">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">recursive_delete_data</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dir_to_delete</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#recursive_delete_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.recursive_delete_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Search in the directory tree the input directory, so it remove its
content recursively before remove the root directory.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>dir_to_delete</strong> (<em>str</em>) – Directory path to delete its content.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.remove_correlation">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">remove_correlation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">thresh</span><span class="o">=</span><span class="default_value">0.0</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#remove_correlation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.remove_correlation" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove the elements that has high correlation</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dataframe with high correlation removed</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.remove_low_reads">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">remove_low_reads</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">thresh</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">al_least</span><span class="o">=</span><span class="default_value">2</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#remove_low_reads"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.remove_low_reads" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the best n variables of a DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>thresh</strong> (<em>float</em>) – threshold that indicates the limit to remove feature less than it</p></li>
<li><p><strong>at_least</strong> (<em>int</em>) – Indicates the minimum match to pass the selection</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dataframe with removed features which have less then ‘threshold’ reads all columns</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.rename_columname">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">rename_columname</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">pos_ini</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">pos_fin</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#rename_columname"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.rename_columname" title="Permalink to this definition">¶</a></dt>
<dd><p>Rename all names of the columns triping the name</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>pos_ini</strong> (<em>int</em>) – position of the first character</p></li>
<li><p><strong>pos_fin</strong> (<em>int</em>) – position of the last character</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A DataFrame with the column renamed with a substring of the inner column name</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.rename_index">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">rename_index</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">new_name</span><span class="o">=</span><span class="default_value">''</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#rename_index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.rename_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Rename the name of the current index of DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>new_name</strong> (<em>str</em>) – name that will replace current one</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The same DataFrame with the index changed</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.replace">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">replace</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">to_replace</span><span class="o">=</span><span class="default_value">'Unknown'</span></em>, <em class="sig-param"><span class="n">replaced_by</span><span class="o">=</span><span class="default_value">nan</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#replace"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.replace" title="Permalink to this definition">¶</a></dt>
<dd><p>Replace DataFrame’s selected values with another.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>to_replace</strong> (<em>str</em><em>, </em><em>float</em><em>, </em><em>int</em><em>, </em><em>bool</em><em>, </em><em>list</em><em>, </em><em>ndarray</em>) – values to be replaced</p></li>
<li><p><strong>replaced_by</strong> (<em>str</em><em>, </em><em>float</em><em>, </em><em>int</em><em>, </em><em>bool</em>) – new value to replaced the value before</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a DataFrame with all values indicates replaced by other value</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.replicate_workflow">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">replicate_workflow</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">workflow</span></em>, <em class="sig-param"><span class="n">data_object</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#replicate_workflow"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.replicate_workflow" title="Permalink to this definition">¶</a></dt>
<dd><p>Replicate a workflow to obtain the same result as previous experiment</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>workflow</strong> (<a class="reference internal" href="src.html#src.objects.Workflow" title="src.objects.Workflow"><em>Workflow</em></a>) – a flow of another object that contains step of its experiment</p></li>
<li><p><strong>data_object</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – a DataObject to which the same process will be applied through the Workflow</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a DataObject</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.rna_gen_length">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">rna_gen_length</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">genemodel</span></em>, <em class="sig-param"><span class="n">gen</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#rna_gen_length"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.rna_gen_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns lenght of a specific gen indicates</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genemodel</strong> (<em>DataFrame</em>) – a DataFrame with genemodel information</p></li>
<li><p><strong>gen_id</strong> (<em>str</em>) – identifier of the gene on a genemodel</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the lenght of a specific gen selected</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.rna_genes_lenght">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">rna_genes_lenght</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">genemodel</span></em>, <em class="sig-param"><span class="n">colnm_lenght</span><span class="o">=</span><span class="default_value">'width'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#rna_genes_lenght"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.rna_genes_lenght" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a DataFrame with the genemodel information, return all lenght of the actual studied genes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genemodel</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>colnm_lenght</strong> (<em>str</em>) – name of the column that contains information about their lenght</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a dataframe with lenghts of the genes</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.rna_rpk">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">rna_rpk</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em>, <em class="sig-param"><span class="n">genemodel</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#rna_rpk"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.rna_rpk" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalize values of DataFrame using RPK method</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>genemodel</strong> (<em>DataFrame</em>) – a DataFrame with gene information, like gene lenght</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a DataFrame with computed RPK</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.rna_rpkm">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">rna_rpkm</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em>, <em class="sig-param"><span class="n">genemodel</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#rna_rpkm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.rna_rpkm" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalize values of DataFrame using RPKM method</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>genemodel</strong> (<em>DataFrame</em>) – a DataFrame with gene information, like gene lenght</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a DataFrame with computed RPKM</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.rna_sample_condition">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">rna_sample_condition</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">sep</span><span class="o">=</span><span class="default_value">'-'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#rna_sample_condition"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.rna_sample_condition" title="Permalink to this definition">¶</a></dt>
<dd><p>Classify a given sample analyze its barcode</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – The current str to be process</p></li>
<li><p><strong>sep</strong> (<em>str</em>) – indicates with character is used to separate values</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>classification of the sample by its barcode passed on input</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.rna_stat_test">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">rna_stat_test</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em>, <em class="sig-param"><span class="n">grouped_by</span></em>, <em class="sig-param"><span class="n">group_1</span></em>, <em class="sig-param"><span class="n">group_2</span></em>, <em class="sig-param"><span class="n">clinical_data</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#rna_stat_test"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.rna_stat_test" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a DataFrame with several statistic information like FoldChange, T, P-Value and a DiffExpr</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>grouped_by</strong> (<em>str</em>) – name of the column of clinical data to organize data</p></li>
<li><p><strong>group_1</strong> (<em>str</em>) – one of the type to create a group</p></li>
<li><p><strong>group_2</strong> (<em>str</em>) – another type to create a group</p></li>
<li><p><strong>clinical_data</strong> (<em>DataFrame</em>) – a dataframe with clinical information of the data</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dataframe with several stadistic information: FoldChange, T, P-Value and a DiffExpr</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.rna_tpm">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">rna_tpm</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em>, <em class="sig-param"><span class="n">genemodel</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#rna_tpm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.rna_tpm" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalize values of DataFrame using TPM method</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>genemodel</strong> (<em>DataFrame</em>) – a DataFrame with gene information, like gene lenght</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a DataFrame with computed TPM</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.rownames">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">rownames</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#rownames"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.rownames" title="Permalink to this definition">¶</a></dt>
<dd><p>Return on a list the name of the rows of the DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>DataFrame</em>) – a dataframe to obtain the name of its rows</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list with the row names of the data</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.save_image">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">save_image</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">fig</span></em>, <em class="sig-param"><span class="n">fig_name</span><span class="o">=</span><span class="default_value">'fig'</span></em>, <em class="sig-param"><span class="n">img_format</span><span class="o">=</span><span class="default_value">'png'</span></em>, <em class="sig-param"><span class="n">os_path</span><span class="o">=</span><span class="default_value">'images'</span></em>, <em class="sig-param"><span class="n">xlabel</span><span class="o">=</span><span class="default_value">'x axis'</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="o">=</span><span class="default_value">'y axis'</span></em>, <em class="sig-param"><span class="n">x_ini_range</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">x_fin_range</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">y_ini_range</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">y_fin_range</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">legend</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">legend_title</span><span class="o">=</span><span class="default_value">''</span></em>, <em class="sig-param"><span class="n">vheight</span><span class="o">=</span><span class="default_value">600</span></em>, <em class="sig-param"><span class="n">vwidth</span><span class="o">=</span><span class="default_value">800</span></em>, <em class="sig-param"><span class="n">title</span><span class="o">=</span><span class="default_value">'Plot of the figure'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#save_image"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.save_image" title="Permalink to this definition">¶</a></dt>
<dd><p>Save a figure as image on disk.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fig</strong> (<em>Figure</em>) – Figure to update.</p></li>
<li><p><strong>fig_name</strong> (<em>str</em>) – Name of file.</p></li>
<li><p><strong>img_format</strong> (<em>str</em>) – Format of the file saved. Available formats: ‘png’, ‘jpeg’,
‘webp’, ‘svg’, ‘pdf’.</p></li>
<li><p><strong>os_path</strong> (<em>str</em>) – Path to save the image.</p></li>
<li><p><strong>xlabel</strong> (<em>str</em>) – Name of the X axis. Defaults to <cite>x axis</cite>.</p></li>
<li><p><strong>ylabel</strong> (<em>str</em>) – Name of the Y axis. Defaults to <cite>y axis</cite>.</p></li>
<li><p><strong>x_ini_range</strong> (<em>int</em>) – Initial value of X axis range. Defaults to None.</p></li>
<li><p><strong>x_fin_range</strong> (<em>int</em>) – Final value of X axis range. Defaults to None.</p></li>
<li><p><strong>y_ini_range</strong> (<em>int</em>) – Initial value of Y axis range. Defaults to None.</p></li>
<li><p><strong>y_fin_range</strong> (<em>int</em>) – Final value of Y axis range. Defaults to None.</p></li>
<li><p><strong>legend</strong> (<em>bool</em>) – Indicates if the figure will have legend or not. Defaults to <cite>True</cite>.</p></li>
<li><p><strong>legend_title</strong> (<em>str</em>) – Name of the legend’s label</p></li>
<li><p><strong>vheight</strong> (<em>int</em>) – Height of the figure as image. Defaults to 600.</p></li>
<li><p><strong>vwidth</strong> (<em>int</em>) – Width of the figure as image. Defaults to 800.</p></li>
<li><p><strong>title</strong> (<em>str</em>) – Title of the figure. Defaults to <cite>Plot of the figure</cite>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.server_check_status">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">server_check_status</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">endpt_status</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#server_check_status"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.server_check_status" title="Permalink to this definition">¶</a></dt>
<dd><p>Check the status of the server.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>endpt_status</strong> (<em>str</em>) – URL (endpoint) to be checked.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.server_status_code">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">server_status_code</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">stat_code</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#server_status_code"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.server_status_code" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the status response from the server</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>stat_code</strong> (<em>int</em>) – Current status of the checked URL endpoint.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Message explaining the obtained server’s code.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.set_counts">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">set_counts</code><span class="sig-paren">(</span><em class="sig-param">dobject: src.dataobject.DataObject</em>, <em class="sig-param">new_counts=Empty DataFrame Columns: [] Index: []</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#set_counts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.set_counts" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Set DataObject’s Count Matrix to new one.</dt><dd><p>Control that the rest of the data match with new Count Matrix
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>new_counts</strong> (<em>DataFrame</em>) – new Count Matrix</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.set_obs">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">set_obs</code><span class="sig-paren">(</span><em class="sig-param">dobject: src.dataobject.DataObject</em>, <em class="sig-param">new_obs=Empty DataFrame Columns: [] Index: []</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#set_obs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.set_obs" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Set DataObject’s Clinical Information to new one.</dt><dd><p>Control that the rest of the data match with new Observations
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>new_obs</strong> (<em>DataFrame</em>) – new Clinical Information</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.set_vars">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">set_vars</code><span class="sig-paren">(</span><em class="sig-param">dobject: src.dataobject.DataObject</em>, <em class="sig-param">new_var=Empty DataFrame Columns: [] Index: []</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#set_vars"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.set_vars" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Set DataObject’s Variables to new one.</dt><dd><p>Control that the rest of the data match with new Variables
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>new_var</strong> (<em>DataFrame</em>) – new Gene Information</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.show_figure">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">show_figure</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">fig</span></em>, <em class="sig-param"><span class="n">xlabel</span><span class="o">=</span><span class="default_value">'x axis'</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="o">=</span><span class="default_value">'y axis'</span></em>, <em class="sig-param"><span class="n">x_ini_range</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">x_fin_range</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">y_ini_range</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">y_fin_range</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">legend</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">legend_title</span><span class="o">=</span><span class="default_value">''</span></em>, <em class="sig-param"><span class="n">vheight</span><span class="o">=</span><span class="default_value">600</span></em>, <em class="sig-param"><span class="n">vwidth</span><span class="o">=</span><span class="default_value">800</span></em>, <em class="sig-param"><span class="n">title</span><span class="o">=</span><span class="default_value">'Plot of the figure'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#show_figure"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.show_figure" title="Permalink to this definition">¶</a></dt>
<dd><p>Show in the screen the figure plot with the specified figured parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fig</strong> (<em>Figure</em>) – Figure to update.</p></li>
<li><p><strong>xlabel</strong> (<em>str</em>) – Name of the X axis. Defaults to <cite>x axis</cite>.</p></li>
<li><p><strong>ylabel</strong> (<em>str</em>) – Name of the Y axis. Defaults to <cite>y axis</cite>.</p></li>
<li><p><strong>x_ini_range</strong> (<em>int</em>) – Initial value of X axis range. Defaults to None.</p></li>
<li><p><strong>x_fin_range</strong> (<em>int</em>) – Final value of X axis range. Defaults to None.</p></li>
<li><p><strong>y_ini_range</strong> (<em>int</em>) – Initial value of Y axis range. Defaults to None.</p></li>
<li><p><strong>y_fin_range</strong> (<em>int</em>) – Final value of Y axis range. Defaults to None.</p></li>
<li><p><strong>legend</strong> (<em>bool</em>) – Indicates if the figure will have legend or not. Defaults to <cite>True</cite>.</p></li>
<li><p><strong>legend_title</strong> (<em>str</em>) – Name of the legend’s label</p></li>
<li><p><strong>vheight</strong> (<em>int</em>) – Height of the figure as image. Defaults to 600.</p></li>
<li><p><strong>vwidth</strong> (<em>int</em>) – Width of the figure as image. Defaults to 800.</p></li>
<li><p><strong>title</strong> (<em>str</em>) – Title of the figure. Defaults to <cite>Plot of the figure</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The figure with the layout’s features updated.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.show_image_web">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">show_image_web</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">fig</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#show_image_web"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.show_image_web" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Show an image on a web browser. If there is not an available web browser,</dt><dd><p>it shows the image on the default application to visualize images.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>fig</strong> (<em>Figure</em>) – Figure to dislay.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.thresh_by_perc">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">thresh_by_perc</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em>, <em class="sig-param"><span class="n">perc</span><span class="o">=</span><span class="default_value">10.0</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#thresh_by_perc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.thresh_by_perc" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate a limit by percentage by number of rows or columns</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – the dataframe to calculate that limit according to number of rows or columns.</p></li>
<li><p><strong>perc</strong> (<em>float</em>) – portion of total elements (0.00 to 100.0).</p></li>
<li><p><strong>axis</strong> (<em>int</em>) – Axis 0 means by rows, 1 by columns.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the mininum number of elements that consider a limit.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.top_variables">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">top_variables</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">n_var</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#top_variables"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.top_variables" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the best n variables of a DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>n_var</strong> (<em>int</em>) – Number of variables selected as the best</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dataframe with n_var rows that are the best</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.transpose">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">transpose</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">index_name</span><span class="o">=</span><span class="default_value">''</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#transpose"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.transpose" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Do transpose to a DataFrame keeping the name of columns and index</dt><dd><p>without multi-index</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p></li>
<li><p><strong>index_name</strong> (<em>str</em><em>, </em><em>int</em>) – the name of the index before transpose to keep it</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a DataFrame with transpose of the input taking into account the extra row that it is created after the transpose</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.update_counts">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">update_counts</code><span class="sig-paren">(</span><em class="sig-param">dobject: src.dataobject.DataObject</em>, <em class="sig-param">new_counts=Empty DataFrame Columns: [] Index: []</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#update_counts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.update_counts" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Set DataObject’s Count Matrix to new one but only modify its values.</dt><dd><p>In this case, the function not control that the rest of the data match with new Count Matrix
because any row or column will be modify. Will have the same shape
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>new_counts</strong> (<em>DataFrame</em>) – new Count Matrix</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.var_correlation">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">var_correlation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#var_correlation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.var_correlation" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates correlation between variables of a DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dataframe with the correlation between the variables
with values from 0.0 to 1.0, being 1.0 the highest value of correlation</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.variance">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">variance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#variance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.variance" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the variance of a matrix</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>DataFrame</em>) – The current DataFrame to be process</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dictionary with the variance of each variable</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.vars_add_column">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">vars_add_column</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">src.dataobject.DataObject</a></span></em>, <em class="sig-param"><span class="n">name_column</span><span class="o">=</span><span class="default_value">''</span></em>, <em class="sig-param"><span class="n">values</span><span class="o">=</span><span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#vars_add_column"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.vars_add_column" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Add to DataObject’s Variables a new column</dt><dd><p>It is not neccesary control that the rest of the data match with new Variables.
Add new column doesn’t affect the rest of the DataObject
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>name_column</strong> (<em>str</em>) – Name of the new column</p></li>
<li><p><strong>values</strong> (<em>list</em><em>, </em><em>ndarray</em>) – values of the new column. Must have the same lenght as number of rows of Variables</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.vars_drop_duplicates">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">vars_drop_duplicates</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">src.dataobject.DataObject</a></span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">keep</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">by_name</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#vars_drop_duplicates"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.vars_drop_duplicates" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Drop DataObject’s Variables duplicates.</dt><dd><p>Control that the rest of the data match with new Variables
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>axis</strong> (<em>int</em>) – axis 0 means by rows, 1 by columns</p></li>
<li><p><strong>keep</strong> (<em>str</em><em>, </em><em>bool</em>) – Indicates which duplicates will keep on Observations.
“first” keeps only the first duplicated column/row; “last” keeps only the last duplicated column/row; False keeps both duplicated columns/rows</p></li>
<li><p><strong>by_name</strong> (<em>bool</em>) – True if only evaluate the name of the column/row, False if evaluate the content of the column/row to decide if is duplicated or not</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.vars_projection">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">vars_projection</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">src.dataobject.DataObject</a></span></em>, <em class="sig-param"><span class="n">col_list</span><span class="o">=</span><span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#vars_projection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.vars_projection" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Set DataObject’s Variables selecting a specific column list.</dt><dd><p>Control that the rest of the data match with new Variables
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>col_list</strong> (<em>list</em>) – A list with the names of the columns to keep</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.vars_replace">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">vars_replace</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">src.dataobject.DataObject</a></span></em>, <em class="sig-param"><span class="n">to_replace</span><span class="o">=</span><span class="default_value">'Unknown'</span></em>, <em class="sig-param"><span class="n">replaced_by</span><span class="o">=</span><span class="default_value">nan</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#vars_replace"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.vars_replace" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Replace DataObject’s Variables values with another.</dt><dd><p>Control that the rest of the data match with new Variables
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>to_replace</strong> (<em>str</em><em>, </em><em>float</em><em>, </em><em>int</em><em>, </em><em>bool</em><em>, </em><em>list</em><em>, </em><em>ndarray</em>) – values to be replaced</p></li>
<li><p><strong>replaced_by</strong> (<em>str</em><em>, </em><em>float</em><em>, </em><em>int</em><em>, </em><em>bool</em>) – new value to replaced the value before</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.vars_selection">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">vars_selection</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dobject</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">src.dataobject.DataObject</a></span></em>, <em class="sig-param"><span class="n">row_list</span><span class="o">=</span><span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#vars_selection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.vars_selection" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Set DataObject’s Variables selecting a specific row list.</dt><dd><p>Control that the rest of the data match with new Variables
This class is a Task which allows generate a Workflow
When this class is called, save his instance on DataObject’s Log</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dobject</strong> (<a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject"><em>DataObject</em></a>) – The current DataObject to be process</p></li>
<li><p><strong>row_list</strong> (<em>list</em>) – A list with the names of the rows to keep</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The DataObject passed with the Task applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="src.html#src.dataobject.DataObject" title="src.dataobject.DataObject">DataObject</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.volcano_plot">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">volcano_plot</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em>, <em class="sig-param"><span class="n">clinical_data</span></em>, <em class="sig-param"><span class="n">grouped_by</span></em>, <em class="sig-param"><span class="n">group_1</span></em>, <em class="sig-param"><span class="n">group_2</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#volcano_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.volcano_plot" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Creates a volcano plot with the data received as the input.</dt><dd><p>To do a volcano plot is necessary to indicates two types of information
collected in clinical data.
It will be neccesary calculated a stadistic information</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – The data used to create the plot.</p></li>
<li><p><strong>clinical_data</strong> (<em>DataFrame</em>) – A dataframe with clinical information of the count matrix</p></li>
<li><p><strong>grouped_by</strong> (<em>str</em>) – Name of the clinical data column to organize the plot</p></li>
<li><p><strong>group_1</strong> (<em>str</em>) – A type of the clinical data column selected</p></li>
<li><p><strong>group_2</strong> (<em>str</em>) – A second type of the clinical data column selected</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Figure that contains the volcano plot.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.workflow_from_json">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">workflow_from_json</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path</span><span class="o">=</span><span class="default_value">''</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#workflow_from_json"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.workflow_from_json" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Open a json file with some Log saved as dictionary</dt><dd><p>and creates a new Workflow throught them</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>path</strong> – The file that contains object Task as dictionary for lines</p>
</dd>
<dt class="field-even">Type</dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a replicate Workflow from json dictionary</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="src.html#src.objects.Workflow" title="src.objects.Workflow">Workflow</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geneflow.geneflow.workflow_to_json">
<code class="sig-prename descclassname">geneflow.geneflow.</code><code class="sig-name descname">workflow_to_json</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">workfw</span><span class="p">:</span> <span class="n"><a class="reference internal" href="src.html#src.objects.Workflow" title="src.objects.Workflow">src.objects.Workflow</a></span></em>, <em class="sig-param"><span class="n">path</span><span class="o">=</span><span class="default_value">''</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geneflow/geneflow.html#workflow_to_json"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geneflow.geneflow.workflow_to_json" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate an external file to save every Log object of Workflow</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>path</strong> – file where dictionary generates will be saved</p>
</dd>
<dt class="field-even">Type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="GeneFlow: Tool to analyze gene expression data" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="modules.html" class="btn btn-neutral float-right" title="&lt;no title&gt;" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Alba Casillas Rodríguez.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>  

<style>
    .wy-nav-content-wrap {
        height: 100%;
        width: 100%;
        background: hsl(0, 0%, 100%);
        display: block;
    }

    .wy-nav-content {
        height: 100%;
        width: 100%;
        background: hsl(0, 0%, 100%);
    }

    .rst-content {
        width: 100%;
        height: 100%;
    }

    /* Sidebar header (and topbar for mobile) */
    .wy-side-nav-search,
    .wy-nav-top {
        background: hsl(0, 2%, 10%);
    }

    /* Sidebar */
    .wy-nav-side {
        background: #85e4a5;
    }

    .caption {
        background: hsl(0, 0%, 100%);
    }

    .caption-text {
        color: hsl(0, 0%, 0%)
    }

    .wy-menu-vertical a {
        color: hsl(0, 0%, 0%)
    }

    .wy-menu-vertical a:hover {
        color: hsl(0, 0%, 100%)
    }
</style>


</body>
</html>